---
title: "expression_tree"
author: "Matthew Davis"
date: "2024-12-02"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

# Setup
## Libraries
```{r, message=FALSE}
library(polymorphology2)
library(tidyverse)
library(phangorn)
library(scales)
```

## Extra functions
```{r}
parse_columns <- function(table, sample){
  
  # Parse the columns
  parse.table <- table %>%
    mutate(rna = sub("rna-([^=]+).*", "\\1", target_id),
           gene = sub(".*gene=([^=]+).*", "\\1", target_id),
           name = sub(".*name=([^=]+).*", "\\1", target_id),
           seq_id = sub(".*seq_id=([^=]+).*", "\\1", target_id),
           type = sub(".*type=([^=]+)", "\\1", target_id),
           rna = gsub("gene", "", rna),
           gene = gsub("gene-|name", "", gene),
           name = gsub("seq_id", "", name)) %>%
    dplyr::select(-target_id) %>%
    dplyr::select(seq_id, everything()) %>%
    mutate(Source = sample)
  
  # Make the chromosome column numeric
  parse.table$seq_id <- as.numeric(gsub("NC_0499.*?([0-9]+).*", "\\1", parse.table$seq_id))
  
  # Add groups for later coloring
  final.table <- parse.table %>%
    mutate(Dup_Group = case_when(
      seq_id == 4 ~ "chr_4",
      seq_id == 9 ~ "chr_9",
      TRUE ~ "Other")) %>%
    mutate(Hom_Group = case_when(
      seq_id == 11 ~ "chr_11",
      seq_id == 12 ~ "chr_12",
      seq_id == 16 ~ "chr_16",
      TRUE ~ "Other"))
  return(final.table)
}

plot_exp_branch <- function(i, R){
  
  all_exp_data_tmp<-all_exp_data[mintpm!=0]
  
  node1 <- tree_exp$data$branch[i, 1]; node2 <- tree_exp$data$branch[i, 2]
  
  # Get the tips (samples) downstream of node2
  descendants <- Descendants(tree_exp, node2, type = "tips")[[1]]
  tips_in_branch <- tree_exp$tip.label[descendants]
  
  all_exp_data_tmp[, tip := Source %in% tips_in_branch]  # Add the tip column once
  all_exp_data_RNA <- all_exp_data_tmp[RNA == R]
  
  # Perform t-test based on random or not
  ttest <- if (random) {
    t.test((all_exp_data_RNA$tpm) ~ sample(all_exp_data_RNA$tip))
  } else {
    t.test(all_exp_data_RNA$tpm ~ all_exp_data_RNA$tip)
  }
  
  wilcoxtest <- if (random) {
    wilcox.test((all_exp_data_RNA$tpm) ~ sample(all_exp_data_RNA$tip))
  } else {
    wilcox.test(all_exp_data_RNA$tpm ~ all_exp_data_RNA$tip)
  }
  
  plot<-ggplot(all_exp_data_RNA, aes(x=tip, y=tpm))+
    geom_boxplot()
  
  tests<-data.table(
    RNA = R,
    Gene = unique(all_exp_data_RNA$Gene),
    p = ttest$p.value,
    stat = ttest$statistic,
    wilcoxp=wilcoxtest$p.value,
    w=wilcoxtest$statistic,
    tips = paste(tips_in_branch, collapse = " "),
    branch = i
  )
  return(list(plot, tests))
}

exp_tree_tests <- function(tree_exp, all_exp_data, Gene_sub = NULL, random = FALSE) {
  
  # Create a copy and subset once
  all_exp_data_tmp <- if (!is.null(Gene_sub)) {
    all_exp_data[Gene %in% Gene_sub]
  } else {
    all_exp_data
  }
  all_exp_data_tmp<-all_exp_data[mintpm!=0]
  unique_RNA <- unique(all_exp_data_tmp$RNA)  # Extract unique RNA once
  
  exp_tests <- rbindlist(lapply(seq_len(nrow(tree_exp$edge)), function(i) {
    node1 <- tree_exp$edge[i, 1]; node2 <- tree_exp$edge[i, 2]
    
    # Get the tips (samples) downstream of node2
    descendants <- Descendants(tree_exp, node2, type = "tips")[[1]]
    tips_in_branch <- tree_exp$tip.label[descendants]
    
    if (length(tips_in_branch) > 1) {
      all_exp_data_tmp[, tip := Source %in% tips_in_branch]  # Add the tip column once
      
      # Use pblapply to iterate over unique RNA values
      ttests <- rbindlist(pblapply(unique_RNA, function(R) {
        all_exp_data_RNA <- all_exp_data_tmp[RNA == R]
        
        # Perform t-test based on random or not
        ttest <- if (random) {
          t.test((all_exp_data_RNA$tpm) ~ sample(all_exp_data_RNA$tip))
        } else {
          t.test(all_exp_data_RNA$tpm ~ all_exp_data_RNA$tip)
        }
        
        wilcoxtest <- if (random) {
          wilcox.test((all_exp_data_RNA$tpm) ~ sample(all_exp_data_RNA$tip))
        } else {
          wilcox.test(all_exp_data_RNA$tpm ~ all_exp_data_RNA$tip)
        }
        
        # Return a data.table for each RNA test
        data.table(
          RNA = R,
          Gene = unique(all_exp_data_RNA$Gene),
          p = ttest$p.value,
          stat = ttest$statistic,
          wilcoxp=wilcoxtest$p.value,
          w=wilcoxtest$statistic,
          tips = paste(tips_in_branch, collapse = " "),
          branch = i
        )
      }))
      
      return(ttests)
    } else {
      return(NULL)
    }
  }))
  
  return(exp_tests)
}


make_exp_windows<-function(test, windows=50){
  
  expression_windows<-real_tests2[,.(stat=median(stat, na.rm=T),
                                     sig=median(-log10(p), na.rm=T)
  ), by=.(CHROM, POS=as.numeric(cut(START, breaks=windows)), branch)]
  
  ggplot(expression_windows, aes(x=POS, y=branch, fill=stat, alpha=sig))+
    geom_tile()+
    facet_grid(branch~CHROM, scales="free",space="free", switch = "y")+
    theme_chrom()+
    scale_fill_gradient2()+
    scale_y_discrete(expand=c(0,0))+
    scale_x_continuous(expand=c(0,0))
}
```

## Read in data
```{r}
# Tree object from mutation tree creation
tree_exp <- readRDS("./Data/Plot_Objects/snp_tree_rna_samples_plot.rds")
# RNA Seq data
cr11a <- fread("./Data/Kallisto/11A_abundance.tsv")
cr12a <- fread("./Data/Kallisto/12A_abundance.tsv")
cr13a <- fread("./Data/Kallisto/13A_abundance.tsv")
cr15a <- fread("./Data/Kallisto/15A_abundance.tsv")
cr15a1 <- fread("./Data/Kallisto/15A1_abundance.tsv")
cr16a <- fread("./Data/Kallisto/16A_abundance.tsv")
cr16a1 <- fread("./Data/Kallisto/16A1_abundance.tsv")
cr16a2 <- fread("./Data/Kallisto/16A2_abundance.tsv")
cr17a1 <- fread("./Data/Kallisto/17A1_abundance.tsv")
cr17a2 <- fread("./Data/Kallisto/17A2_abundance.tsv")
cr18a <- fread("./Data/Kallisto/18A_abundance.tsv")
cr18a1 <- fread("./Data/Kallisto/18A1_abundance.tsv")
# GFF genes
genes_good <- fread("./Data/GFFs/Parsed/genes_good.tsv")
```

## Parse column information and create large data frame
```{r}
cr11a <- parse_columns(cr11a, "E 14")
cr12a <- parse_columns(cr12a, "E 12")
cr13a <- parse_columns(cr13a, "E 13")
cr15a <- parse_columns(cr15a, "E 8")
cr15a1 <- parse_columns(cr15a1, "E 7")
cr16a <- parse_columns(cr16a, "E 6")
cr16a1 <- parse_columns(cr16a1, "E 2")
cr16a2 <- parse_columns(cr16a2, "E 1")
cr17a1 <- parse_columns(cr17a1, "E 4")
cr17a2 <- parse_columns(cr17a2, "E 3")
cr18a <- parse_columns(cr18a, "E 9")
cr18a1 <- parse_columns(cr18a1, "E 5")

all_counts <- rbind(cr11a, cr12a, cr13a, cr15a, cr15a1, cr16a, cr16a1, cr16a2, cr17a1, cr17a2, cr18a, cr18a1)
```

# Prepare data
## Renam some columns
```{r}
names(all_counts)[names(all_counts) == "gene"] <- "Gene"
names(all_counts)[names(all_counts) == "rna"] <- "RNA"
names(all_counts)[names(all_counts) == "seq_id"] <- "CHROM"
```

## Calculate min and mean tpm for transcripts
```{r}
all_counts[,meantpm:=mean(tpm), by=RNA]
all_counts[,mintpm:=min(tpm), by=RNA]
```


## Convert to phangorn
```{r}
# all_exp_data_tmp<-all_counts[mintpm!=0]
# unique_RNA <- unique(all_exp_data_tmp$RNA)  # Extract unique RNA once


tree <- as.phylo(tree_exp)
```


## Trying function
```{r}
## takes a few minutes:
real_tests <- exp_tree_tests(tree, all_counts, Gene_sub=NULL, random=F)
real_tests <- merge(real_tests, genes_good, by.x = "Gene", by.y = "GENE")
mintpm <- all_counts[,.(RNA, mintpm)]
real_tests$mintpm <- mintpm$mintpm[match(real_tests$RNA, mintpm$RNA)]
real_tests <- real_tests[mintpm>0]
real_tests[,padjust:=p.adjust(p, method="fdr"), by=branch]
real_tests$sig<- -log10(real_tests$p)
real_tests$sig<-ifelse(real_tests$stat<0, real_tests$sig, -real_tests$sig)
```
## Set levels for plotting
```{r}

```

## Plot tree to figure out individuals in node
```{r}
library(ggtree)
ggtree(tree) +
  geom_tiplab() +
  geom_text2(aes(label = node), hjust = -0.3, size = 3)
```


## Plotting all
```{r}
ggplot(real_tests[padjust<0.05], aes(x = START, col = stat > 0, y = sig)) +
  geom_point(size = 0.1) +
  geom_segment(aes(x = START, xend = START, y = sig, yend = 0), size = 0.1) +
  facet_grid(branch~CHROM, space = "free_x", scales = "free_x", switch = "y") +
  scale_color_manual(values = c("orangered3", "royalblue1")) +
  scale_y_continuous(position = "right") +
  scale_x_continuous(labels = label_number(scale = 1e-7, suffix = "")) +
  labs(x = "Position", y = "-log10(p-value)") +
  theme_linedraw(base_size = 6) +
  theme(strip.background = element_blank(),
        strip.text.x = element_text(colour = "black", margin = ggplot2::margin(t = 0, b = 0)),
        panel.spacing.x = unit(0.01,"cm"),
        panel.spacing.y = unit(0.01,"cm"),
        strip.text.y = element_text(color = "black"),
        legend.position = "none",
        axis.text.x = element_blank(),
        axis.ticks.x = element_blank(),
        plot.margin = ggplot2::margin(t = 0.0,r = 0.0,b = 0.0,l = 0.05, unit = "cm")
  )
ggsave("./Figures/RNA_Sequencing/clade_gene_diff_expression.pdf", height = 2, width = 4.5)
ggsave("./Figures/RNA_Sequencing/clade_gene_diff_expression.png", height = 2, width = 4.5)
```

## Plotting only chromosome 12 deletions
```{r}
ggplot(real_tests[padjust<0.05 & CHROM %in% 12 & branch %in% 22], aes(x = START, col = stat > 0, y = sig)) +
  geom_point(size = 0.5) +
  geom_hline(yintercept = 0.5, size = 0.25, color = "grey50", linetype = 2) +
  #geom_segment(aes(x = START, xend = START, y = sig, yend = 0), size = 0.2) +
  #facet_grid(branch~CHROM, space = "free_x", scales = "free_x", switch = "y") +
  scale_color_manual(values = c("orangered3", "royalblue1")) +
  scale_y_continuous(limits = c(-8.5,8.5), position = "right") +
  scale_x_continuous(labels = label_number(scale = 1e-7, suffix = "")) +
  labs(title = "Chromosome 12", x = "Chromosome 12", y = "-log10(p-value)") +
  theme_classic(base_size = 6) +
  theme(strip.background = element_blank(),
        strip.text.x = element_blank(),
        panel.spacing.x = unit(0.01,"cm"),
        panel.spacing.y = unit(0.01,"cm"),
        strip.text.y = element_text(color = "black"),
        legend.position = "none",
        axis.text.x = element_blank(),
        axis.ticks.x = element_blank(),
        #axis.title.x = element_blank(),
        plot.title = element_blank(),
        #panel.border = element_rect(color = "black", fill = NA, linewidth = 0.5), 
        plot.margin = ggplot2::margin(t = 0.015,r = 0.05,b = 0.0,l = 0.05, unit = "cm")
  )
ggsave("./Figures/RNA_Sequencing/clade_gene_diff_expression_chr12_del.pdf", height = 1, width = 1.6)
ggsave("./Figures/RNA_Sequencing/clade_gene_diff_expression_chr12_del.png", height = 1, width = 1.6)
```

## Plotting only chromosome 4 deletions
```{r}
ggplot(real_tests[padjust<0.05 & CHROM %in% 4 & branch %in% 18], aes(x = START, col = stat > 0, y = sig)) +
  geom_point(size = 0.5) +
  geom_hline(yintercept = 0.5, size = 0.25, color = "grey50", linetype = 2) +
  #geom_segment(aes(x = START, xend = START, y = sig, yend = 0), size = 0.2) +
  #facet_grid(branch~CHROM, space = "free_x", scales = "free_x", switch = "y") +
  scale_color_manual(values = c("orangered3", "royalblue1")) +
  scale_y_continuous(limits = c(-8.5,8.5), position = "right") +
  scale_x_continuous(labels = label_number(scale = 1e-7, suffix = "")) +
  labs(title = "Chromosome 4", x = "Chromosome 4", y = "-log10(p-value)") +
  theme_classic(base_size = 6) +
  theme(strip.background = element_blank(),
        strip.text.x = element_blank(),
        panel.spacing.x = unit(0.01,"cm"),
        panel.spacing.y = unit(0.01,"cm"),
        strip.text.y = element_text(color = "black"),
        legend.position = "none",
        axis.text.x = element_blank(),
        axis.ticks.x = element_blank(),
        #axis.title.x = element_blank(),
        axis.title.y = element_text(color = "white"),
        axis.ticks.y = element_line(color = "white"),
        axis.text.y = element_text(color = "white"),
        axis.line.y = element_line(color = "white"),
        plot.title = element_blank(),
        #panel.border = element_rect(color = "black", fill = NA, linewidth = 0.5),
        plot.margin = ggplot2::margin(t = 0.015,r = 0.05,b = 0.0,l = 0.05, unit = "cm")
  )
ggsave("./Figures/RNA_Sequencing/clade_gene_diff_expression_chr4_del.pdf", height = 1, width = 1.6)
ggsave("./Figures/RNA_Sequencing/clade_gene_diff_expression_chr4_del.png", height = 1, width = 1.6)
```