---
title: "hemizygosity_embryos"
author: "Matthew Davis"
date: "2024-07-19"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

# Setup
## Libraries
```{r, message=FALSE}
library(tidyverse)
library(data.table)
library(gridExtra)
```

## Functions
```{r}
quick_qual_dist <- function(vcf){
  
  all <- vcf %>%
  ggplot(aes(x = QUAL)) +
  geom_histogram() +
  geom_vline(aes(xintercept = mean(QUAL))) + 
  labs(title = "All QUAL") +
  theme_classic()

qual_5 <- vcf %>%
  filter(QUAL >= 5) %>%
  ggplot(aes(x = QUAL)) +
  geom_histogram() +
  geom_vline(aes(xintercept = mean(QUAL))) +
  labs(title = "QUAL >= 5") +
  theme_classic()

grid.arrange(all, qual_5, ncol = 2)

}

VAF_point_plot <- function(vcf, quality = 0, size  = 1, alpha = 1, 
                           het_color = "deepskyblue4", hom_color = "darksalmon",
                           col_num = 1, samples = c("ref_chandler", "tree2", "cr2_hifi", "cr85", "cr10",
                                                    "cr13", "cr22", "cr21_1", "cr21_2", "cr2_11A","cr2_12A",
                                                    "cr2_13A", "cr2_15A", "cr2_15A1", "cr2_16A", "cr2_16A1",
                                                    "cr2_16A2", "cr2_17A1", "cr2_17A2", "cr2_18A", "cr2_18A1")) {
  
  # Select which sources to plot, heterozygous and homozygous sites, quality threshold
  vcf_filt <- vcf %>%
    filter(Source %in% samples) %>%
    filter(str_length(REF) == 1) %>%
    filter(str_length(ALT) == 1) %>%
    filter(GT %in% c("0/1", "1/1")) %>%
    filter(QUAL >= quality)
  
  # Get log depth by Source
  vcf_tran <- vcf_filt %>%
    group_by(Source) %>%
    mutate(DP_log = log(DP)) %>%
    mutate(DP_log_mean = mean(DP_log)) %>%
    ungroup()

  #Plot the data
  plot <- vcf_tran %>%
    ggplot(aes(x = POS/1000000, y = VAF, col = GT)) +
    geom_point(size = size, alpha = alpha) +
    facet_wrap(~ Source, ncol = col_num, scales = "free") +
    scale_color_manual(values=c("0/1" = het_color, "1/1" = hom_color)) +
    labs(x = "Position (Mb)",
         y = "VAF",
         color = "Genotype") +
    theme_classic(base_size = 6) +
    theme(legend.position = "bottom",
          axis.title.x = element_text(face = "bold"),
          axis.title.y = element_text(face = "bold"))

  return(plot)
}

window_depth_plot <- function(vcf, window_size = 1e6, quality = 0, col_num = 1, 
                              samples = c("ref_chandler", "tree2", "cr2_hifi", "cr85", "cr10",
                                          "cr13", "cr22", "cr21_1", "cr21_2", "cr2_11A", "cr2_12A",
                                          "cr2_13A", "cr2_15A", "cr2_15A1", "cr2_16A", "cr2_16A1",
                                          "cr2_16A2", "cr2_17A1", "cr2_17A2", "cr2_18A", "cr2_18A1")){
  
  # Define the window size
window_size <- window_size

# Create a new column to define windows based on the POS column
vcf_windows <- vcf[, window := floor((POS - 1) / window_size) + 1, by = .(CHROM, Source)]

# Select which sources to plot, heterozygous and homozygous sites, quality threshold
  vcf_filt <- vcf_windows %>%
    filter(Source %in% samples) %>%
    filter(str_length(REF) == 1) %>%
    filter(str_length(ALT) == 1) %>%
    filter(GT %in% c("0/1", "1/1")) %>%
    filter(QUAL >= quality)
  
  # Get log depth by Source
  vcf_tran <- vcf_filt %>%
    group_by(window) %>%
    mutate(mean_DP = mean(DP)) %>%
    mutate(DP_log = log(DP)) %>%
    mutate(DP_log_mean = mean(DP_log)) %>%
    ungroup()
  
  # Plot
  plot <- vcf_tran %>%
    ggplot(aes(x = window, y = mean_DP)) +
    geom_line() +
    facet_wrap(~ Source, ncol = col_num, scales = "free") +
    labs(x = "Position (Windows)",
         y = "Mean Depth in Windows") +
    theme_classic(base_size = 6) +
    theme(legend.position = "bottom",
          axis.title.x = element_text(face = "bold"),
          axis.title.y = element_text(face = "bold"))
  
  return(plot)
}

point_plot <- function(vcf, window_size = 1e6, het_color = "deepskyblue4", hom_color = "darksalmon",
                       quality = 0, col_num = 1, size = 1, alpha = 1, samples = c("ref_chandler", "tree2", "cr2_hifi", "cr85", "cr10",
                                                              "cr13", "cr22", "cr21_1", "cr21_2", "cr2_11A", "cr2_12A",
                                                              "cr2_13A", "cr2_15A", "cr2_15A1", "cr2_16A", "cr2_16A1",
                                                              "cr2_16A2", "cr2_17A1", "cr2_17A2", "cr2_18A", "cr2_18A1")) {
  
   # Define the window size
  window_size <- window_size

  # Create a new column to define windows based on the POS column
  vcf_windows <- vcf[, window := floor((POS - 1) / window_size) + 1, by = .(CHROM, Source)]

# Select which sources to plot, heterozygous and homozygous sites, quality threshold
  vcf_filt <- vcf_windows %>%
    filter(Source %in% samples) %>%
    filter(str_length(REF) == 1) %>%
    filter(str_length(ALT) == 1) %>%
    filter(GT %in% c("0/1", "1/1")) %>%
    filter(QUAL >= quality)
  
  # # Select which sources to plot, heterozygous and homozygous sites, quality threshold
  # vcf_filt <- vcf %>%
  #   filter(Source %in% samples) %>%
  #   filter(str_length(REF) == 1) %>%
  #   filter(str_length(ALT) == 1) %>%
  #   filter(GT %in% c("0/1", "1/1")) %>%
  #   filter(QUAL >= quality)
  
  # Get log depth by Source
  vcf_tran <- vcf_filt %>%
    group_by(window) %>%
    mutate(DP_log = log(DP)) %>%
    mutate(DP_log_mean = mean(DP_log)) %>%
    ungroup()

  #Plot the data
  plot <- vcf_tran %>%
    ggplot(aes(x = POS/1000000, y = DP_log, col = GT)) +
    geom_point(size = size, alpha = alpha) +
    geom_line(aes(y = DP_log_mean, x = window)) +
    scale_color_manual(values=c("0/1" = het_color, "1/1" = hom_color)) +
    facet_wrap(~ Source, ncol = col_num) +
    labs(x = "Position (Mb)",
         y = "Depth (natural log)",
         color = "Genotype") +
    theme_classic(base_size = 6) +
    theme(legend.position = "bottom",
          axis.title.x = element_text(face = "bold"),
          axis.title.y = element_text(face = "bold"))

  return(plot)
}

fraction_hom_plot <- function(vcf, window_size = 1e4, quality = 0, col_num = 1, low_color = "grey90",
                              high_color = "red3", samples = c("ref_chandler", "tree2", "cr2_hifi", 
                                                               "cr85", "cr10", "cr13", "cr22", "cr21_1",
                                                               "cr21_2", "cr2_11A", "cr2_12A", "cr2_13A", 
                                                               "cr2_15A", "cr2_15A1", "cr2_16A", "cr2_16A1",
                                                               "cr2_16A2", "cr2_17A1", "cr2_17A2", "cr2_18A",
                                                               "cr2_18A1")){
  
  # Define the window size
  window_size <- window_size

  # Create a new column to define windows based on the POS column
  vcf_windows <- vcf[, window := floor((POS - 1) / window_size) + 1, by = .(CHROM, Source)]

# Select which sources to plot, heterozygous and homozygous sites, quality threshold
  vcf_filt <- vcf_windows %>%
    filter(Source %in% samples) %>%
    filter(str_length(REF) == 1) %>%
    filter(str_length(ALT) == 1) %>%
    filter(GT %in% c("0/1", "1/1")) %>%
    filter(QUAL >= quality)
  
  # Get log depth by Source
  vcf_tran <- vcf_filt %>%
    group_by(window) %>%
    mutate(fraction_1_1 = sum(GT == "1/1") / n()) %>%
    ungroup()
  
  # Plot
  plot <- vcf_tran %>%
    ggplot(aes(x = window, y = 1, fill = fraction_1_1)) +
    geom_tile() + 
    facet_wrap(~ Source, ncol = col_num, scales = "free" ) +
    scale_fill_gradient(low = low_color, high = high_color) +
    labs(x = "Position (Window)", y = element_blank(), fill = "Fraction of\nHomozygous Regions") +
    theme_classic(base_size = 6) +
    theme(axis.text.y = element_blank(),
          axis.ticks.y = element_blank(),
          axis.line.y = element_blank())
  
  return(plot)
}
```

## Read in data
```{r}
chr11_an <- fread("./Data/VCFs/Filtered/all_an_snps_chr_11.vcf")
chr12_an <- fread("./Data/VCFs/Filtered/all_an_snps_chr_12.vcf")
chr16_an <- fread("./Data/VCFs/Filtered/all_an_snps_chr_16.vcf")
```

# Brief data check
## Checking distribution of quality scores
```{r}
quick_qual_dist(chr11_an)
quick_qual_dist(chr12_an)
quick_qual_dist(chr16_an)
```

# Look at VAF by position
## Plot SNPs by VAF and position
```{r}
VAF_point_plot(chr11_an, quality = 30, alpha = 0.3, col_num = 2, samples = c("cr2_hifi", "cr2_16A", "cr2_16A1", "cr2_16A2",
                                              "cr2_17A1", "cr2_17A2", "cr2_18A1"))

VAF_point_plot(chr12_an, quality = 30, alpha = 0.3, samples = c("cr2_15A", "cr2_15A1"))

VAF_point_plot(chr16_an, quality = 30, alpha = 0.3, col_num = 2, samples = c("cr2_hifi", "cr2_15A", "cr2_15A1", "cr2_16A", "cr2_16A1",
                                 "cr2_16A2", "cr2_17A1", "cr2_17A2", "cr2_18A", "cr2_18A1"))
```
# Look at average depth in windows
## Plot
```{r}
window_depth_plot(chr11_an, col_num = 2, samples = c("cr2_hifi", "cr2_16A", "cr2_16A1", "cr2_16A2",
                                                     "cr2_17A1", "cr2_17A2", "cr2_18A1"))

window_depth_plot(chr12_an, samples = c("cr2_15A", "cr2_15A1"))

window_depth_plot(chr16_an, col_num = 2, samples = c("cr2_hifi", "cr2_15A", "cr2_15A1", "cr2_16A", "cr2_16A1",
                                                     "cr2_16A2", "cr2_17A1", "cr2_17A2", "cr2_18A", "cr2_18A1"))
```

# Looking at position by depth by genotype
## Plot SNPs by position and depth, colored by genotype
```{r}
# Plot individual chromosomes and affected samples
point_plot(chr11_an, quality = 30, col_num = 2, alpha = 0.3, samples = c("cr2_hifi", "cr2_16A", "cr2_16A1", "cr2_16A2",
                                              "cr2_17A1", "cr2_17A2", "cr2_18A1"))
#ggsave("./Figures/Hemizygosity/chr11_position_depth_genotype.png", height = 5, width = 7)

point_plot(chr12_an, quality = 30, alpha = 0.3, samples = c("cr2_15A", "cr2_15A1"))
#ggsave("./Figures/Hemizygosity/chr12_position_depth_genotype.png", height = 3, width = 4)

point_plot(chr16_an,  quality = 30, col_num = 2,  alpha = 0.3, samples = c("cr2_hifi", "cr2_15A", "cr2_15A1", "cr2_16A", "cr2_16A1",
                                 "cr2_16A2", "cr2_17A1", "cr2_17A2", "cr2_18A", "cr2_18A1"))
#ggsave("./Figures/Hemizygosity/chr16_position_depth_genotype.png", height = 5, width = 7)


# Combine chromosome data
dup_chr_snps <- rbind(chr11_an, chr12_an, chr16_an)

# Plot all samples and chromsomes for large scale comparison
dup_chr_snps  %>%
    filter(GT %in% c("0/1", "1/1")) %>%
    filter(QUAL >= 0) %>%
    group_by(Source) %>%
    mutate(DP_log = log(DP)) %>%
    mutate(DP_log_mean = mean(DP_log)) %>%
  ggplot(aes(x = POS/1000000, y = DP_log, col = GT)) +
    geom_point() +
    geom_hline(aes(yintercept = DP_log_mean)) +
    scale_color_manual(values=c("0/1" = "deepskyblue4", "1/1" = "darksalmon")) +
    facet_grid(Source ~ CHROM, space = "free", scales = "free_x") +
    labs(x = "Position (Mb)",
         y = "Depth (natural log)",
         color = "Genotype") +
    theme_classic() +
    theme(legend.position = "bottom",
          axis.title.x = element_text(face = "bold"),
          axis.title.y = element_text(face = "bold"))
# ggsave("./Figures/Hemizygosity/all_source_chr11_12_16_position_depth_genotype.png", height = 15, width = 9)
```

# Find the fraction of homoozygous alternative calls by window
## Generate plots
```{r}
fraction_hom_plot(chr11_an, window_size = 1e6, quality = 0, col_num = 2, samples = c("cr2_hifi","cr2_16A","cr2_16A1","cr2_16A2",
                                                                   "cr2_17A1", "cr2_17A2","cr2_18A1"))

fraction_hom_plot(chr12_an, window_size = 1e6, quality = 0, samples = c("cr2_15A", "cr2_15A1"))

fraction_hom_plot(chr16_an,  window_size = 1e6, quality = 0, col_num = 2, samples = c("cr2_hifi","cr2_15A","cr2_15A1","cr2_16A",
                                                                    "cr2_16A1","cr2_16A2","cr2_17A1",
                                                                    "cr2_17A2","cr2_18A","cr2_18A1"))
```


## Generate the data
```{r}

quality = 0
samples = c("cr2_hifi", "cr2_16A", "cr2_16A1", "cr2_16A2",
                                              "cr2_17A1", "cr2_17A2", "cr2_18A1")
   # Define the window size
  window_size <- 1e4

  # Create a new column to define windows based on the POS column
  vcf_windows <- chr11_an[, window := floor((POS - 1) / window_size) + 1, by = .(CHROM, Source)]

# Select which sources to plot, heterozygous and homozygous sites, quality threshold
  vcf_filt <- vcf_windows %>%
    filter(Source %in% samples) %>%
    filter(str_length(REF) == 1) %>%
    filter(str_length(ALT) == 1) %>%
    filter(GT %in% c("0/1", "1/1")) %>%
    filter(QUAL >= quality)
  
  # Get log depth by Source
  vcf_tran <- vcf_filt %>%
    group_by(window) %>%
    mutate(fraction_1_1 = sum(GT == "1/1") / n()) %>%
    ungroup()
  
  unique(vcf_tran$fraction_1_1)
  
  ggplot(vcf_tran, aes(x = window, y = 1, fill = fraction_1_1)) +
    geom_tile() + 
    facet_wrap(~Source, scales = "free") +
    scale_fill_gradient(low = "grey90", high = "dodgerblue") +
    labs(x = "Position (Window)", y = element_blank(), fill = "Fraction of\nHomozygous Regions") +
    theme_classic(base_size = 6) +
    theme(axis.text.y = element_blank(),
          axis.ticks.y = element_blank(),
          axis.line.y = element_blank())
```

## Plot the calls
```{r}

```


# Work on creating a smooth line graph
## Create homozygous and heterozyogus files
```{r}
# chr16_hom <- chr16_filt %>%
#   filter(GT %in% "1/1")
```

## Take the average depth in the window of hets and homs
```{r}
# # Function to calculate average depth starting from position 1
# avg_depth_window <- function(data, window_size = 10000, genotype) {
#   # Filter data based on genotype
#   filtered_data <- data[GT == genotype]
#   
#   # Initialize an empty list to store results
#   result_list <- list()
#   
#   # Determine the maximum POS value
#   max_pos <- max(filtered_data$POS)
#   
#   # Loop through the POS values in steps of window_size, starting from 1
#   for (start in seq(1, max_pos, by = window_size)) {
#     end <- start + window_size - 1
#     
#     # Subset data for the current window
#     window_data <- filtered_data[POS >= start & POS <= end]
#     
#     # Calculate the average depth (DP)
#     if (nrow(window_data) > 0) {
#       avg_depth <- mean(window_data$DP, na.rm = TRUE)
#     } else {
#       avg_depth <- NA
#     }
#     
#     # Store the results
#     result_list[[length(result_list) + 1]] <- data.table(
#       start = start,
#       end = end,
#       avg_depth = avg_depth
#     )
#   }
#   
#   # Combine all results into a data table
#   result <- rbindlist(result_list)
#   return(result)
# }
# 
# chr16_avg_het <- avg_depth_window(data = chr16_an, genotype = "0/1")
# chr16_avg_hom <- avg_depth_window(data = chr16_an, genotype = "1/1")
```

## Try something new
```{r}
# Define window size
window_size <- 1000000

# Create a column for window number
chr16_an <- chr16_an %>%
  mutate(Window = floor((POS - 1) / window_size) + 1)

 het <- chr16_an %>%
   group_by(Source, Window) %>%
   filter(GT == "0/1") %>%
   summarize(het_avg_dp = mean(DP, na.rm = TRUE), .groups = 'drop')
 
 hom <- chr16_an %>%
   group_by(Source, Window) %>%
   filter(GT == "1/1") %>%
   summarize(hom_avg_dp = mean(DP, na.rm = TRUE), .groups = 'drop')
 
 # Combine both data frames into one
 combined_df <- left_join(het, hom, by = c("Source", "Window"))
 
 # combined_df <- combined_df %>%
 #  replace_na(list(het_avg_dp = 0, hom_avg_dp = 0))
```

## Plot the thing
```{r}
# Reshape data from wide to long format
long_df <- combined_df %>%
  pivot_longer(cols = c(het_avg_dp, hom_avg_dp), 
               names_to = "type", 
               values_to = "average_depth")

# Scale 
# Calculate the min and max of depth_count for each Source and type
min_max_counts <- long_df %>%
  group_by(Source, type) %>%
  summarise(
    min_depth = min(average_depth, na.rm = TRUE),
    max_depth = max(average_depth, na.rm = TRUE),
    .groups = 'drop'
  )

# Join the min_max_counts back to the original data frame and normalize the depth_count
long_df_scaled <- long_df %>%
  left_join(min_max_counts, by = c("Source", "type")) %>%
  mutate(
    scaled_depth_count = 2 * ((average_depth - min_depth) / (max_depth - min_depth)) - 1
  ) %>%
  select(-min_depth, -max_depth) %>%
  mutate(Position = ((Window*window_size) / 2) / 1e6)# Remove the min and max columns if not needed

# Plot the data
ggplot(long_df_scaled, aes(x = Position, y = scaled_depth_count, color = type)) +
  geom_line() +
  facet_wrap(~ Source) +
  labs(x = "Position (Mb)",
       y = "Average Depth",
       color = "Genotype") +
  theme_classic()
```

## Stuff
```{r}
# Define window size
window_size <- 1000000

 het2 <- chr16_an %>%
   group_by(Source, Window) %>%
   filter(GT == "0/1") %>%
   summarize(het_dp_count = sum(DP, na.rm = TRUE), .groups = 'drop')
 
 hom2 <- chr16_an %>%
   group_by(Source, Window) %>%
   filter(GT == "1/1") %>%
   summarize(hom_dp_count = sum(DP, na.rm = TRUE), .groups = 'drop')
 
 # Combine both data frames into one
 combined_df2 <- left_join(het2, hom2, by = c("Source", "Window"))
```

## Plot 2
```{r}
# Reshape data from wide to long format
long_df2 <- combined_df2 %>%
  pivot_longer(cols = c(het_dp_count, hom_dp_count), 
               names_to = "type", 
               values_to = "depth_count")

# Scale 
# Calculate the min and max of depth_count for each Source and type
min_max_counts <- long_df2 %>%
  group_by(Source, type) %>%
  summarise(
    min_depth = min(depth_count, na.rm = TRUE),
    max_depth = max(depth_count, na.rm = TRUE),
    .groups = 'drop'
  )

# Join the min_max_counts back to the original data frame and normalize the depth_count
long_df2_scaled <- long_df2 %>%
  left_join(min_max_counts, by = c("Source", "type")) %>%
  mutate(
    scaled_depth_count = 2 * ((depth_count - min_depth) / (max_depth - min_depth)) - 1
  ) %>%
  select(-min_depth, -max_depth) %>%
  mutate(Position = ((Window*window_size) / 2) / 1e6)# Remove the min and max columns if not needed

long_df2_scaled %>%
ggplot(aes(x = Position, y = scaled_depth_count, color = type)) +
  geom_line() +
  facet_wrap(~ Source) +
  labs(x = "Position (Mb)",
       y = "Scaled Depth Count",
       color = "Genotype") +
  theme_classic()
ggsave("./Figures/Hemizygosity/chr16_position_depth_genotype_1mb_window.pdf", height = 5, width = 7)
ggsave("./Figures/Hemizygosity/chr16_position_depth_genotype_1mb_window.png", height = 5, width = 7)
```

## Try the counts, but using log
```{r}
long_df2_log <- long_df2 %>%
  group_by(Source) %>%
  mutate(depth_count_log = log(depth_count)) %>%
  mutate(depth_count_mean = mean(depth_count_log)) %>%
  mutate(POS = ((Window*window_size)/2)/1e6) %>%
  ungroup()
    
long_df2_log %>%
ggplot(aes(x = POS, y = depth_count_log, color = type)) +
  geom_line() +
  geom_hline(aes(yintercept = depth_count_mean)) +
  facet_wrap(~ Source) +
  labs(x = "Position (Mb)",
       y = "Mean Depth per Window (natural log)",
       color = "Genotype") +
  theme_classic()
```

