---
title: "parsing_data"
author: "Matthew Davis"
date: "2024-10-31"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

# Setup
## Libraries
```{r, message=FALSE}
library(polymorphology2)
library(tidyverse)
library(pbapply)
```

## Functions
```{r}
mutations_embryo_tree<-function(mutations_sub){
  mutations_sub<-mutations_sub[Type=="Embryo"]
  mutations_matrix<-data.table(table(UNIQUE=mutations_sub$UNIQUE,
                                     SOURCE=mutations_sub$Source))
  mutations_matrix<-dcast(mutations_matrix, SOURCE~UNIQUE, value.var = "N")
  distance_mat<-dist(mutations_matrix[,-1], method = "binary")
  clust<-hclust(distance_mat, method = "average")
  clust$labels<-mutations_matrix$SOURCE
  clustplot<-plot(clust)
  return(list(dist=distance_mat, plot=clustplot))
}


annot_mut_effect<-function(i){
  genic <- mutations_means$CDS[i]
  if(!genic) return("nonCDS")
  row <- mutations_means[i]
  POS <- row$POS
  REF <- row$REF
  ALT <- row$ALT
  CHROM = row$CHROM
  overlap <- sites_in_features(cds_good, row, mode = "any")[any==T]
  mod <- unique(cds_good[ID%in%overlap$ID]$Parent)[1]
  CDS_model <- cds_good[Parent%in%mod]
  DIR <- CDS_model$DIRECTION[1]
  CDS_POS <- unlist(lapply(1:nrow(CDS_model), function(j){
    CDS_model$START[j]:CDS_model$STOP[j]
  }))
  BPs <- CDS[[mod]]
  AA <- seqinr::translate(toupper(CDS[[mod]]))
  genome_BPs<-fasta[[CHROM]][CDS_POS]
  genome_AA<-seqinr::translate(toupper(genome_BPs))
  
  REF_genome<-toupper(fasta[[CHROM]][POS])
  if(REF_genome!=REF) return("REF_genome_mismatch")
  
  CDS_dt<-data.table(cPOS=CDS_POS, BP=toupper(cds[[mod]]))
  if(DIR=="-") {
    CDS_dt$cPOS<-rev(CDS_dt$cPOS)
    REF<-revcomp(REF)
    ALT<-revcomp(ALT)
    genome_AA<-seqinr::translate(unlist(strsplit(revcomp(toupper(fasta[[CHROM]][CDS_POS])), split="")))
  }
  if(!all(genome_AA==AA)) return("REF_genome_mismatch_AA")
  if(CDS_dt[cPOS==POS]$BP!=REF) return(CDS_dt[cPOS==POS]$BP)
  CDS_dt$AA<-rep(seqinr::translate(CDS_dt$BP), each=3)
  CDS_dt$BP2<-CDS_dt$BP
  CDS_dt$BP2[which(CDS_dt$cPOS==POS)]<-ALT
  CDS_dt[BP2!=BP]
  CDS_dt$newAA<-rep(seqinr::translate(CDS_dt$BP2), each=3)
  CDS_dt$AAPOS<-rep(1:length(seqinr::translate(CDS_dt$BP)), each=3)
  SYN=all(CDS_dt$AA==CDS_dt$newAA)
  message(i, mod, CDS_dt[which(CDS_dt$cPOS==POS)]$AA," to ", CDS_dt[which(CDS_dt$cPOS==POS)]$newAA)
  message(ifelse( SYN, "\tSyn","\tNon-Syn"))
  
  effect<-CDS_dt[which(CDS_dt$cPOS==POS)]
  effect$gene_model<-mod
  effect$CHROM<-CHROM
  ifelse(SYN, "Syn","Non-Syn")
}

vstrsplit<-function(x,...){
  unlist(strsplit(x,...))
}



CDS_codon_mutation_table <- function(CDS) {
  # Precompute mutations table
  muts <- data.table(table(REF = c("A", "T", "C", "G"), ALT = c("A", "T", "C", "G")))[REF != ALT]
  
  # Create codon table with all possible mutations
  codons <- data.table(SEQINR.UTIL$CODON.AA)
  codons[, CODON := toupper(CODON)]
  
  # Precompute the CODON_mutations_table
  CODON_mutations_table <- rbindlist(lapply(1:nrow(codons), function(i) {
    CODON <- codons$CODON[i]
    CODON_dt <- data.table(REF = strsplit(CODON, split = "")[[1]], POS = 1:3, REFCODON = CODON)
    CODON_dt <- unique(merge(CODON_dt, muts, by = "REF", allow.cartesian = TRUE))
    CODON_dt[, ALTCODON := sapply(1:.N, function(j) {
      REFCODON <- strsplit(CODON_dt$REFCODON[j], split = "")[[1]]
      REFCODON[CODON_dt$POS[j]] <- CODON_dt$ALT[j]
      paste0(REFCODON, collapse = "")
    })]
    CODON_dt[, `:=`(REFAA = codons$AA[match(REFCODON, codons$CODON)],
                    ALTAA = codons$AA[match(ALTCODON, codons$CODON)])]
    CODON_dt[, effect := ifelse(REFAA == ALTAA, "S", "N")]
    return(CODON_dt)
  }))
  
  # Convert each CDS into a mutation table
  pb <- txtProgressBar(min = 0, max = length(CDS), style = 3)
  CDS_dt_all <- rbindlist(lapply(1:length(CDS), function(i) {
    random_seq <- toupper(CDS[[i]])
    remainder <- length(random_seq) %% 3
    if (remainder != 0) {
      message("Gene not multiple of 3")
      return(NULL)
    }
    
    original_protein <- seqinr::translate(random_seq)
    CDS_dt <- data.table(POS = rep(1:3, times = length(random_seq) / 3),
                         DNAPOS = 1:length(random_seq),
                         REF = random_seq,
                         AA = rep(original_protein, each = 3))
    
    # Extract codon for each amino acid
    codons_list <- sapply(seq(1, length(random_seq), by = 3), function(start) {
      paste(random_seq[start:(start + 2)], collapse = "")
    })
    CDS_dt[, REFCODON := rep(codons_list, each = 3)]
    
    # Merge with mutations to add ALT column of all possible mutations at each codon
    CDS_dt_mutations_table <- unique(merge(CDS_dt, muts, by = "REF", allow.cartesian = TRUE))[order(DNAPOS)]
    
    # Merge with CODON_mutations_table to annotate effect
    CDS_dt_mutations_table <- merge(CDS_dt_mutations_table, CODON_mutations_table, by = c("REFCODON", "REF", "POS", "ALT"))
    setTxtProgressBar(pb, i)
    CDS_dt_mutations_table[, GENE := names(CDS)[i]]
    return(CDS_dt_mutations_table[, .(REF, ALT, effect, GENE)])
  }))
  close(pb)
  
  return(CDS_dt_all)
}


mutation_probs<-function(muts, genome, cds=NULL){
  message("Calculating base pair frequencies from genome...")
  bp_freq<-rbindlist(lapply(1:length(genome), function(i){
    message(paste("\tChr",i))
    bp_freq<-data.table(table(REF=genome[[i]]))[REF!="n"]
    bp_freq$REF<-toupper(bp_freq$REF)
    return(bp_freq)
  }))
  bp_freq<-bp_freq[REF %in% c("A","T","C","G"),.(N=sum(N)), by="REF"]
  
  mutations<-data.table((table(REF=muts$REF, ALT=muts$ALT)))[REF!=ALT]
  mutations<-merge(mutations, bp_freq, by="REF")
  mutations$rawPROB=(mutations$N.x/(mutations$N.y))
  
  if(!is.null(cds)){
    message("Calculating base pair frequencies from CDS...")
    bp_freqCDS<-rbindlist(lapply(1:length(cds), function(i){
      
      bp_freq<-data.table(table(REF=cds[[i]]))[REF!="n"]
      bp_freq$REF<-toupper(bp_freq$REF)
      return(bp_freq)
    }))
    bp_freqCDS<-bp_freqCDS[REF %in% c("A","T","C","G"),.(CDSN=sum(N)), by="REF"]
    
    mutations<-merge(mutations, bp_freqCDS, by="REF")
    mutations$rawPROB=(mutations$N.x/(mutations$N.y-mutations$CDSN))
  }
  return(mutations)
}


sample_neutrals<-function(neutrals, N, i){
  samples<-sapply(1:i, function(i){
    samp<-neutrals[sample(1:length(neutrals), N)]
    sum(samp=="N")/sum(samp=="S")
  })
}

GvsIG<-function(mutations_data, REGIONS){
  REGIONS$ID<-1:nrow(REGIONS)
  REGIONS_mut<-sites_in_features(REGIONS,mutations_data, mode = "counts")
  REGIONS$mut<-REGIONS_mut$counts[match(REGIONS$ID, REGIONS_mut$ID)]
  
  REGIONS_means<-REGIONS[,.(mut=sum(mut), LENGTH=sum(MAPP_LENGTH2), mutrate=sum(mut)/sum(MAPP_LENGTH2)), by=REGION]
  chitest<-chisq.test(REGIONS_means[,2:3])
  ratio<-REGIONS_means[REGION=="GENE"]$mutrate/REGIONS_means[REGION=="IG"]$mutrate
  return(data.table(p=chitest$p.value, ratio=ratio))
}

my_minimal<-function(){
  theme_minimal(base_size = 6)+
    theme(panel.background = element_blank(),
          plot.background = element_blank(),
          panel.grid = element_blank(),
          legend.key.size = unit(0.25, "cm"),
          legend.background = element_blank(),
          plot.title = element_text(hjust=0.5, size=6))
}
```

# Genome and annotations
## Preparing data
### Read in data
```{r}
# Mappability
mapp <- read.bedGraph("./Data/Genmap/K150E0/ref_chandler_primary_default_scaffold.genmap.bedgraph")

# Genome
fasta <- read.fasta("./Data/Fastas/ref_chandler_primary_default_scaffold.fasta")
fastah1 <- read.fasta("./Data/Fastas/ref_chandler_hap1_default_scaffold.fasta")
fastah2 <- read.fasta("./Data/Fastas/ref_chandler_hap2_default_scaffold.fasta")

# Annotations
gff <- read.GFF("./Data/GFFs/Chromosome_Only_Liftoff/ref_chandler_primary_default_scaffold_chr_only_liftoff_a99s99.gff")
h1gff <- read.GFF("./Data/GFFs/Chromosome_Only_Liftoff/ref_chandler_hap1_default_scaffold_chr_only_liftoff_a99s99.gff")
h2gff <- read.GFF("./Data/GFFs/Chromosome_Only_Liftoff/ref_chandler_hap2_default_scaffold_chr_only_liftoff_a99s99.gff")
```

### Rename chromosomes
```{r}
gff$CHROM<-as.numeric(gsub("NC_0499(..).1_RagTag", "\\1", gff$CHROM))
unique(gff$CHROM)
h1gff$CHROM<-as.numeric(gsub("NC_0499(..).1_RagTag", "\\1", h1gff$CHROM))
unique(gff$CHROM)
h2gff$CHROM<-as.numeric(gsub("NC_0499(..).1_RagTag", "\\1", h2gff$CHROM))
unique(gff$CHROM)

mapp$CHROM<-as.numeric(gsub("NC_0499(..).1_RagTag", "\\1", mapp$CHROM))
mapp <- mapp[!is.na(CHROM)]
unique(mapp$CHROM)

names(fasta)<-as.numeric(gsub("NC_0499(..).1_RagTag", "\\1", names(fasta)))
fasta <- fasta[!is.na(names(fasta))]
unique(names(fasta))

names(fastah1)<-as.numeric(gsub("NC_0499(..).1_RagTag", "\\1", names(fastah1)))
fastah1 <- fastah1[!is.na(names(fastah1))]
unique(names(fastah1))

names(fastah2)<-as.numeric(gsub("NC_0499(..).1_RagTag", "\\1", names(fastah2)))
fastah2 <- fastah2[!is.na(names(fastah2))]
unique(names(fastah2))
```


## Extract data
### Extract genes and cds sequences 
```{r}
# Primary
genes <- gff %>%
  filter(TYPE %in% "gene")

genes$GENE <- gsub("ID=gene-(.+);Dbxref.+", "\\1", genes$INFO)

sum(duplicated(genes$GENE))

cds <- gff %>%
  filter(TYPE %in% "CDS")

cds$Parent <- gsub(".+Parent=(.+);Dbxref.+", "\\1", cds$INFO)
sum(duplicated(genes$Parent))

cds$Product <- gsub(".+product=(.+);protein_id.+", "\\1", cds$INFO)
sum(duplicated(genes$Product))

cds$GENE <- gsub(".+CDS;gene=(.+);product.+", "\\1", cds$INFO)
sum(duplicated(cds$GENE))


cds$ID <- 1:nrow(cds)

# Hap 1

h1genes <- h1gff %>%
  filter(TYPE %in% "gene")

h1genes$GENE <- gsub("ID=gene-(.+);Dbxref.+", "\\1", h1genes$INFO)

sum(duplicated(h1genes$GENE))

h1cds <- h1gff %>%
  filter(TYPE %in% "CDS")

h1cds$Parent <- gsub(".+Parent=(.+);Dbxref.+", "\\1", h1cds$INFO)
sum(duplicated(h1cds$Parent))

h1cds$Product <- gsub(".+product=(.+);protein_id.+", "\\1", h1cds$INFO)
sum(duplicated(h1cds$Product))

h1cds$GENE <- gsub(".+CDS;gene=(.+);product.+", "\\1", h1cds$INFO)
sum(duplicated(h1cds$GENE))


h1cds$ID <- 1:nrow(h1cds)

# Hap 2

h2genes <- h2gff %>%
  filter(TYPE %in% "gene")

h2genes$GENE <- gsub("ID=gene-(.+);Dbxref.+", "\\1", h2genes$INFO)

sum(duplicated(h2genes$GENE))

h2cds <- h2gff %>%
  filter(TYPE %in% "CDS")

h2cds$Parent <- gsub(".+Parent=(.+);Dbxref.+", "\\1", h2cds$INFO)
sum(duplicated(h2cds$Parent))

h2cds$Product <- gsub(".+product=(.+);protein_id.+", "\\1", h2cds$INFO)
sum(duplicated(h2cds$Product))

h2cds$GENE <- gsub(".+CDS;gene=(.+);product.+", "\\1", h2cds$INFO)
sum(duplicated(h2cds$GENE))


h2cds$ID <- 1:nrow(h2cds)

```

### Protein amino acids
```{r}
# Primary
proteins <- pblapply(unique(cds$Parent), function(p){
  cds_p <- cds[Parent == p]
  dir <- unique(cds_p$DIRECTION)
  CHROM = cds_p$CHROM[1]
  genome_seq <- paste(sapply(1:nrow(cds_p), function(r){
    START = cds_p$START[r]
    STOP = cds_p$STOP[r]
    paste(fasta[[CHROM]][START:STOP], collapse="")
  }), collapse = "")
  if(dir == "-"){
    genome_seq <- revcomp(genome_seq)
  }
  genome_seq <- unlist(strsplit(genome_seq, split = ""))
  AA_seq<-seqinr::translate(genome_seq)
})

names(proteins) <- unique(cds$Parent)

#proteins
sum(duplicated(names(proteins)))

# Hap 1
h1proteins <- pblapply(unique(h1cds$Parent), function(p){
  h1cds_p <- h1cds[Parent == p]
  dir <- unique(h1cds_p$DIRECTION)
  CHROM = h1cds_p$CHROM[1]
  genome_seq <- paste(sapply(1:nrow(h1cds_p), function(r){
    START = h1cds_p$START[r]
    STOP = h1cds_p$STOP[r]
    paste(fastah1[[CHROM]][START:STOP], collapse="")
  }), collapse = "")
  if(dir == "-"){
    genome_seq <- revcomp(genome_seq)
  }
  genome_seq <- unlist(strsplit(genome_seq, split = ""))
  AA_seq<-seqinr::translate(genome_seq)
})

names(h1proteins) <- unique(h1cds$Parent)

#proteins
sum(duplicated(names(h1proteins)))


# Hap 2
h2proteins <- pblapply(unique(h2cds$Parent), function(p){
  h2cds_p <- h2cds[Parent == p]
  dir <- unique(h2cds_p$DIRECTION)
  CHROM = h2cds_p$CHROM[1]
  genome_seq <- paste(sapply(1:nrow(h2cds_p), function(r){
    START = h2cds_p$START[r]
    STOP = h2cds_p$STOP[r]
    paste(fastah2[[CHROM]][START:STOP], collapse="")
  }), collapse = "")
  if(dir == "-"){
    genome_seq <- revcomp(genome_seq)
  }
  genome_seq <- unlist(strsplit(genome_seq, split = ""))
  AA_seq<-seqinr::translate(genome_seq)
})

names(h2proteins) <- unique(h2cds$Parent)

#proteins
sum(duplicated(names(h2proteins)))
```

### Extract protein DNA sequences
```{r}
# Primary
protein_cds <- pblapply(unique(cds$Parent), function(p){
  cds_p <- cds[Parent == p]
  dir <- unique(cds_p$DIRECTION)
  CHROM = cds_p$CHROM[1]
  genome_seq <- paste(sapply(1:nrow(cds_p), function(r){
    START = cds_p$START[r]
    STOP = cds_p$STOP[r]
    paste(fasta[[CHROM]][START:STOP], collapse = "")
  }), collapse = "")
  if(dir == "-"){
    genome_seq<-revcomp(genome_seq)
  }
  genome_seq<-unlist(strsplit(genome_seq, split=""))
  return(genome_seq)
})

names(protein_cds)<-unique(cds$Parent)

##protein_cds
sum(duplicated(names(protein_cds)))

# Hap 1
protein_h1cds <- pblapply(unique(h1cds$Parent), function(p){
  h1cds_p <- h1cds[Parent == p]
  dir <- unique(h1cds_p$DIRECTION)
  CHROM = h1cds_p$CHROM[1]
  genome_seq <- paste(sapply(1:nrow(h1cds_p), function(r){
    START = h1cds_p$START[r]
    STOP = h1cds_p$STOP[r]
    paste(fastah1[[CHROM]][START:STOP], collapse = "")
  }), collapse = "")
  if(dir == "-"){
    genome_seq<-revcomp(genome_seq)
  }
  genome_seq<-unlist(strsplit(genome_seq, split=""))
  return(genome_seq)
})

names(protein_h1cds)<-unique(h1cds$Parent)

##protein_cds
sum(duplicated(names(protein_h1cds)))

# Hap 2
protein_h2cds <- pblapply(unique(h2cds$Parent), function(p){
  h2cds_p <- h2cds[Parent == p]
  dir <- unique(h2cds_p$DIRECTION)
  CHROM = h2cds_p$CHROM[1]
  genome_seq <- paste(sapply(1:nrow(h2cds_p), function(r){
    START = h2cds_p$START[r]
    STOP = h2cds_p$STOP[r]
    paste(fastah2[[CHROM]][START:STOP], collapse = "")
  }), collapse = "")
  if(dir == "-"){
    genome_seq<-revcomp(genome_seq)
  }
  genome_seq<-unlist(strsplit(genome_seq, split=""))
  return(genome_seq)
})

names(protein_h2cds)<-unique(h2cds$Parent)

##protein_cds
sum(duplicated(names(protein_h2cds)))
```

## Checking sequences
### Make sure protein sequences are correct (start with M, ends with *)
```{r}
#Primary
prot_check <- rbindlist(pblapply(names(proteins), function(p){
  codon1 <- proteins[[p]][1]
  codonend <- proteins[[p]][length(proteins[[p]])]
  return(data.table(p, codon1, codonend))
}), 
fill=T)

prot_check$CHECK <- prot_check$codon1 == "M" & prot_check$codonend == "*"
table(prot_check$CHECK)

# Hap 1
h1prot_check <- rbindlist(pblapply(names(h1proteins), function(p){
  codon1 <- h1proteins[[p]][1]
  codonend <- h1proteins[[p]][length(h1proteins[[p]])]
  return(data.table(p, codon1, codonend))
}), 
fill=T)

h1prot_check$CHECK <- h1prot_check$codon1 == "M" & h1prot_check$codonend == "*"
table(h1prot_check$CHECK)

# Hap 2
h2prot_check <- rbindlist(pblapply(names(h2proteins), function(p){
  codon1 <- h2proteins[[p]][1]
  codonend <- h2proteins[[p]][length(h2proteins[[p]])]
  return(data.table(p, codon1, codonend))
}), 
fill=T)

h2prot_check$CHECK <- h2prot_check$codon1 == "M" & h2prot_check$codonend == "*"
table(h2prot_check$CHECK)
```

### Only keep valid proteins in annotations
```{r}
# Primary
cds_good <- cds[Parent %in% prot_check[CHECK==T]$p]
genes_good <- genes[GENE %in% cds_good$GENE]

nrow(cds)
nrow(cds_good)

nrow(genes)
nrow(genes_good)

# Hap 1
h1cds_good <- h1cds[Parent %in% h1prot_check[CHECK==T]$p]
h1genes_good <- h1genes[GENE %in% h1cds_good$GENE]

nrow(h1cds)
nrow(h1cds_good)

nrow(h1genes)
nrow(h1genes_good)

# Hap 2
h2cds_good <- h2cds[Parent %in% h2prot_check[CHECK==T]$p]
h2genes_good <- h2genes[GENE %in% h2cds_good$GENE]

nrow(h2cds)
nrow(h2cds_good)

nrow(h2genes)
nrow(h2genes_good)
```

## Intergenic space and mappability
### Identify intergenic space
```{r}
# Primary
intergenic <- genes_good[, .(
  TYPE = "intergenic",
  START = STOP + 1L,
  STOP = data.table::shift(START, type = "lead") - 1L
), by = CHROM]

intergenic <- intergenic[!is.na(STOP)]

intergenic <- intergenic[START <= STOP]

intergenic$INFO <- paste("intergenic", 1:nrow(intergenic), sep ="_" )
intergenic$SOURCE <- paste("Manual")
intergenic$SCORE <- paste(".")
intergenic$DIRECTION <- paste(".")
intergenic$PHASE <- paste(".")

# Hap 1
h1intergenic <- h1genes_good[, .(
  TYPE = "intergenic",
  START = STOP + 1L,
  STOP = data.table::shift(START, type = "lead") - 1L
), by = CHROM]

h1intergenic <- h1intergenic[!is.na(STOP)]

h1intergenic <- h1intergenic[START <= STOP]

h1intergenic$INFO <- paste("intergenic", 1:nrow(h1intergenic), sep ="_" )
h1intergenic[, UNIQUE := paste("intergenic", START, STOP, sep ="_" )]
h1intergenic$SOURCE <- paste("Manual")
h1intergenic$SCORE <- paste(".")
h1intergenic$DIRECTION <- paste(".")
h1intergenic$PHASE <- paste(".")
h1intergenic$GENE <- paste("none")

# Hap 2
h2intergenic <- h2genes_good[, .(
  TYPE = "intergenic",
  START = STOP + 1L,
  STOP = data.table::shift(START, type = "lead") - 1L
), by = CHROM]

h2intergenic <- h2intergenic[!is.na(STOP)]

h2intergenic <- h2intergenic[START <= STOP]

h2intergenic$INFO <- paste("intergenic", 1:nrow(h2intergenic), sep ="_" )
h2intergenic[, UNIQUE := paste("intergenic", START, STOP, sep ="_" )]
h2intergenic$SOURCE <- paste("Manual")
h2intergenic$SCORE <- paste(".")
h2intergenic$DIRECTION <- paste(".")
h2intergenic$PHASE <- paste(".")
h2intergenic$GENE <- paste("none")
```

### Add hap column to haplotype gffs
```{r}
h1cds_good[, HAP := "A"]
h1genes_good[, HAP := "A"]

h2cds_good[, HAP := "B"]
h2genes_good[, HAP := "B"]

h1intergenic[, HAP := "A"]
h2intergenic[,HAP := "B"]
```


### Defining mappable regions
```{r}
# Genic space
genes_good$ID<-1:nrow(genes_good)

genes_good$LENGTH <- genes_good$STOP - genes_good$START

mapp_genic<-features_in_features(genes_good, mapp[DEPTH==1], mode = "length")

# Intergenic space
intergenic$ID <- 1:nrow(intergenic)

intergenic$LENGTH <- intergenic$STOP - intergenic$START

mapp_genic <- features_in_features(intergenic, mapp[DEPTH==1], mode = "length")
```

### Ratio of gene to intergenic
```{r}
sum(genes_good$LENGTH)/sum(intergenic$LENGTH)
```

### Add in genic features and intergenic regions into 1 gff
```{r}
# h1gff_good <- h1gff[h1genes_good, on = .(CHROM, START >= START, START <= STOP)]
# nrow(unique(h1gff_good, by = c("TYPE", "CHROM", "START")) %>%
#   filter(TYPE %in% "gene"))
# nrow(unique(h1genes_good, by = c("TYPE", "CHROM", "START")) %>%
#   filter(TYPE %in% "gene"))
# 
# h2gff_good <- h2gff[h2genes_good, on = .(CHROM, START >= START, START <= STOP)]
# nrow(unique(h2gff_good, by = c("i.TYPE", "CHROM", "i.START", "i.STOP")) %>%
#   filter(i.TYPE %in% "gene"))
# nrow(unique(h2genes_good, by = c("TYPE", "CHROM", "START", "STOP")) %>%
#   filter(TYPE %in% "gene"))
# 
# h1gff_good[, HAP := "A"]
# h2gff_good[, HAP := "B"]
# 
# h1gff_good <- rbind(h1gff_good, h1intergenic)
# h2gff_good <- rbind(h2gff_good, h2intergenic)
# 
# h1gff_good[, gene_priority := (TYPE == "gene")]
# setorder(h1gff_good, CHROM, START, -gene_priority)
# h1gff_good[, gene_priority := NULL]
# 
# h1gff_good <- unique(h1gff_good)
# 
# # Ensure both tables are keyed properly for foverlaps
# setorder(h2gff, CHROM, START)
# setorder(h2genes_good, CHROM, START)
# 
# setkey(h2gff, CHROM, START, STOP)
# setkey(h2genes_good, CHROM, START, STOP)
# 
# result <- foverlaps(h2gff, h2genes_good, by.x = c("CHROM", "START", "STOP"), by.y = c("CHROM", "START", "STOP"))
# #h2gff_good <- h2gff_good[, grep("^i\\.", names(h2gff_good)), with = FALSE]
# 
# result[, UNIQUE := paste(CHROM, i.TYPE, i.START, i.STOP, sep = "_")]
# 
# unique(result, by = "UNIQUE")%>%
#   filter(i.TYPE == "gene")
# 
# 
# #######################
# pos <- h1genes_good %>%
#   select(CHROM, START, STOP)
# pos[h1gff, on = .(CHROM, START <= START, STOP >= START)]
# 
# 
# h1gff[, .(CHROM = pos$CHROM, START >= pos$START, START <= pos$STOP)]
# 
# setkey(h1gff, CHROM, START, STOP)
# setkey(pos, CHROM, START, STOP)
# 
# filtered_h1gff <- foverlaps(h1gff, pos, nomatch = 0)
# filtered_h1gff[, UNIQUE := paste(CHROM, TYPE, i.START, i.STOP, sep = "_")]
# 
# uniq_filt_h1gff <- unique(filtered_h1gff, by = "UNIQUE")
# 
# uniq_filt_h1gff %>%
#   filter(TYPE %in% "gene")
# h1genes_good %>%
#   filter(TYPE %in% "gene")
# 
# ################### This one seems to work
h1gff$GENE <- gsub(".+;gene=(LOC[^;]+).*", "\\1", h1gff$INFO)
h1gff_good <- h1gff[h1gff$GENE %in% h1genes_good$GENE, ]

h1gff_good[, UNIQUE := paste(TYPE, CHROM, START, STOP, sep = "_")]

unique(h1gff_good, by = "UNIQUE") %>%
  filter(TYPE == "gene")

h1gff_good[, gene_priority := (TYPE == "gene")]
setorder(h1gff_good, CHROM, START, -gene_priority)
h1gff_good[, gene_priority := NULL]
h1gff_good[, HAP := "A"]

h1gff_good_uniq <- unique(h1gff_good, by = "UNIQUE")

h2gff$GENE <- gsub(".+;gene=(LOC[^;]+).*", "\\1", h2gff$INFO)
h2gff_good <- h2gff[h2gff$GENE %in% h2genes_good$GENE, ]

h2gff_good[, UNIQUE := paste(TYPE, CHROM, START, STOP, sep = "_")]

unique(h2gff_good, by = "UNIQUE") %>%
  filter(TYPE == "gene")

h2gff_good[, gene_priority := (TYPE == "gene")]
setorder(h2gff_good, CHROM, START, -gene_priority)
h2gff_good[, gene_priority := NULL]
h2gff_good[, HAP := "B"]

h2gff_good_uniq <- unique(h2gff_good, by = "UNIQUE")
```

### Calculate UTRs
```{r}
# Hap 1
h1utr_rows <- h1gff_good_uniq[, {
  gene_start <- START[TYPE == "gene"][1]   # Get start of gene
  cds_start <- START[TYPE == "CDS"][1]     # Get first CDS start
  
  if (!is.na(gene_start) & !is.na(cds_start)) {
    utr_start <- gene_start
    utr_stop <- cds_start - 1
    
    # Only add the row if STOP - START > 1
    if (utr_stop - utr_start > 1) {
      list(
        CHROM = CHROM[1],
        SOURCE = SOURCE[1],
        TYPE = "UTR",
        START = utr_start,
        STOP = utr_stop,
        SCORE = ".",
        DIRECTION = DIRECTION[1],
        PHASE = ".",
        INFO = paste0("ID=utr-", GENE[1], ";Parent=gene-", GENE[1]),
        UNIQUE = paste0("UTR_", CHROM[1], "_", utr_start, "_", utr_stop),
        HAP = HAP[1]
      )
    } else {
      NULL
    }
  } else {
    NULL
  }
}, by = GENE]

h1utr_rows[, TYPE := fifelse(DIRECTION == "+", "5UTR", "3UTR")]

h1gff_good_utr <- rbind(h1gff_good_uniq, h1utr_rows)

setorder(h1gff_good_utr, CHROM, START)

# Hap 2
h2utr_rows <- h2gff_good_uniq[, {
  gene_start <- START[TYPE == "gene"][1]   # Get start of gene
  cds_start <- START[TYPE == "CDS"][1]     # Get first CDS start
  
  if (!is.na(gene_start) & !is.na(cds_start)) {
    utr_start <- gene_start
    utr_stop <- cds_start - 1
    
    # Only add the row if STOP - START > 1
    if (utr_stop - utr_start > 1) {
      list(
        CHROM = CHROM[1],
        SOURCE = SOURCE[1],
        TYPE = "UTR",
        START = utr_start,
        STOP = utr_stop,
        SCORE = ".",
        DIRECTION = DIRECTION[1],
        PHASE = ".",
        INFO = paste0("ID=utr-", GENE[1], ";Parent=gene-", GENE[1]),
        UNIQUE = paste0("UTR_", CHROM[1], "_", utr_start, "_", utr_stop),
        HAP = HAP[1]
      )
    } else {
      NULL
    }
  } else {
    NULL
  }
}, by = GENE]

h2utr_rows[, TYPE := fifelse(DIRECTION == "+", "5UTR", "3UTR")]

h2gff_good_utr <- rbind(h2gff_good_uniq, h2utr_rows)

setorder(h2gff_good_utr, CHROM, START)
```

### Calculate introns
```{r}
# Hap 1
h1intron_rows <- h1gff_good_utr[TYPE == "exon", .SD[order(START)], by = GENE][, {
  if (.N > 1) {  # Only process if there are at least two exons
    intron_start <- STOP[-.N] + 1  # Intron starts after previous exon STOP
    intron_stop <- START[-1] - 1   # Intron stops before next exon START
    
    .(CHROM = CHROM[1],
      SOURCE = SOURCE[1],
      TYPE = "intron",
      START = intron_start,
      STOP = intron_stop,
      SCORE = ".",
      DIRECTION = DIRECTION[1],
      PHASE = ".",
      INFO = paste0("ID=intron-", GENE[1], ";Parent=gene-", GENE[1]),
      UNIQUE = paste0("intron_", CHROM[1], "_", intron_start, "_", intron_stop),
      HAP = HAP[1])
  } else {
    NULL  # Skip genes with only one exon
  }
}, by = GENE]

h1gff_good_intron <- rbind(h1gff_good_utr, h1intron_rows)

setorder(h1gff_good_intron, CHROM, START)

# Hap 2
h2intron_rows <- h2gff_good_utr[TYPE == "exon", .SD[order(START)], by = GENE][, {
  if (.N > 1) {  # Only process if there are at least two exons
    intron_start <- STOP[-.N] + 1  # Intron starts after previous exon STOP
    intron_stop <- START[-1] - 1   # Intron stops before next exon START
    
    .(CHROM = CHROM[1],
      SOURCE = SOURCE[1],
      TYPE = "intron",
      START = intron_start,
      STOP = intron_stop,
      SCORE = ".",
      DIRECTION = DIRECTION[1],
      PHASE = ".",
      INFO = paste0("ID=intron-", GENE[1], ";Parent=gene-", GENE[1]),
      UNIQUE = paste0("intron_", CHROM[1], "_", intron_start, "_", intron_stop),
      HAP = HAP[1])
  } else {
    NULL  # Skip genes with only one exon
  }
}, by = GENE]

h2gff_good_intron <- rbind(h2gff_good_utr, h2intron_rows)

setorder(h2gff_good_intron, CHROM, START)
```

### Add in intergenic regions
```{r}
h1gff_good_int <- rbind(h1gff_good_intron, h1intergenic)
setorder(h1gff_good_int, CHROM, START)

h2gff_good_int <- rbind(h2gff_good_intron, h2intergenic)
setorder(h2gff_good_int, CHROM, START)
```

### Write out the gffs of verified genes, UTR, intron, intergenic space 
```{r}
fwrite(h1gff_good_int, "./Data/GFFs/Parsed/hap1_gff_good.tsv", sep = "\t")
fwrite(h2gff_good_int, "./Data/GFFs/Parsed/hap2_gff_good.tsv", sep = "\t")
```

## Identify Homopolymers
### Generate Homopolymer data
```{r, eval=FALSE}
# Identify HPs in the genome
HPs<-make_homopolymer(fasta, 3)
fwrite(HPs,"./Data/Homopolymers/ref_chandler_primary_omnic.fasta.HPs.tsv", sep = "\t")
```

### Read homopolymer data in
```{r}
HPs <- fread("./Data/Homopolymers/ref_chandler_primary_omnic.fasta.HPs.tsv")
```

# Mutations
## Preparing data
### Read data
```{r}
mutations <- fread("./Data/VCFs/Filtered/all_dn_snps.vcf")
# mutations_ns <-fread("./Data/VCFs/Filtered/dn_snps_no_share.vcf")
```

### Rename default column
```{r}
mutations <- mutations %>%
  rename("SAMPLE"= "default", "SOURCE" = "Source") %>%
  dplyr::select(-c(ID, INFO, Hap, unique))

# mutations_ns <- mutations_ns %>%
#   rename("SAMPLE" = "default") %>%
#   select(-c(ID, INFO, Hap, unique))
```

### Adjusting the unique column and identifying SNPs
```{r}
# Mutations
mutations$UNIQUE <- paste(mutations$CHROM, mutations$POS, mutations$ALT, sep = "_")
mutations[, NUMIND := .N, by = UNIQUE] # NUMIND column is the number of Sources with that mutation
SNPS <- nchar(mutations$REF) == 1 & nchar(mutations$ALT) == 1
mutations$SNP <- SNPS

# No share mutations
# mutations_ns$UNIQUE <- paste(mutations_ns$CHROM, mutations_ns$POS, mutations_ns$ALT, sep = "_")
# mutations_ns[, N:=.N, by = UNIQUE]
# SNPS_ns <- nchar(mutations_ns$REF)==1 & nchar(mutations_ns$ALT)==1
# mutations_ns$SNP <- SNPS_ns
```

## Annotate the mutations with CDS and Genic regions
### Annotating with CDS
```{r}
# Mutations
mutations$ID <- 1:nrow(mutations)
mutations_cds <- features_in_sites(cds_good, mutations)
mutations$CDS <- mutations_cds$overlaps[match(mutations$ID, mutations_cds$ID)]

table(mutations$CDS)
# No share mutations
# mutations_ns$ID<-1:nrow(mutations_ns)
# mutations_ns_cds<-features_in_sites(cds_good, mutations_ns)
# mutations_ns$CDS<-mutations_ns_cds$overlaps[match(mutations_ns$ID, mutations_ns_cds$ID)]
```

### Annotating with genes
```{r}
# Mutations
mutations_gene <- features_in_sites(genes_good, mutations)
mutations$GENIC <- mutations_gene$overlaps[match(mutations$ID, mutations_gene$ID)]

table(mutations$GENIC)
# No share mutations
# mutations_ns_gene<-features_in_sites(genes_good, mutations_ns)
# mutations_ns$GENIC<-mutations_ns_gene$overlaps[match(mutations_ns$ID, mutations_ns_gene$ID)]
```

## Split AD
```{r}
mutations$ALTDP <- sapply(mutations$AD, function(x) as.numeric(vstrsplit(x, split=",")[2]))
mutations$REFDP <- sapply(mutations$AD, function(x) as.numeric(vstrsplit(x, split=",")[1]))
```

## Calculation of deviation from heterozygous expection
```{r, warning=FALSE}
# Does the Chi square test warning matter?
mutations$CHIPHET <- pbsapply(1:nrow(mutations), function(i) {
  
  ALTDP <- mutations$ALTDP[1]
  REFDP <- mutations$REFDP[1]
  chitest <- chisq.test(c(ALTDP, REFDP))
  
  return(chitest$p.value)
})
```

## Distribution of shared SNPs
```{r}
mutations %>%
  ggplot(aes(x = as.factor(NUMIND))) +
  geom_histogram(stat = "count") +
  theme_classic()

counts<-data.table(table(table(mutations$UNIQUE)))
ggplot(counts, aes(x=as.numeric(V1), y=N))+
  geom_bar(stat="identity")+
  scale_y_log10()
```

## Recalculating DP and VAF
### Calculate DP based on informative reads and normalize
```{r}
mutations$DPREAL <- mutations$ALTDP + mutations$REFDP

mutations[, DPNORM:=log(DP/mean(DP)), by = SOURCE]
mutations[, DPNORM_REAL:=log(DPREAL/mean(DPREAL)), by = SOURCE]
```

### Calculate VAF based on informative reads
```{r}
mutations$VAFREAL <- mutations$ALTDP/mutations$DPREAL
```

## Assign labels for samples
```{r}
mutations <- mutations %>%
  mutate(CULTURETYPE = case_when(
    SOURCE %in% c("ref_chandler", "tree2") ~ "tree",
    SOURCE %in% c("cr2_hifi", "cr21_1", "cr21_2", "cr2_11A", "cr2_12A", "cr2_13A", "cr2_15A",
                  "cr2_15A1", "cr2_16A", "cr2_16A1", "cr2_16A2", "cr2_17A1", "cr2_17A2",
                  "cr2_18A", "cr2_18A1") ~ "embryo",
    SOURCE %in% c("cr85", "cr10", "cr13", "cr22") ~ "shoot")) %>%
  mutate(SEQTYPE = case_when(
    SOURCE %in% c("ref_chandler", "tree2", "cr2_hifi") ~ "long_read",
    SOURCE %in% c("cr21_1", "cr21_2", "cr2_11A", "cr2_12A", "cr2_13A", "cr2_15A", "cr2_15A1", 
                  "cr2_16A", "cr2_16A1","cr2_16A2", "cr2_17A1", "cr2_17A2", "cr2_18A",
                  "cr2_18A1", "cr85", "cr10", "cr13", "cr22") ~ "short_read"))
```

## Calculate summary stats for each SNP across all samples
```{r}
# Reset object as data table to avoid warnings
mutations <- as.data.table(mutations)
# Mean VAF
mutations[, meanVAF:=mean(VAF), by = UNIQUE]
mutations[, meanVAFREAL:=mean(VAFREAL), by = UNIQUE]

# Median depth
mutations[, medDP:=median(DP), by = UNIQUE]
mutations[, medDPREAL:=median(DPREAL), by = UNIQUE]

# Median Quality
mutations[, medQUAL:=median(QUAL), by = UNIQUE]
```

## Annotate mappability to means
```{r}
mutations_MAPP <- features_in_sites(mapp[DEPTH==1], mutations)
mutations$MAPP <- mutations_MAPP$overlaps[match(mutations$ID, mutations_MAPP$ID)]
table(mutations$MAPP, mutations$SEQTYPE)
```

<!-- ## Annotate mutations near homopolymeric regions -->
<!-- ```{r} -->
<!-- # Seems to be about 40 minutes -->
<!-- mutations_neighbors <- rbindlist(pblapply((1:nrow(mutations)), function(i){ -->
<!--   CHROM = mutations$CHROM[i] -->
<!--   POS = mutations$POS[i] -->
<!--   UNIQUE = mutations$UNIQUE[i] -->
<!--   up = paste0(toupper(fasta[[CHROM]][(POS-10):(POS-1)]), collapse="") -->
<!--   down = paste0(toupper(fasta[[CHROM]][(POS+1):(POS+10)]), collapse="") -->
<!--  return(data.table(UNIQUE, up, down)) -->
<!-- })) -->

<!-- mutations<-merge(mutations, mutations_neighbors, by="UNIQUE") -->
<!-- ``` -->

## Annotate trimer context for mutations
```{r}
# mutations contexts are only created for SNPs with this function, not INDELs. INDELS are filled with NAs.
# Takes around 15 minutes
mutations$TRICONTEXT <- tricontexts(mutations, fasta)
mutations$MUT <- paste0(substr(mutations$TRICONTEXT, 2, 2), substr(mutations$TRICONTEXT, 4, 4), substr(mutations$TRICONTEXT, 6, 6))

mutations$MUT[mutations$MUT == "NANANA"] <- NA
```

<!-- ## Calculate distance -->
<!-- ```{r} -->

<!-- ## run post filter -->
<!-- mutations_split<-split(mutations, by=c("CHROM","SOURCE")) -->
<!-- mutations$DIST<- pbsapply(1:nrow(mutations), function(i){ -->
<!--   C = mutations$CHROM[i] -->
<!--   P = mutations$POS[i] -->
<!--   S = mutations$SOURCE[i] -->
<!--   listcoord<-paste0(C,".",S) -->
<!--   other <- mutations_split[[listcoord]][POS!=P] -->
<!--   min(abs(other$POS-P)) -->
<!-- }) -->
<!-- ``` -->

## Organize mutations to save
```{r}
mutations_save <- mutations %>%
  select(CHROM, POS, REF, ALT, UNIQUE, QUAL, FILTER, GT, GQ, REFDP, ALTDP, DPREAL, VAFREAL, DPNORM_REAL, medQUAL, medDPREAL, meanVAFREAL, SNP, GENIC, CDS, NUMIND, CULTURETYPE, SEQTYPE, MAPP, MUT, TRICONTEXT, SOURCE)
```

# Write out new files
```{r, eval=T}
fwrite(mutations_save, "./Data/VCFs/Parsed/mutations.tsv", sep = "\t")
fwrite(cds_good, "./Data/GFFs/Parsed/cds_good.tsv", sep = "\t")
fwrite(genes_good, "./Data/GFFs/Parsed/genes_good.tsv", sep = "\t")
fwrite(h1cds_good, "./Data/GFFs/Parsed/hap1_cds_good.tsv", sep = "\t")
fwrite(h1genes_good, "./Data/GFFs/Parsed/hap1_genes_good.tsv", sep = "\t")
fwrite(h2cds_good, "./Data/GFFs/Parsed/hap2_cds_good.tsv", sep = "\t")
fwrite(h2genes_good, "./Data/GFFs/Parsed/hap2_genes_good.tsv", sep = "\t")
```

# Try to implement HPs
## Clear environment to save memory
```{r}
rm(list = ls())
```

## Read in saved mutations and HPs
```{r}
muts <- fread("./Data/VCFs/Parsed/mutations.tsv")
fasta <- read.fasta("./Data/Fastas/ref_chandler_primary_default_scaffold.fasta")
HPs <- fread("./Data/Homopolymers/ref_chandler_primary_omnic.fasta.HPs.tsv")
```

## Select muts to make as small as possible
```{r}
muts_small <- muts %>%
  select(CHROM, POS, UNIQUE)
```


## Run HPs
## Annotate mutations near homopolymeric regions
```{r}
# Seems to be about 40 minutes
mutations_neighbors <- rbindlist(pblapply((1:nrow(muts_small)), function(i){
  CHROM = muts_small$CHROM[i]
  POS = muts_small$POS[i]
  UNIQUE = muts_small$UNIQUE[i]
  up = paste0(toupper(fasta[[CHROM]][(POS-10):(POS-1)]), collapse="")
  down = paste0(toupper(fasta[[CHROM]][(POS+1):(POS+10)]), collapse="")
  return(data.table(UNIQUE, up, down))
}))

muts_hp <- merge(muts_small, mutations_neighbors, by="UNIQUE")
```

# Create a trimmed down set of shared ancestral snps
## Function for reading ancestral snps
```{r}
read.ANC <- function(path, genotype = c("0/0","0/1","1/1")) {  
  # Load data
  snps <- fread(path)
  
  setnames(snps, "Source", "SOURCE")
  
  # Identify SNPs with single-character REF and ALT
  snps[, SNP := nchar(REF) == 1 & nchar(ALT) == 1]
  
  # Filter SNPs with heterozygous genotype and calculate DPNORM by SOURCE
  het_snps <- snps[SNP == TRUE & GT %in% genotype]
  
  # Select relevant columns
  small_snps <- het_snps[, .(CHROM, POS, REF, ALT, GT, VAF, AD, DP, QUAL, SOURCE)]
  
  # Split AD column into REFDP and ALTDP
  small_snps[, c("REFDP", "ALTDP") := tstrsplit(AD, ",", type.convert = TRUE)]
  
  # Create REALDP Column
  small_snps[, REALDP := REFDP + ALTDP]
  
  # Normalize REFDP and ALTDP within each SOURCE
  small_snps[, `:=`(
    REFDP_NORM = REFDP / mean(REFDP, na.rm = TRUE),
    ALTDP_NORM = ALTDP / mean(ALTDP, na.rm = TRUE),
    REAL_DPNORM = REALDP / mean(REALDP, na.rm = TRUE)
  ), by = SOURCE]
  
  # Drop the original AD column
  small_snps[, AD := NULL]
  small_snps[, DP := NULL]
  
  return(small_snps)
}
```

## Read and combine
```{r}
chr1 <- read.ANC("./Data/VCFs/Filtered/all_an_snps_chr_1.vcf")
chr2 <- read.ANC("./Data/VCFs/Filtered/all_an_snps_chr_2.vcf")
chr3 <- read.ANC("./Data/VCFs/Filtered/all_an_snps_chr_3.vcf")
chr4 <- read.ANC("./Data/VCFs/Filtered/all_an_snps_chr_4.vcf")
chr5 <- read.ANC("./Data/VCFs/Filtered/all_an_snps_chr_5.vcf")
chr6 <- read.ANC("./Data/VCFs/Filtered/all_an_snps_chr_6.vcf")
chr7 <- read.ANC("./Data/VCFs/Filtered/all_an_snps_chr_7.vcf")
chr8 <- read.ANC("./Data/VCFs/Filtered/all_an_snps_chr_8.vcf")
chr9 <- read.ANC("./Data/VCFs/Filtered/all_an_snps_chr_9.vcf")
chr10 <- read.ANC("./Data/VCFs/Filtered/all_an_snps_chr_10.vcf")
chr11 <- read.ANC("./Data/VCFs/Filtered/all_an_snps_chr_11.vcf")
chr12 <- read.ANC("./Data/VCFs/Filtered/all_an_snps_chr_12.vcf")
chr13 <- read.ANC("./Data/VCFs/Filtered/all_an_snps_chr_13.vcf")
chr14 <- read.ANC("./Data/VCFs/Filtered/all_an_snps_chr_14.vcf")
chr15 <- read.ANC("./Data/VCFs/Filtered/all_an_snps_chr_15.vcf")
chr16 <- read.ANC("./Data/VCFs/Filtered/all_an_snps_chr_16.vcf")

an_snps <- rbind(chr1, chr2, chr3, chr4, chr5, chr6, chr7, chr8, chr9, chr10, chr11, chr12, chr13, chr14, chr15, chr16)
rm(chr1, chr2, chr3, chr4, chr5, chr6, chr7, chr8, chr9, chr10, chr11, chr12, chr13, chr14, chr15, chr16)

fwrite(an_snps, "./Data/VCFs/Ancestral_SNPs/ancestral_snps_smaller.tsv")
```

# Structural Variants
## Libraries
```{r, message=FALSE}
library(ggenomics)
library(polymorphology2)
library(ggVennDiagram)
library(vcfR)
```

## Functions
```{r}
read_pbsv <- function(file, chrom_label = "Chr", col_to_separate = "UnnamedSample", name){
  # Read the VCF file using vcfR
  vcf <- read.vcfR(file, verbose = FALSE)
  
  # Convert @fix and @gt slots to data.tables
  fix <- as.data.table(vcf@fix)
  gt <- as.data.table(vcf@gt)
  all <- cbind(fix, gt)
  
  # Select chromosomes
  vcf.table <- all[grepl(chrom_label, CHROM), ]
  
  # Remove columns with no information
  vcf.table[, c("QUAL") := NULL]
  
  # Select passing variants then remove FILTER
  vcf.table <- vcf.table[FILTER == "PASS"]
  vcf.table[, FILTER := NULL]
  
  # Separate genotype information into distinct columns
  ## Need to verify this is working properly
  vcf.table <- vcf.table[, c("GT", "AD", "DP", "SAC") := tstrsplit(get(col_to_separate), ":", type.convert = TRUE)]
  
  # Removing unresolved SVs. Need to find a better way to do
  vcf.table <- vcf.table[!grepl("pbsv.BND", ID), ] 
  vcf.table <- vcf.table[!grepl("IMPRECISE", INFO), ]
  vcf.table <- vcf.table[!grepl("NotFullySpanned", INFO), ]
  
  # Separate sv information
  vcf.table <- vcf.table[, c("SVTYPE", "END", "SVLEN") := tstrsplit(get("INFO"), ";", type.convert = TRUE)]
  
  # Remove extraneous info from columns
  vcf.table$SVTYPE <- sub(".*=", "", vcf.table$SVTYPE)
  vcf.table$END <- sub(".*=", "", vcf.table$END)
  vcf.table$SVLEN <- sub(".*=", "", vcf.table$SVLEN)
  
  # Separate allele depths into ref and alt
  vcf.table[, c("REFAD", "ALTAD") := tstrsplit(AD, ",")]
  vcf.table[, AD := NULL]
  
  # Remove the DP
  vcf.table[, DP := NULL]
  
  # Change classes
  vcf.table[, END := as.numeric(END)]
  vcf.table[, SVLEN := as.numeric(SVLEN)]
  vcf.table[, REFAD := as.numeric(REFAD)]
  vcf.table[, ALTAD := as.numeric(ALTAD)]
  
  # Create unique identifier
  vcf.table[, UNIQUE := paste(CHROM, POS, ALT, sep = "_")]
  
  # Recalculate DEPTH and VAF columns
  vcf.table[, VAF := ALTAD/(REFAD + ALTAD)]
  vcf.table[, DEPTH := REFAD + ALTAD]
  
  # Add source column
  vcf.table[, SOURCE := name]
  
  # Remove extraneous columns to save space
  vcf.table[, UnnamedSample := NULL]
  vcf.table[, FORMAT := NULL]
  vcf.table[, INFO := NULL]
  
  # Filter for Insertions and deletions
  vcf.table <- vcf.table[SVTYPE %in% c("INS", "DEL")]
  
  # Change column order
  setcolorder(vcf.table, c("CHROM", "POS", "END", "ID", "REF", "ALT", "GT", "SVTYPE", "SVLEN", "REFAD", "ALTAD", "DEPTH", "VAF", "SAC", "UNIQUE", "SOURCE"))
  
  return(vcf.table)
  
}

rename_chr_haps <- function(pbsv){
  
  pbsv$HAP <- gsub("NC_0499(..).1_RagTag_(hap.)", "\\2", pbsv$CHROM)
  
  pbsv$CHROM <- gsub("NC_0499(..).1_RagTag_(hap.)", "\\1", pbsv$CHROM)
  
  pbsv$CHROM <- as.numeric(pbsv$CHROM)
  
  pbsv <- pbsv %>% 
    filter(!grepl("h1t", CHROM)) %>%
    filter(!grepl("h2t", CHROM)) %>%
    filter(!grepl("NW_", CHROM))
  
  pbsv$HAP <- gsub("hap1", "A", pbsv$HAP)
  pbsv$HAP <- gsub("hap2", "B", pbsv$HAP)
  
  return(pbsv)
}
```

## Read in data
```{r}
cr2sv <- read_pbsv("./Data/VCFs/SV_VCF/cr2_combohap.sv.vcf", chrom_label = "NC_0499", name = "cr2_hifi")
refsv <- read_pbsv("./Data/VCFs/SV_VCF/ref_chandler_combohap.sv.vcf", chrom_label = "NC_0499", name = "ref_chandler")
termsv <- read_pbsv("./Data/VCFs/SV_VCF/term_chandler_combohap.sv.vcf", chrom_label = "NC_0499", name = "term_chandler")
```

## Filtering data
## Select for chromosomes and rename
```{r}
cr2sv <- rename_chr_haps(cr2sv)
refsv <- rename_chr_haps(refsv)
termsv <- rename_chr_haps(termsv)
```

## Create UNIQUE column
```{r}
cr2sv <- cr2sv %>%
  mutate(UNIQUE = paste(CHROM,HAP,POS, sep = "_"))

refsv <- refsv %>%
  mutate(UNIQUE = paste(CHROM,HAP,POS, sep = "_"))

termsv <- termsv %>%
  mutate(UNIQUE = paste(CHROM,HAP,POS, sep = "_"))
```

## Look at overlap of structural variants
```{r}
## Make a list
sv_list <- list(
  cr2 = cr2sv$UNIQUE,
  ref = refsv$UNIQUE,
  term = termsv$UNIQUE
)

# Create the Venn diagram
ggVennDiagram(sv_list)
```

## Find denovo sites
```{r}
dn_cr2sv <- cr2sv %>%
  anti_join(refsv, by = "UNIQUE") %>%
  anti_join(termsv, by = "UNIQUE")

dn_refsv <- refsv %>%
  anti_join(cr2sv, by = "UNIQUE") %>%
  anti_join(termsv, by = "UNIQUE")

dn_termsv <- termsv %>%
  anti_join(refsv, by = "UNIQUE") %>%
  anti_join(cr2sv, by = "UNIQUE")

dn_sv <- rbind(dn_cr2sv, dn_refsv, dn_termsv)
```

## Verify de novo sites
```{r}
## Make a list
dn_sv_list <- list(
  cr2 = dn_cr2sv$UNIQUE,
  ref = dn_refsv$UNIQUE,
  term = dn_termsv$UNIQUE
)

# Create the Venn diagram
ggVennDiagram(dn_sv_list)
```

## Write the insertion and deletion table
```{r}
fwrite(dn_sv, "./Data/VCFs/SV_VCF/Parsed/parsed_ins_dels_pbsv.vcf", sep = "\t")
```

# Parsing primary SVs
## Functions
```{r}
rename_chr_pri <- function(pbsv){

  pbsv$CHROM <- gsub("NC_0499(..).1_RagTag", "\\1", pbsv$CHROM)
  
  pbsv$CHROM <- as.numeric(pbsv$CHROM)
  
  pbsv <- pbsv %>% 
    filter(!grepl("h1t", CHROM))
  
  return(pbsv)
}
```

## Read svs
```{r}
cr2sv <- read_pbsv("./Data/VCFs/SV_VCF/cr2_primary.sv.vcf", chrom_label = "NC_0499", name = "cr2_hifi")
refsv <- read_pbsv("./Data/VCFs/SV_VCF/ref_chandler_primary.sv.vcf", chrom_label = "NC_0499", name = "ref_chandler")
termsv <- read_pbsv("./Data/VCFs/SV_VCF/term_chandler_primary.sv.vcf", chrom_label = "NC_0499", name = "term_chandler")
```

## Filtering data
## Select for chromosomes and rename
```{r}
cr2sv <- rename_chr_pri(cr2sv)
refsv <- rename_chr_pri(refsv)
termsv <- rename_chr_pri(termsv)
```

## Create UNIQUE column
```{r}
cr2sv <- cr2sv %>%
  mutate(UNIQUE = paste(CHROM,POS, sep = "_"))

refsv <- refsv %>%
  mutate(UNIQUE = paste(CHROM,POS, sep = "_"))

termsv <- termsv %>%
  mutate(UNIQUE = paste(CHROM,POS, sep = "_"))
```

## Look at overlap of structural variants
```{r}
## Make a list
sv_list <- list(
  cr2 = cr2sv$UNIQUE,
  ref = refsv$UNIQUE,
  term = termsv$UNIQUE
)

# Create the Venn diagram
ggVennDiagram(sv_list)
```

## Find denovo sites
```{r}
dn_cr2sv <- cr2sv %>%
  anti_join(refsv, by = "UNIQUE") %>%
  anti_join(termsv, by = "UNIQUE")

dn_refsv <- refsv %>%
  anti_join(cr2sv, by = "UNIQUE") %>%
  anti_join(termsv, by = "UNIQUE")

dn_termsv <- termsv %>%
  anti_join(refsv, by = "UNIQUE") %>%
  anti_join(cr2sv, by = "UNIQUE")

dn_sv <- rbind(dn_cr2sv, dn_refsv, dn_termsv)
```

## Verify de novo sites
```{r}
## Make a list
dn_sv_list <- list(
  cr2 = dn_cr2sv$UNIQUE,
  ref = dn_refsv$UNIQUE,
  term = dn_termsv$UNIQUE
)

# Create the Venn diagram
ggVennDiagram(dn_sv_list)
```

```{r}
dn_sv %>%
  ggplot(aes(x = SVLEN)) +
  geom_histogram() +
  facet_wrap(~SOURCE)

min(abs(dn_sv$SVLEN))
```


## Write the insertion and deletion table
```{r}
fwrite(dn_sv, "./Data/VCFs/SV_VCF/Parsed/parsed_ins_dels_pbsv_primary.vcf", sep = "\t")
```

# Parse TE BLAST location tables
## Read in data
```{r}
te_5453 <- fread("./Data/Blast_Tables/all_5000_trimmed_blastout.txt")
te_899 <- fread("./Data/Blast_Tables/all_900_trimmed_blastout.txt")
all_sv <- fread("./Data/Blast_Tables/dn_sv_10bp_trimmed_topmatch_blastout.txt")
```

## Add column headers
```{r}
colnames(te_5453) <- c("QSEQID", "SSEQID", "PCT_IDENTITY", "LENGTH", "MISMATCH", "GAPOPEN", "QSTART", "QEND", "SSTART", "SEND", "EVALUE", "BITSCORE")
colnames(te_899) <- c("QSEQID", "SSEQID", "PCT_IDENTITY", "LENGTH", "MISMATCH", "GAPOPEN", "QSTART", "QEND", "SSTART", "SEND", "EVALUE", "BITSCORE")
colnames(all_sv) <- c("QSEQID", "SSEQID", "PCT_IDENTITY", "LENGTH", "MISMATCH", "GAPOPEN", "QSTART", "QEND", "SSTART", "SEND", "EVALUE", "BITSCORE")
```


## Rename Chromosomes and create new hap column
```{r}
te_5453[, HAP := sub("NC_0499([0-9]+).1_RagTag_(hap[0-9]+)", "\\2", SSEQID)]
te_5453[, HAP := fifelse(HAP == "hap1", "A", "B")]
te_5453[, CHROM := as.numeric(sub("NC_0499([0-9]+).1_RagTag_(hap[0-9]+)", "\\1", SSEQID))]

te_899[, HAP := sub("NC_0499([0-9]+).1_RagTag_(hap[0-9]+)", "\\2", SSEQID)]
te_899[, HAP := fifelse(HAP == "hap1", "A", "B")]
te_899[, CHROM := as.numeric(sub("NC_0499([0-9]+).1_RagTag_(hap[0-9]+)", "\\1", SSEQID))]

all_sv <- all_sv[grepl("^NC_0499", SSEQID)]
all_sv[, HAP := sub("NC_0499([0-9]+).1_RagTag_(hap[0-9]+)", "\\2", SSEQID)]
all_sv[, HAP := fifelse(HAP == "hap1", "A", "B")]
all_sv[, CHROM := as.numeric(sub("NC_0499([0-9]+).1_RagTag_(hap[0-9]+)", "\\1", SSEQID))]
```

## Assign strands
The orientation of the strand affects which bp is start and end
```{r}
# Create strand column
all_sv[, STRAND := fifelse(SSTART < SEND, "+", "-")]
te_899[, STRAND := fifelse(SSTART < SEND, "+", "-")]
te_5453[, STRAND := fifelse(SSTART < SEND, "+", "-")]

# Make sure start is lower than end
all_sv[SSTART > SEND, c("SSTART", "SEND") := .(SEND, SSTART)]
te_899[SSTART > SEND, c("SSTART", "SEND") := .(SEND, SSTART)]
te_5453[SSTART > SEND, c("SSTART", "SEND") := .(SEND, SSTART)]
```

## Create SVLEN column
```{r}
all_sv[, SVLEN := as.numeric(sub("([0-9]+)_.*", "\\1", QSEQID))]
te_5453[, SVLEN := as.numeric(sub("([0-9]+)_.*", "\\1", QSEQID))]
te_899[, SVLEN := as.numeric(sub("([0-9]+)_.*", "\\1", QSEQID))]
```

## Look at two odd TEs
```{r}
te_5453 %>%
  group_by(QSEQID) %>%
  filter(QSEQID %in% c("5454_7_42221860_A_3", "5454_11_4069930_B_8") ) %>%
  mutate(MATCH = QEND - QSTART) %>%
  slice_max(order_by = BITSCORE, n = 15)
```


## Select sequences with highest bitscore
```{r}
top_te_5453 <- te_5453[ , .SD[which.max(BITSCORE)], by = QSEQID]

top_te_899 <- te_899[ , .SD[which.max(BITSCORE)], by = QSEQID]

top_sv <- all_sv[, .SD[which.max(BITSCORE)], by = QSEQID]
```

## Write out top hit tables
```{r}
fwrite(top_te_5453, "./Data/Blast_Tables/Parsed/te_5453_table.tsv", sep = "\t")
fwrite(top_te_899, "./Data/Blast_Tables/Parsed/te_899_table.tsv", sep = "\t")
fwrite(top_sv, "./Data/Blast_Tables/Parsed/sv_table.tsv", sep = "\t")
```

# Create fasta table for TEs
## Read in fasta
```{r}
fasta_5000 <- ggread_fasta("./Data/Fastas/TEs/all_5000_trimmed.fasta")
fasta_900 <- ggread_fasta("./Data/Fastas/TEs/all_900_trimmed.fasta")
```

## Convert to table
```{r}
table_5000 <- data.table(NAME = names(fasta_5000), SEQUENCE = as.character(fasta_5000))
table_900 <- data.table(NAME = names(fasta_900), SEQUENCE = as.character(fasta_900))
```

## Write out the fasta
```{r}
fwrite(table_5000, "./Data/Fastas/TEs/all_5000_fasta_table.tsv", sep = "\t")
fwrite(table_900, "./Data/Fastas/TEs/all_900_fasta_table.tsv", sep = "\t")
```

