---
title: "rna_seq_expression"
author: "Matthew Davis"
date: "2024-07-17"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

# Setup
## Libraries
```{r, message=FALSE}
library(data.table)
library(tidyverse)
library(ggpubr)
```

# STAR data
## Functions
```{r}
# read_rna_counts <- function(path, source){
#   
#   # Read in data
#   table <- fread(path)
# 
#   # Manipulate table
#   colnames(table)[7] <- "Counts" # Change column 7 to counts
#   table$Chr <- sub(";.*", "",table$Chr) # Keep entry for first exon
#   table$Strand <- sub(";.*", "",table$Strand) # Keep first entry for strand
#   table$Start <- sub(";.*", "",table$Start) # Keep first entry for start
#   table$End <- sub(".*;([^;]+)$", "\\1",table$End) # keep last entry for end. This gives total gene position. Lengths aren't quite right though.
#   
#   # Add Source column
#   table <- table %>%
#     mutate(Source = source)
#   
#   # Rename chromosome names
#   table <- table %>%
#     mutate(Chr = gsub("NC_0499.*?([0-9]+).*", "\\1", Chr) %>%
#              str_remove("^0+") %>%
#              as.numeric())
#   
#   #Reorganize the table
#   table <- table %>%
#     select(Chr, Start, End, Strand, Length, Counts, Geneid, Source)
#   
#   return(table)
#   
# }
# 
# site_rel_gene_count <- function(rna_table) {
#   
#   tot_genes <- rna_table %>%
#   group_by(Chr) %>%
#   add_count(name = "Total_genes")
# 
#   table <- tot_genes %>%
#     mutate(Relative_count = Counts / Total_genes) %>%
#     mutate(Group = case_when(
#       Chr == 4 ~ "chr_4",
#       Chr == 9 ~ "chr_9",
#       TRUE ~ "Other"))
#   
#   return(table)
# }
# 
# chr_rel_gene_count <- function(rna_table) {
#   
#   # Total genes per chromosome
#   tot_genes <- rna_table %>%
#     group_by(Source, Chr) %>%
#     summarise(Total_genes = n()) %>%
#     ungroup
#   
#   # Sum of all RNA counts per chromosome
#   tot_counts <- rna_table %>%
#     group_by(Source, Chr) %>%
#     summarise(Total_counts = sum(Counts)) %>%
#     ungroup()
#   
#   # Sum of all lengths per chromosome
#   tot_length <- rna_table %>%
#     group_by(Source, Chr) %>%
#     summarise(Total_length = sum(Length)) %>%
#     ungroup()
#   
#   #combine tables
#   tmp_table <- full_join(tot_counts, tot_genes)
#   
#   tot_table <- full_join(tmp_table, tot_length)
#   
#   # Calculate relative count and give colors to chromosomes
#   tot_table <- tot_table %>%
#     mutate(Rel_count_per_gene = Total_counts / Total_genes) %>%
#     mutate(Rel_count_per_length = Total_counts / Total_length) %>%
#     mutate(Group = case_when(
#       Chr == 4 ~ "chr_4",
#       Chr == 9 ~ "chr_9",
#       TRUE ~ "Other"))
#   
#   return(tot_table)
#   
# }
```

## Read in data
```{r}
# cr2_11A <- read_rna_counts("./Data/RNASeq_Counts/11A_I2_S29_counts.txt", source = "cr2_11A")
# cr2_12A <- read_rna_counts("./Data/RNASeq_Counts/12A_I2_S30_counts.txt", source = "cr2_12A")
# cr2_13A <- read_rna_counts("./Data/RNASeq_Counts/13A_I1_S31_counts.txt", source = "cr2_13A")
# cr2_15A <- read_rna_counts("./Data/RNASeq_Counts/15A_I1_S32_counts.txt", source = "cr2_15A")
# cr2_15A1 <- read_rna_counts("./Data/RNASeq_Counts/15A1_I1_S40_counts.txt", source = "cr2_15A1")
# cr2_16A <- read_rna_counts("./Data/RNASeq_Counts/16A_I2_S33_counts.txt", source = "cr2_16A")
# cr2_16A1 <- read_rna_counts("./Data/RNASeq_Counts/16A1_I1_S34_counts.txt", source = "cr2_16A1")
# cr2_16A2 <- read_rna_counts("./Data/RNASeq_Counts/16A2_I2_S35_counts.txt", source = "cr2_16A2")
# cr2_17A1 <- read_rna_counts("./Data/RNASeq_Counts/17A1_I2_S36_counts.txt", source = "cr2_17A1")
# cr2_17A2 <- read_rna_counts("./Data/RNASeq_Counts/17A2_I2_S37_counts.txt", source = "cr2_17A2")
# cr2_18A <- read_rna_counts("./Data/RNASeq_Counts/18A_I2_S38_counts.txt", source = "cr2_18A")
# cr2_18A1 <- read_rna_counts("./Data/RNASeq_Counts/18A1_I1_S39_counts.txt", source = "cr2_18A1")
# 
# all_rna <- rbind(cr2_11A, cr2_12A, cr2_13A, cr2_15A, cr2_15A1, cr2_16A, cr2_16A1, cr2_16A2, cr2_17A1, cr2_17A2, cr2_18A, cr2_18A1)
```

# Searching for the HDA genes
## Try filtering for them
```{r}
# all_rna %>%
#   filter(Geneid == "LOC108981459" | Geneid == "LOC109003842") %>%
#   mutate(hda_id = "hda6")
# 
# all_rna %>%
#   filter(Geneid == "LOC108997370" | Geneid == "LOC109000511" | Geneid == "LOC109007743") %>%
#   mutate(hda_id = "hda19")
```
# Explore the data
## Check distributions of data
```{r}
# all_rna %>%
#   ggplot(aes(x = Counts)) +
#   geom_density() +
#   facet_wrap(~Source) +
#   theme_classic()
# 
# all_rna %>%
#   ggplot(aes(x = Length)) +
#   geom_density() +
#   facet_wrap(~Source, scales = "free") +
#   theme_classic()
# 
# all_rna %>%
#   group_by(Source, Chr) %>%
#   summarise(median(Counts))

```


# Relative expression per chromosome
## Generate relative expression tables
```{r}
# all_rna_rel_chr <- chr_rel_gene_count(all_rna)
# all_rna_rel_site <- site_rel_gene_count(all_rna)

```


## Plot
```{r}
# Wondering if there is an impact of number of exons in genes. Genes with more exons will havew higher counts. Maybe all of the high exon count genes are also on chromosome 7?

# all_rna_rel_chr %>%
#   ggplot(aes(x = as.factor(Chr), y = Rel_count_per_gene, fill = Group)) + 
#   geom_bar(stat = "identity") +
#   facet_wrap(~Source) +
#   scale_fill_manual(values = c("chr_4" = "red", "chr_9" = "red", "Other" = "gray")) +
#   labs(x = "Chromosome", y = "RNA Count per Gene") +
#   theme_classic() +
#   theme(legend.position = "none")
# 
# all_rna_rel_chr %>%
#   ggplot(aes(x = as.factor(Chr), y = Rel_count_per_length, fill = Group)) + 
#   geom_bar(stat = "identity") +
#   facet_wrap(~Source) +
#   scale_fill_manual(values = c("chr_4" = "red", "chr_9" = "red", "Other" = "gray")) +
#   labs(x = "Chromosome", y = "RNA Count per Exon Length") +
#   theme_classic() +
#   theme(legend.position = "none")
# 
# all_rna_rel_site %>%
#   ggplot(aes(x = as.factor(Chr), y = log(Relative_count), fill = Group)) + 
#   geom_boxplot() +
#   facet_wrap(~Source) +
#   scale_fill_manual(values = c("chr_4" = "red", "chr_9" = "red", "Other" = "gray")) +
#   labs(x = "Chromosome", y = "Relative per gene RNA Count") +
#   theme_classic() +
#   theme(legend.position = "none")
# 
# all_rna_rel_site %>%
#   ggplot(aes(x = as.factor(Chr), y = Relative_count, col = Group)) + 
#   geom_point() +
#   facet_wrap(~Source) +
#   scale_fill_manual(values = c("chr_4" = "red", "chr_9" = "red", "Other" = "gray")) +
#   labs(x = "Chromosome", y = "Relative per gene RNA Count") +
#   theme_classic() +
#   theme(legend.position = "none")
# 
# all_rna_rel_site %>%
#   ggplot(aes(x = as.factor(Chr), y = log(Counts + 1), fill = Group)) + 
#   geom_boxplot() +
#   facet_wrap(~Source) +
#   scale_fill_manual(values = c("chr_4" = "red", "chr_9" = "red", "Other" = "gray")) +
#   labs(x = "Chromosome", y = "RNA Count") +
#   theme_classic() +
#   theme(legend.position = "none")
# 
# all_rna_rel_site %>%
#   ggplot(aes(x = as.factor(Chr %in% c(4,9)), y = log(Counts), fill = Group)) + 
#   geom_boxplot() +
#   facet_wrap(~Source) +
#   scale_fill_manual(values = c("chr_4" = "red", "chr_9" = "red", "Other" = "gray")) +
#   labs(x = "Chromosome", y = "RNA Count") +
#   theme_classic() +
#   theme(legend.position = "none")
```
# Working with the Kallisto Counts
## Functions
```{r}
parse_columns <- function(table, sample){
  
  # Parse the columns
  parse.table <- table %>%
    mutate(rna = sub("rna-([^=]+).*", "\\1", target_id),
           gene = sub(".*gene=([^=]+).*", "\\1", target_id),
           name = sub(".*name=([^=]+).*", "\\1", target_id),
           seq_id = sub(".*seq_id=([^=]+).*", "\\1", target_id),
           type = sub(".*type=([^=]+)", "\\1", target_id),
           rna = gsub("gene", "", rna),
           gene = gsub("gene-|name", "", gene),
           name = gsub("seq_id", "", name)) %>%
    select(-target_id) %>%
    select(seq_id, everything()) %>%
    mutate(Source = sample)
  
  # Make the chromosome column numeric
  parse.table$seq_id <- as.numeric(gsub("NC_0499.*?([0-9]+).*", "\\1", parse.table$seq_id))

  # Add groups for later coloring
  final.table <- parse.table %>%
  mutate(Dup_Group = case_when(
    seq_id == 4 ~ "chr_4",
    seq_id == 9 ~ "chr_9",
    TRUE ~ "Other")) %>%
  mutate(Hom_Group = case_when(
    seq_id == 11 ~ "chr_11",
    seq_id == 12 ~ "chr_12",
    seq_id == 16 ~ "chr_16",
    TRUE ~ "Other"))
  return(final.table)
}
  

read_gff <- function(file){
  
  gff <- fread(file, skip = 3)
  
  colnames(gff) <- c("CHROM", "SOURCE", "ID", "START", "STOP", "SCORE", "STRAND", "FRAME", "ATTRIBUTE")
  
  # Parse the columns
  parse.table <- gff %>%
    mutate(gene = sub(".*gene-([^=]+).*", "\\1", ATTRIBUTE),
           POS = (STOP + START) / 2) %>%
    select(-ATTRIBUTE) %>%
    filter(!grepl(";", gene))
  
  gff <- parse.table[grep("^NC_0499", parse.table$CHROM), ]
  gff$CHROM <- as.numeric(gsub("NC_0499(..).1_RagTag", "\\1.", gff$CHROM))
  
  return(gff)  
}
```

## Read in data
```{r}
cr11a <- fread("./Data/Kallisto/11A_abundance.tsv")
cr12a <- fread("./Data/Kallisto/12A_abundance.tsv")
cr13a <- fread("./Data/Kallisto/13A_abundance.tsv")
cr15a <- fread("./Data/Kallisto/15A_abundance.tsv")
cr15a1 <- fread("./Data/Kallisto/15A1_abundance.tsv")
cr16a <- fread("./Data/Kallisto/16A_abundance.tsv")
cr16a1 <- fread("./Data/Kallisto/16A1_abundance.tsv")
cr16a2 <- fread("./Data/Kallisto/16A2_abundance.tsv")
cr17a1 <- fread("./Data/Kallisto/17A1_abundance.tsv")
cr17a2 <- fread("./Data/Kallisto/17A2_abundance.tsv")
cr18a <- fread("./Data/Kallisto/18A_abundance.tsv")
cr18a1 <- fread("./Data/Kallisto/18A1_abundance.tsv")
```

new_labels = c(
  "ref_chandler" = "Reference", "tree2" = "Tree",
  "cr2_hifi" = "E HiFi", "cr21_1" = "E 11", "cr21_2" = "E 10",
  "cr2_11A" = "E 14", "cr2_12A" = "E 12", "cr2_13A" = "E 13",
  "cr2_15A" = "E 8", "cr2_15A1" = "E 7", "cr2_16A" = "E 6",
  "cr2_16A1" = "E 2", "cr2_16A2" = "E 1", "cr2_17A1" = "E 4",
  "cr2_17A2" = "E 3", "cr2_18A" = "E 9", "cr2_18A1" = "E 5",
  "cr85" = "S 1985", "cr10" = "S 2010",
  "cr13" = "S 2013", "cr22" = "S 2022"
)

## Parse column information and create large data frame
```{r}
cr11a <- parse_columns(cr11a, "E 14")
cr12a <- parse_columns(cr12a, "E 12")
cr13a <- parse_columns(cr13a, "E 13")
cr15a <- parse_columns(cr15a, "E 8")
cr15a1 <- parse_columns(cr15a1, "E 7")
cr16a <- parse_columns(cr16a, "E 6")
cr16a1 <- parse_columns(cr16a1, "E 2")
cr16a2 <- parse_columns(cr16a2, "E 1")
cr17a1 <- parse_columns(cr17a1, "E 4")
cr17a2 <- parse_columns(cr17a2, "E 3")
cr18a <- parse_columns(cr18a, "E 9")
cr18a1 <- parse_columns(cr18a1, "E 5")

all_counts <- rbind(cr11a, cr12a, cr13a, cr15a, cr15a1, cr16a, cr16a1, cr16a2, cr17a1, cr17a2, cr18a, cr18a1)
```

## Calculate the counts per chromsome
```{r}
chr.table <- all_counts %>%
  filter(tpm > 1.0) %>%
  group_by(Source, seq_id) %>%
  summarise(counts_by_length = sum(est_counts)/sum(length), 
            counts_by_efflength = sum(est_counts)/sum(eff_length), 
            tpm_by_length = sum(tpm)/sum(length),
            tpm_by_efflength = sum(tpm)/sum(eff_length),
            mean_counts = mean(est_counts),
            mean_tpm = mean(tpm), 
            mean_length = mean(length)) %>%
  ungroup() %>%
  mutate(Dup_Group = case_when(
    seq_id == 4 ~ "chr_4",
    seq_id == 9 ~ "chr_9",
    TRUE ~ "Other")) %>%
  mutate(Hom_Group = case_when(
    seq_id == 11 ~ "chr_11",
    seq_id == 12 ~ "chr_12",
    seq_id == 16 ~ "chr_16",
    TRUE ~ "Other"))

```

## Plot
```{r}
chr.table %>%
  ggplot(aes(x = as.factor(seq_id), y = counts_by_length, fill = Dup_Group)) + 
  geom_bar(stat = "identity") +
  facet_wrap(~ Source, scales = "free") +
  scale_fill_manual(values = c("chr_4" = "red", "chr_9" = "red", "Other" = "gray")) +
  labs(x = "Chromosome", y = "Gene transcripts / total exon length") +
  theme_classic() +
  theme(legend.position = "none")
# ggsave("./Figures/RNA_Sequencing/kallisto_counts_over_length_per_chr.pdf", height = 8, width = 12)
# ggsave("./Figures/RNA_Sequencing/kallisto_counts_over_length_per_chr.png", height = 8, width = 12)

chr.table %>%
  ggplot(aes(x = as.factor(seq_id), y = counts_by_efflength, fill = Dup_Group)) + 
  geom_bar(stat = "identity") +
  facet_wrap(~ Source, scales = "free") +
  scale_fill_manual(values = c("chr_4" = "red", "chr_9" = "red", "Other" = "gray")) +
  labs(x = "Chromosome", y = "Gene transcripts / total effective exon length") +
  theme_classic() +
  theme(legend.position = "none")
# ggsave("./Figures/RNA_Sequencing/kallisto_counts_over_efflength_per_chr.pdf", height = 8, width = 12)
# ggsave("./Figures/RNA_Sequencing/kallisto_counts_over_efflength_per_chr.png", height = 8, width = 12)

chr.table %>%
  ggplot(aes(x = as.factor(seq_id), y = tpm_by_length, fill = Dup_Group)) + 
  geom_bar(stat = "identity") +
  facet_wrap(~ Source, scales = "free") +
  scale_fill_manual(values = c("chr_4" = "red", "chr_9" = "red", "Other" = "gray")) +
  labs(x = "Chromosome", y = "Transcript per Million / total exon length") +
  theme_classic() +
  theme(legend.position = "none")
# ggsave("./Figures/RNA_Sequencing/kallisto_tpm_over_length_per_chr.pdf", height = 8, width = 12)
# ggsave("./Figures/RNA_Sequencing/kallisto_tpm_over_length_per_chr.png", height = 8, width = 12)

chr.table %>%
  ggplot(aes(x = as.factor(seq_id), y = tpm_by_efflength, fill = Dup_Group)) + 
  geom_bar(stat = "identity") +
  facet_wrap(~ Source, scales = "free") +
  scale_fill_manual(values = c("chr_4" = "red", "chr_9" = "red", "Other" = "gray")) +
  labs(x = "Chromosome", y = "Transcript per Million / total  effective exon length") +
  theme_classic() +
  theme(legend.position = "none")
# ggsave("./Figures/RNA_Sequencing/kallisto_tpm_over_efflength_per_chr.pdf", height = 8, width = 12)
# ggsave("./Figures/RNA_Sequencing/kallisto_tpm_over_efflength_per_chr.png", height = 8, width = 12)

chr.table %>%
  ggplot(aes(x = as.factor(seq_id), y = mean_counts, fill = Dup_Group)) + 
  geom_bar(stat = "identity") +
  facet_wrap(~ Source, scales = "free") +
  scale_fill_manual(values = c("chr_4" = "red", "chr_9" = "red", "Other" = "gray")) +
  labs(x = "Chromosome", y = "Mean Counts") +
  theme_classic() +
  theme(legend.position = "none")
# ggsave("./Figures/RNA_Sequencing/kallisto_mean_counts_per_chr.pdf", height = 8, width = 12)
# ggsave("./Figures/RNA_Sequencing/kallisto_mean_counts_per_chr.png", height = 8, width = 12)

chr.table %>%
  ggplot(aes(x = as.factor(seq_id), y = mean_tpm, fill = Dup_Group)) + 
  geom_bar(stat = "identity") +
  facet_wrap(~ Source, scales = "free") +
  scale_fill_manual(values = c("chr_4" = "red", "chr_9" = "red", "Other" = "gray")) +
  labs(x = "Chromosome", y = "Mean TPM") +
  theme_classic() +
  theme(legend.position = "none")
# ggsave("./Figures/RNA_Sequencing/kallisto_mean_tpm_per_chr.pdf", height = 8, width = 12)
# ggsave("./Figures/RNA_Sequencing/kallisto_mean_tpm_per_chr.png", height = 8, width = 12)

chr.table %>%
  ggplot(aes(x = as.factor(seq_id), y = mean_length, fill = Dup_Group)) + 
  geom_bar(stat = "identity") +
  facet_wrap(~ Source, scales = "free") +
  scale_fill_manual(values = c("chr_4" = "red", "chr_9" = "red", "Other" = "gray")) +
  labs(x = "Chromosome", y = "Mean Length") +
  theme_classic() +
  theme(legend.position = "none")
# ggsave("./Figures/RNA_Sequencing/kallisto_mean_length_per_chr.pdf", height = 8, width = 12)
# ggsave("./Figures/RNA_Sequencing/kallisto_mean_length_per_chr.png", height = 8, width = 12)


all_counts %>%
  filter(tpm > 1.0) %>%
  ggplot(aes(x = as.factor(seq_id), y = log(est_counts), fill = Dup_Group)) + 
  geom_boxplot() +
  facet_wrap(~ Source, scales = "free") +
  scale_fill_manual(values = c("chr_4" = "red", "chr_9" = "red", "Other" = "gray")) +
  labs(x = "Chromosome", y = "Estimated Counts") +
  theme_classic() +
  theme(legend.position = "none")
# ggsave("./Figures/RNA_Sequencing/kallisto_boxplot_counts_per_chr.pdf", height = 8, width = 12)
# ggsave("./Figures/RNA_Sequencing/kallisto_boxplot_counts_per_chr.png", height = 8, width = 12)

all_counts %>%
  filter(tpm > 1.0) %>%
  ggplot(aes(x = as.factor(seq_id), y = log(est_counts/length), fill = Dup_Group)) + 
  geom_boxplot() +
  facet_wrap(~Source, scales = "free") +
  scale_fill_manual(values = c("chr_4" = "red", "chr_9" = "red", "Other" = "gray")) +
  labs(x = "Chromosome", y = "Count / total exon length (log)") +
  theme_classic() +
  theme(legend.position = "none")
# ggsave("./Figures/RNA_Sequencing/kallisto_boxplot_counts_over_length_per_chr.pdf", height = 8, width = 12)
# ggsave("./Figures/RNA_Sequencing/kallisto_boxplot_counts_over_length_per_chr.png", height = 8, width = 12)

all_counts %>%
  filter(tpm > 1.0) %>%
  ggplot(aes(x = as.factor(seq_id), y = log(tpm), fill = Dup_Group)) + 
  geom_boxplot() +
  facet_wrap(~Source, scales = "free") +
  scale_fill_manual(values = c("chr_4" = "red", "chr_9" = "red", "Other" = "gray")) +
  labs(x = "Chromosome", y = "TPM") +
  theme_classic() +
  theme(legend.position = "none")
# ggsave("./Figures/RNA_Sequencing/kallisto_boxplot_tpm_per_chr.pdf", height = 8, width = 12)
# ggsave("./Figures/RNA_Sequencing/kallisto_boxplot_tpm_per_chr.png", height = 8, width = 12)

all_counts %>%
  filter(tpm > 1.0) %>%
  ggplot(aes(x = as.factor(seq_id %in% c(4,9)), y = log(tpm), fill = Dup_Group)) + 
  geom_boxplot() +
  facet_wrap(~Source, scales = "free") +
  scale_fill_manual(values = c("chr_4" = "red", "chr_9" = "red", "Other" = "gray")) +
  labs(x = "Chromosome", y = "Log TPM") +
  theme_classic() +
  theme(legend.position = "none")
```

# Plot to compare all chr to 4 and 9
```{r}
# Set levels
levels <- c("E 1", "E 2", "E 3", "E 4", "E 5", "E 6",
            "E 7", "E 8", "E 9", 
            "E 12", "E 13", "E 14")
grp_levels <- c("Other", "chr_4", "chr_9")
all_counts$Source <- factor(all_counts$Source, levels = levels)
all_counts$Dup_Group <- factor(all_counts$Dup_Group, levels = grp_levels)

# Specifying t.test comparisons
comparisons <- list(c("chr_4", "Other"), c("chr_9", "Other"))


all_counts %>%
  filter(tpm > 1.0) %>%
  ggplot(aes(x = Dup_Group, y = log(tpm), fill = Dup_Group)) + 
  geom_boxplot() +
  facet_wrap(~ Source, scales = "free") +
  scale_fill_manual(values = c("chr_4" = "firebrick", "chr_9" = "firebrick", "Other" = "gray90")) +
  scale_x_discrete(labels = c("Other" = "Others", "chr_4" = "4", "chr_9" = "9")) +
  labs(x = "Chromosome", y = "TPM") +
  stat_compare_means(method = "t.test", comparisons = comparisons, label = "p.signif") +
  theme_classic(base_size = 6) +
  theme(legend.position = "none",
        axis.title = element_text(face = "bold"),
        strip.background = element_blank(),
        strip.text = element_text(face = "bold"))

all_counts %>%
  filter(tpm > 1.0) %>%
  ggplot(aes(x = Dup_Group, y = log(tpm), fill = Dup_Group)) + 
  geom_boxplot() +
  facet_wrap(~ Source, scales = "free") +
  scale_fill_manual(values = c("chr_4" = "firebrick", "chr_9" = "firebrick", "Other" = "gray90")) +
  scale_x_discrete(labels = c("Other" = "Others", "chr_4" = "4", "chr_9" = "9")) +
  labs(x = "Chromosome", y = "TPM") +
  stat_compare_means(comparisons = comparisons) +
  theme_classic(base_size = 6) +
  theme(legend.position = "none",
        axis.title = element_text(face = "bold"),
        strip.background = element_blank(),
        strip.text = element_text(face = "bold"))

all_counts %>% 
  filter(tpm > 1.0) %>%
  ggplot(aes(x = Dup_Group, y = log(tpm), fill = Dup_Group)) + 
  geom_boxplot(outlier.shape = NA) +
  facet_grid(~ Source, scales = "free") +
  scale_fill_manual(values = c("chr_4" = "firebrick", "chr_9" = "firebrick", "Other" = "gray90")) +
  scale_x_discrete(labels = c("Other" = "Others", "chr_4" = "4", "chr_9" = "9")) +
  coord_cartesian(ylim = c(0,9)) +
  labs(x = "Chromosome", y = "Natural Log TPM") +
  stat_compare_means(method = "t.test", comparisons = comparisons, label = "p.signif", bracket.size = .3, size = 2, label.y = c(7,8)) +
  theme_classic(base_size = 6) +
  theme(legend.position = "none",
        #axis.title = element_text(face = "bold"),
        strip.background = element_blank(),
        strip.text = element_text(face = "bold"),
        plot.margin = margin(t = 0, r = 0, b = 0, l = 0, unit = "cm")
        )
```

## Plot comapring expression in chromosomes with log
```{r}
all_counts %>%
  filter(tpm > 1.0) %>%
  mutate(
    fill_color = case_when(
      Dup_Group %in% c("chr_4") & Source %in% c("E 5", "E 6") ~ "gray80",
      Dup_Group %in% c("chr_4", "chr_9") ~ "orangered2",
      TRUE ~ "gray80"
    )
  ) %>%
  ggplot(aes(x = Dup_Group, y = log(tpm), fill = fill_color, group = Dup_Group)) +  # Explicit grouping by Dup_Group
  geom_boxplot(outlier.shape = NA) +
  facet_grid(~ Source, scales = "free") +
  scale_fill_identity() +
  scale_x_discrete(labels = c("Other" = "All", "chr_4" = "4", "chr_9" = "9")) +
  coord_cartesian(ylim = c(0, 9)) +
  labs(x = "Chromosome", y = "Natural Log TPM") +
  stat_compare_means(method = "t.test", comparisons = comparisons, label = "p.signif", bracket.size = 0.3, size = 2, label.y = c(7, 8)) +
  theme_classic(base_size = 6) +
  theme(
    legend.position = "none",
    strip.background = element_blank(),
    #strip.text = element_text(face = "bold"),
    plot.margin = margin(t = -0, r = -0, b = -0, l = -0, unit = "cm")
  )
ggsave("./Figures/RNA_Sequencing/compare_chr_49_to_all_ttest_log.pdf", height = 1.8, width = 5.5)
ggsave("./Figures/RNA_Sequencing/compare_chr_49_to_all_ttest_log.png", height = 1.8, width = 5.5)
```

## Plot comapring expression in chromosomes
```{r}
all_counts %>%
  filter(tpm > 1.0) %>%
  mutate(
    fill_color = case_when(
      Dup_Group %in% c("chr_4") & Source %in% c("E 5", "E 6") ~ "gray80",
      Dup_Group %in% c("chr_4", "chr_9") ~ "orangered2",
      TRUE ~ "gray80"
    )
  ) %>%
  ggplot(aes(x = Dup_Group, y = tpm, fill = fill_color, group = Dup_Group)) +  # Explicit grouping by Dup_Group
  geom_boxplot(outlier.shape = NA) +
  facet_grid(~ Source, scales = "free") +
  scale_fill_identity() +
  scale_x_discrete(labels = c("Other" = "All", "chr_4" = "4", "chr_9" = "9")) +
  coord_cartesian(ylim = c(0, 125)) +
  labs(x = "Chromosome", y = "TPM") +
  stat_compare_means(method = "t.test", comparisons = comparisons, label = "p.signif", bracket.size = 0.3, size = 2) +
  theme_classic(base_size = 6) +
  theme(
    legend.position = "none",
    strip.background = element_blank(),
    #strip.text = element_text(face = "bold"),
    plot.margin = margin(t = -0, r = -0, b = -0, l = -0, unit = "cm")
  )
ggsave("./Figures/RNA_Sequencing/compare_chr_49_to_all_ttest.pdf", height = 1.8, width = 5.5)
ggsave("./Figures/RNA_Sequencing/compare_chr_49_to_all_ttest.png", height = 1.8, width = 5.5)
```




## Look at homozygous regions of the genome
```{r}
comparisons <- list(c("chr_11", "Other"), c("chr_12", "Other"), c("chr_16", "Other"))

all_counts %>%
  filter(!seq_id %in% c(4,9)) %>%
  filter(tpm >= 1.0) %>%
  ggplot(aes(x = Hom_Group, y = log(tpm), fill = Hom_Group)) + 
  geom_boxplot(outlier.size = .25) +
  facet_grid(~ Source, scales = "free") +
  scale_fill_manual(values = c("chr_11" = "firebrick", "chr_12" = "dodgerblue",  "chr_16" = "firebrick", "Other" = "gray90")) +
  scale_x_discrete(labels = c("Other" = "Others", "chr_11" = "11", "chr_12" = "12", "chr_16" = "16")) +
  labs(x = "Chromosome", y = "Natural Log TPM") +
  stat_compare_means(method = "t.test", comparisons = comparisons, label = "p.signif", bracket.size = .3, size = 2) +
  theme_classic(base_size = 6) +
  theme(legend.position = "none",
        axis.title = element_text(face = "bold"),
        strip.background = element_blank(),
        strip.text = element_text(face = "bold"))

#ggsave("./Figures/RNA_Sequencing/compare_chr_111216_to_all_ttest.pdf", height = 2.2, width = 5.8)
#ggsave("./Figures/RNA_Sequencing/compare_chr_111216_to_all_ttest.png", height = 2.2, width = 5.8)
```


## Look for regions of lower expression along the Chromosomes
# Get gene position
```{r}
gff <- read_gff("./Data/GFFs/Chromosome_Only_Liftoff/ref_chandler_primary_default_scaffold_chr_only_liftoff_a99s99.gff")

gene_pos <- gff %>%
  select(CHROM, POS, gene)

gene_pos <- gene_pos %>%
  mutate(gene = sub("_.*", "", gene))

all_counts <- all_counts %>%
  dplyr::rename(CHROM = seq_id)

all_counts <- left_join(all_counts, gene_pos)


all_counts %>%
  filter(tpm > 1.0) %>%
  filter(Source == "E 3") %>%
  ggplot(aes(x = POS, y = tpm)) + 
  geom_point() +
  facet_wrap(~ CHROM, scales = "free") +
  labs(x = "Position", y = "Log TPM") +
  theme_classic(base_size = 6) +
  theme()

str(all_counts)
```
# Creating windows
```{r}
trial <- all_counts %>%
  filter(tpm > 1) %>%
  filter(!is.na(POS))

# Define the window size
window_size <- 1e6

# Create a new column to define windows based on the POS column
trial[, window := floor((POS - 1) / window_size) + 1, by = CHROM]

# Calculate the mean TPM for each window for every chromosome, keeping the Source information
med_tpm_windows <- trial[, .(med_tpm = median(tpm, na.rm = TRUE)), by = .(CHROM, window, Source)]

# Add a group column
med_tpm_windows <- med_tpm_windows %>%
    mutate(Dup_Group = case_when(
      CHROM == 4 ~ "chr_4",
      CHROM == 9 ~ "chr_9",
      TRUE ~ "Other"))

# Plotting
med_tpm_windows %>%
  ggplot(aes(x = window, y = log(med_tpm), color = Source)) + 
  geom_line() +
  facet_wrap(~ CHROM, scales = "free") +
  labs(x = "Position", y = "Log TPM") +
  scale_color_manual(values = c("E 5" = "firebrick", "E 6" = "gold")) +
  theme_classic(base_size = 6) +
  theme()

med_tpm_windows %>%
  ggplot(aes(x = window, y = log(med_tpm), color = Source)) + 
  geom_point() +
  facet_wrap(~ CHROM, scales = "free") +
  labs(x = "Position", y = "Log TPM") +
  scale_color_manual(values = c("E 5" = "firebrick", "E 6" = "gold")) +
  theme_classic(base_size = 6) +
  theme()
```

# Plot a geom_tile representation of TPM in windows
## Normalization data
```{r}
library(ggenomics)
counts.genes <- all_counts %>%
  filter(tpm > 1, !is.na(POS))

colnames(counts.genes)[colnames(counts.genes) == "Source"] <- "SOURCE"

counts.genes <- sliding_window_table(counts.genes, window_size = 1e8, slide_size = 1e8)

counts.genes <- counts.genes %>%
  group_by(SOURCE, CHROM, POS_WINDOW) %>%
  mutate(med_tpm_window = median(tpm))

counts.genes <- counts.genes %>%
  group_by(SOURCE) %>%
  mutate(mean_tpm_sample = mean(med_tpm_window))

counts.genes <- counts.genes %>%
  mutate(norm_tpm = med_tpm_window/mean_tpm_sample)
```

## Plot Counts genes
```{r}
counts.genes %>%
  mutate(SOURCE = forcats::fct_rev(SOURCE)) %>% # reverse levels
  ggplot(aes(x = POS_WINDOW, y = SOURCE, fill = norm_tpm)) +
  geom_tile() +
  scale_fill_gradientn(colours = c("grey97", "grey95","grey95","grey93","orangered", "orangered2","orangered3","orangered4")) +
  #  viridis::scale_fill_viridis(option = "turbo") + 
  facet_grid(SOURCE~CHROM, scales = "free", space = "free") +
  theme_classic(base_size = 6) +
  theme(strip.background = element_blank(),
        strip.text.x = element_text(margin = margin(t = 1, b = 1, unit = "pt")),
        strip.text.y = element_blank(),  
        legend.position = "none",
        axis.title.y = element_blank(),
        axis.ticks.y = element_line(linewidth = 0.5, linetype = "13"),
        axis.ticks.length.y = unit(0.5, "cm"),
        axis.text.y = element_blank(),
        axis.title.x = element_blank(),
        #axis.line.x = element_blank(),
        axis.text.x = element_blank(),
        axis.ticks.x = element_blank(),
        panel.border = element_rect(fill = NA),
        panel.spacing.y = unit(0.05, "cm"),
        panel.spacing.x = unit(0.05, "cm"),
        plot.margin = margin(t = 0, r = 0.1, b = 0, l = -0, unit = "cm")
        )
ggsave("./Figures/RNA_Sequencing/expression_heatmap_all_dups.pdf", height = 1.5, width = 2.3)
ggsave("./Figures/RNA_Sequencing/expression_heatmap_all_dups.png", height = 1.5, width = 2.3)
```

## Try windows of gene numbers
```{r}
setDT(counts.genes)
setorder(counts.genes, CHROM, POS)

counts.genes[, GENEWINDOW := (1:.N - 1) %/% 5000 + 1, by = CHROM]

counts.genes %>%
  group_by(CHROM) %>%
  summarise(num_windows = length(unique(GENEWINDOW)))

counts.genes <- counts.genes %>%
  group_by(SOURCE, CHROM, GENEWINDOW) %>%
  mutate(med_tpm_genewindow = median(tpm))

counts.genes <- counts.genes %>%
  group_by(SOURCE) %>%
  mutate(genemean_tpm_sample = mean(med_tpm_genewindow))

counts.genes <- counts.genes %>%
  mutate(genenorm_tpm = med_tpm_genewindow/genemean_tpm_sample)
```

## Plot gene windows
```{r}
counts.genes %>%
  mutate(SOURCE = forcats::fct_rev(SOURCE)) %>% # reverse levels
  ggplot(aes(x = GENEWINDOW, y = SOURCE, fill = genenorm_tpm)) +
  geom_tile() +
  scale_fill_gradientn(colours = c("royalblue", "white", "lightyellow", "lightyellow", "orangered1",  "orangered1")) +
  facet_grid(SOURCE~CHROM, scales = "free", space = "free") +
  theme_classic(base_size = 6) +
  theme(strip.background = element_blank(),
        strip.text.x = element_text(margin = margin(t = 1, b = 1, unit = "pt")),
        strip.text.y = element_blank(),  
        legend.position = "none",
        axis.title.y = element_blank(),
        axis.ticks.y = element_line(linewidth = 0.5, linetype = "13"),
        axis.ticks.length.y = unit(0.5, "cm"),
        axis.text.y = element_blank(),
        axis.title.x = element_blank(),
        #axis.line.x = element_blank(),
        axis.text.x = element_blank(),
        axis.ticks.x = element_blank(),
        panel.border = element_rect(fill = NA),
        panel.spacing.y = unit(0.05, "cm"),
        panel.spacing.x = unit(0.05, "cm"),
        plot.margin = margin(t = 0, r = 0.1, b = 0, l = -0, unit = "cm")
        )
```

## Try even segments
```{r}
setDT(counts.genes)

counts.genes[, SEGMENT := cut(1:.N, breaks = 5, labels = FALSE), by = c("SOURCE", "CHROM")]

#counts.genes[, SEGMENT := 1, by = c("SOURCE", "CHROM")]
# whole chromosome isn't enough to tell the expression difference

counts.genes <- counts.genes %>%
  group_by(SOURCE, CHROM, SEGMENT) %>%
  mutate(med_tpm_segment = median(tpm))

counts.genes <- counts.genes %>%
  group_by(SOURCE, CHROM, SEGMENT) %>%
  mutate(mean_tpm_segment = mean(tpm))

counts.genes <- counts.genes %>%
  group_by(SOURCE) %>%
  mutate(segmean_tpm_sample = mean(med_tpm_segment))

counts.genes <- counts.genes %>%
  group_by(SOURCE) %>%
  mutate(segmean_meantpm_sample = mean(mean_tpm_segment))

counts.genes <- counts.genes %>%
  mutate(segnorm_tpm = med_tpm_segment/segmean_tpm_sample)

counts.genes <- counts.genes %>%
  mutate(segnorm_meantpm = mean_tpm_segment/segmean_meantpm_sample)

range(counts.genes$segnorm_tpm)
```

## plot segments
```{r}
counts.genes %>%
  mutate(SOURCE = forcats::fct_rev(SOURCE)) %>% # reverse levels
  ggplot(aes(x = SEGMENT, y = SOURCE, fill = log(segnorm_tpm))) +
  geom_tile() +
  viridis::scale_fill_viridis(option = "turbo") +
   # scale_fill_gradientn(
   # colours = c("royalblue1", "royalblue", "white","white","white", "white","white","white","white","white","white","white","orangered", "orangered","orangered","orangered","orangered"),
   # values = scales::rescale(c(0.62, 0.65, 0.67, 0.68, 0.69, 0.7, 0.71, 0.72, 0.73, 0.74, .75, 1.2, 1.25,1.3, 1.5, 1.6, 1.7, 1.8)) # Specify explicit mapping
  #limits = c(0, 1.6), # Ensure your range covers all values
  #oob = scales::squish # Handle out-of-bounds values
  # #scale_fill_gradientn() 
  #   scale_fill_gradient2(
  #   low = "blue", 
  #   mid = "white", 
  #   high = "brown", 
  #   midpoint = 1
   #) +
  facet_grid(SOURCE~CHROM, scales = "free", space = "free") +
  theme_classic(base_size = 6) +
  theme(strip.background = element_blank(),
        strip.text.x = element_text(margin = margin(t = 1, b = 1, unit = "pt")),
        strip.text.y = element_blank(),  
        legend.position = "right",
        axis.title.y = element_blank(),
        axis.ticks.y = element_line(linewidth = 0.5, linetype = "13"),
        axis.ticks.length.y = unit(0.5, "cm"),
        axis.text.y = element_blank(),
        axis.title.x = element_blank(),
        axis.line.x = element_blank(),
        axis.line.y = element_blank(),
        axis.text.x = element_blank(),
        axis.ticks.x = element_blank(),
        panel.border = element_rect(fill = NA),
        panel.spacing.y = unit(0.05, "cm"),
        panel.spacing.x = unit(0.05, "cm"),
        plot.margin = margin(t = 0, r = 0.1, b = 0, l = -0, unit = "cm")
        )
ggsave("./Figures/RNA_Sequencing/expression_heatmap_all_5seg.pdf", height = 1.5, width = 2.3)
ggsave("./Figures/RNA_Sequencing/expression_heatmap_all_5seg.png", height = 1.5, width = 2.3)
```

# Try comparing mean of each window to every other window in the sample
```{r}
# Function to calculate and compare medians
test <-  counts.genes %>%
  # Calculate SEGMENT median
  group_by(SOURCE, CHROM, SEGMENT) %>%
  summarise(SEGMENT_MEDIAN = median(tpm)) %>%
  # Calculate SOURCE median excluding CHROMs 4, 9, 12
  left_join(
    counts.genes %>%
      #filter(!CHROM %in% c(4, 9, 11, 12, 16)) %>%
      group_by(SOURCE) %>%
      summarise(SOURCE_MEDIAN = median(tpm)),
    by = "SOURCE"
  ) %>%
  # Divide SEGMENT_MEDIAN by SOURCE_MEDIAN
  mutate(
    RATIO = SEGMENT_MEDIAN / SOURCE_MEDIAN,
    FLAG = ifelse(RATIO > 1, "Higher", ifelse(RATIO < 1, "Lower", "Equal")),
    LOGRAT = log(RATIO)
  )

test
```

## Try plotting
This is the closest I've gotten
```{r}
range(test$LOGRAT)
test %>%
  ggplot(aes(x = SEGMENT, y = SOURCE, fill = LOGRAT)) +
  geom_tile() +
  #viridis::scale_fill_viridis(option = "turbo") +
  scale_fill_gradientn(colours = c("royalblue1", "royalblue2", "royalblue", "white","gray98", "gray96", "gray94", "gray92", "grey90","orangered", "orangered2", "orangered3"),
                       #values = scales::rescale(c(-0.86, -0.8, -.6, -.4, 0, .15, .25, .35, .45, .6))
                       )+
  facet_grid(SOURCE~CHROM, scales = "free", space = "free") +
  theme_classic(base_size = 6)
```

# Calculate if there is a difference between expression of each gene and everyother in the individual
## 
```{r}
library(dplyr)
library(purrr)
library(broom)

# Define a function for pairwise comparison
compare_tpm <- function(data) {
  data %>%
    mutate(pvalue = map_dbl(tpm, ~ {
      # Exclude the current tpm value
      others <- tpm[tpm != .x]
      if (length(others) > 1) {
        # Perform Wilcoxon test
        wilcox.test(c(.x), others, exact = FALSE)$p.value
      } else {
        NA  # Not enough data for testing
      }
    }))
}
```

## Test functino
```{r}
# Group by SOURCE and apply the function
genes_pvals <- counts.genes %>%
  ungroup() %>%
  #slice_sample(n = 1e4) %>%
  group_by(SOURCE) %>%
  group_modify(~ compare_tpm(.x)) %>%
  ungroup()

genes_pvals <- genes_pvals %>%
  dplyr::select(-c(Dup_Group, Hom_Group, POS, POS_WINDOW, WINDOW_START, WINDOW_END, med_tpm_window, mean_tpm_sample, norm_tpm, SEGMENT, med_tpm_segment, segmean_tpm_sample, segnorm_tpm, mean_tpm_segment, segmean_meantpm_sample, segnorm_meantpm))

#fwrite(genes_pvals, "./Data/RNASeq_Counts/all_sample_kallisto_counts_w_pvalue.csv")

range(genes_pvals$pvalue)
```

## Compare Chromosomes
```{r}
compare_chromosomes_by_source <- function(data) {
  data %>%
    group_by(SOURCE, CHROM) %>%
    summarize(
      # Extract the TPM values for the current chromosome
      current_chr_tpm = list(tpm[CHROM == first(CHROM)]),
      # Extract the TPM values for all other chromosomes
      other_chr_tpm = list(tpm[CHROM != first(CHROM)]),
      # Perform the Wilcoxon test if there are sufficient points
      pvalue = {
        current_chr = unlist(current_chr_tpm)
        other_chr = unlist(other_chr_tpm)
        
        if (length(current_chr) > 1 && length(other_chr) > 1) {
          wilcox.test(current_chr, other_chr, exact = FALSE)$p.value
        } else {
          NA  # Not enough data for testing
        }
      },
      .groups = "drop"  # Remove grouping from the output
    )
}


# Apply the function
genes_pvals_chr <- counts.genes %>%
  ungroup() %>%
  group_by("SOURCE") %>%
  group_modify(~ compare_chromosomes_by_source(.x)) %>%
  ungroup()
```

## Another way to test
```{r}
library(dplyr)

# Function to calculate p-values for each CHROM
calculate_pvalues <- function(df) {
  df %>%
    group_by(SOURCE, CHROM) %>%
    summarise(
      pvalue = {
        current_chrom <- tpm
        other_chrom <- df %>% filter(SOURCE == first(as.character(SOURCE)) & CHROM != first(CHROM)) %>% pull(tpm)
        wilcox.test(current_chrom, other_chrom)$p.value
      },
      .groups = "drop"
    )
}

# Apply the function and join the p-values back to the original data
result <- counts.genes %>%
  group_split(SOURCE) %>%
  purrr::map_dfr(calculate_pvalues) %>%
  right_join(data, by = c("SOURCE", "CHROM"))

# View the result
head(result)


```

```{r}
library(dplyr)
library(purrr)

# Function to calculate p-values for each CHROM
calculate_pvalues <- function(df) {
  df %>%
    group_by(CHROM) %>%
    summarise(
      pvalue = {
        current_chrom <- tpm
        other_chrom <- df %>% filter(CHROM != cur_group_id()$CHROM) %>% pull(tpm)
        wilcox.test(current_chrom, other_chrom)$p.value
      },
      .groups = "drop"
    )
}

# Group by SOURCE and calculate p-values for each CHROM
result <- counts.genes %>%
  group_split(SOURCE, keep = TRUE) %>%
  map_dfr(~ calculate_pvalues(.x) %>%
            mutate(SOURCE = unique(.x$SOURCE))) %>%
  right_join(counts.genes, by = c("SOURCE", "CHROM"))

# View the result
head(result)

```

