---
title: "rna_seq_expression"
author: "Matthew Davis"
date: "2024-07-17"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

# Setup
## Libraries
```{r, message=FALSE}
library(data.table)
library(tidyverse)
library(ggpubr)
```

# STAR data
## Functions
```{r}
# read_rna_counts <- function(path, source){
#   
#   # Read in data
#   table <- fread(path)
# 
#   # Manipulate table
#   colnames(table)[7] <- "Counts" # Change column 7 to counts
#   table$Chr <- sub(";.*", "",table$Chr) # Keep entry for first exon
#   table$Strand <- sub(";.*", "",table$Strand) # Keep first entry for strand
#   table$Start <- sub(";.*", "",table$Start) # Keep first entry for start
#   table$End <- sub(".*;([^;]+)$", "\\1",table$End) # keep last entry for end. This gives total gene position. Lengths aren't quite right though.
#   
#   # Add Source column
#   table <- table %>%
#     mutate(Source = source)
#   
#   # Rename chromosome names
#   table <- table %>%
#     mutate(Chr = gsub("NC_0499.*?([0-9]+).*", "\\1", Chr) %>%
#              str_remove("^0+") %>%
#              as.numeric())
#   
#   #Reorganize the table
#   table <- table %>%
#     select(Chr, Start, End, Strand, Length, Counts, Geneid, Source)
#   
#   return(table)
#   
# }
# 
# site_rel_gene_count <- function(rna_table) {
#   
#   tot_genes <- rna_table %>%
#   group_by(Chr) %>%
#   add_count(name = "Total_genes")
# 
#   table <- tot_genes %>%
#     mutate(Relative_count = Counts / Total_genes) %>%
#     mutate(Group = case_when(
#       Chr == 4 ~ "chr_4",
#       Chr == 9 ~ "chr_9",
#       TRUE ~ "Other"))
#   
#   return(table)
# }
# 
# chr_rel_gene_count <- function(rna_table) {
#   
#   # Total genes per chromosome
#   tot_genes <- rna_table %>%
#     group_by(Source, Chr) %>%
#     summarise(Total_genes = n()) %>%
#     ungroup
#   
#   # Sum of all RNA counts per chromosome
#   tot_counts <- rna_table %>%
#     group_by(Source, Chr) %>%
#     summarise(Total_counts = sum(Counts)) %>%
#     ungroup()
#   
#   # Sum of all lengths per chromosome
#   tot_length <- rna_table %>%
#     group_by(Source, Chr) %>%
#     summarise(Total_length = sum(Length)) %>%
#     ungroup()
#   
#   #combine tables
#   tmp_table <- full_join(tot_counts, tot_genes)
#   
#   tot_table <- full_join(tmp_table, tot_length)
#   
#   # Calculate relative count and give colors to chromosomes
#   tot_table <- tot_table %>%
#     mutate(Rel_count_per_gene = Total_counts / Total_genes) %>%
#     mutate(Rel_count_per_length = Total_counts / Total_length) %>%
#     mutate(Group = case_when(
#       Chr == 4 ~ "chr_4",
#       Chr == 9 ~ "chr_9",
#       TRUE ~ "Other"))
#   
#   return(tot_table)
#   
# }
```

## Read in data
```{r}
# cr2_11A <- read_rna_counts("./Data/RNASeq_Counts/11A_I2_S29_counts.txt", source = "cr2_11A")
# cr2_12A <- read_rna_counts("./Data/RNASeq_Counts/12A_I2_S30_counts.txt", source = "cr2_12A")
# cr2_13A <- read_rna_counts("./Data/RNASeq_Counts/13A_I1_S31_counts.txt", source = "cr2_13A")
# cr2_15A <- read_rna_counts("./Data/RNASeq_Counts/15A_I1_S32_counts.txt", source = "cr2_15A")
# cr2_15A1 <- read_rna_counts("./Data/RNASeq_Counts/15A1_I1_S40_counts.txt", source = "cr2_15A1")
# cr2_16A <- read_rna_counts("./Data/RNASeq_Counts/16A_I2_S33_counts.txt", source = "cr2_16A")
# cr2_16A1 <- read_rna_counts("./Data/RNASeq_Counts/16A1_I1_S34_counts.txt", source = "cr2_16A1")
# cr2_16A2 <- read_rna_counts("./Data/RNASeq_Counts/16A2_I2_S35_counts.txt", source = "cr2_16A2")
# cr2_17A1 <- read_rna_counts("./Data/RNASeq_Counts/17A1_I2_S36_counts.txt", source = "cr2_17A1")
# cr2_17A2 <- read_rna_counts("./Data/RNASeq_Counts/17A2_I2_S37_counts.txt", source = "cr2_17A2")
# cr2_18A <- read_rna_counts("./Data/RNASeq_Counts/18A_I2_S38_counts.txt", source = "cr2_18A")
# cr2_18A1 <- read_rna_counts("./Data/RNASeq_Counts/18A1_I1_S39_counts.txt", source = "cr2_18A1")
# 
# all_rna <- rbind(cr2_11A, cr2_12A, cr2_13A, cr2_15A, cr2_15A1, cr2_16A, cr2_16A1, cr2_16A2, cr2_17A1, cr2_17A2, cr2_18A, cr2_18A1)
```

# Searching for the HDA genes
## Try filtering for them
```{r}
# all_rna %>%
#   filter(Geneid == "LOC108981459" | Geneid == "LOC109003842") %>%
#   mutate(hda_id = "hda6")
# 
# all_rna %>%
#   filter(Geneid == "LOC108997370" | Geneid == "LOC109000511" | Geneid == "LOC109007743") %>%
#   mutate(hda_id = "hda19")
```
# Explore the data
## Check distributions of data
```{r}
# all_rna %>%
#   ggplot(aes(x = Counts)) +
#   geom_density() +
#   facet_wrap(~Source) +
#   theme_classic()
# 
# all_rna %>%
#   ggplot(aes(x = Length)) +
#   geom_density() +
#   facet_wrap(~Source, scales = "free") +
#   theme_classic()
# 
# all_rna %>%
#   group_by(Source, Chr) %>%
#   summarise(median(Counts))

```


# Relative expression per chromosome
## Generate relative expression tables
```{r}
# all_rna_rel_chr <- chr_rel_gene_count(all_rna)
# all_rna_rel_site <- site_rel_gene_count(all_rna)

```


## Plot
```{r}
# Wondering if there is an impact of number of exons in genes. Genes with more exons will havew higher counts. Maybe all of the high exon count genes are also on chromosome 7?

# all_rna_rel_chr %>%
#   ggplot(aes(x = as.factor(Chr), y = Rel_count_per_gene, fill = Group)) + 
#   geom_bar(stat = "identity") +
#   facet_wrap(~Source) +
#   scale_fill_manual(values = c("chr_4" = "red", "chr_9" = "red", "Other" = "gray")) +
#   labs(x = "Chromosome", y = "RNA Count per Gene") +
#   theme_classic() +
#   theme(legend.position = "none")
# 
# all_rna_rel_chr %>%
#   ggplot(aes(x = as.factor(Chr), y = Rel_count_per_length, fill = Group)) + 
#   geom_bar(stat = "identity") +
#   facet_wrap(~Source) +
#   scale_fill_manual(values = c("chr_4" = "red", "chr_9" = "red", "Other" = "gray")) +
#   labs(x = "Chromosome", y = "RNA Count per Exon Length") +
#   theme_classic() +
#   theme(legend.position = "none")
# 
# all_rna_rel_site %>%
#   ggplot(aes(x = as.factor(Chr), y = log(Relative_count), fill = Group)) + 
#   geom_boxplot() +
#   facet_wrap(~Source) +
#   scale_fill_manual(values = c("chr_4" = "red", "chr_9" = "red", "Other" = "gray")) +
#   labs(x = "Chromosome", y = "Relative per gene RNA Count") +
#   theme_classic() +
#   theme(legend.position = "none")
# 
# all_rna_rel_site %>%
#   ggplot(aes(x = as.factor(Chr), y = Relative_count, col = Group)) + 
#   geom_point() +
#   facet_wrap(~Source) +
#   scale_fill_manual(values = c("chr_4" = "red", "chr_9" = "red", "Other" = "gray")) +
#   labs(x = "Chromosome", y = "Relative per gene RNA Count") +
#   theme_classic() +
#   theme(legend.position = "none")
# 
# all_rna_rel_site %>%
#   ggplot(aes(x = as.factor(Chr), y = log(Counts + 1), fill = Group)) + 
#   geom_boxplot() +
#   facet_wrap(~Source) +
#   scale_fill_manual(values = c("chr_4" = "red", "chr_9" = "red", "Other" = "gray")) +
#   labs(x = "Chromosome", y = "RNA Count") +
#   theme_classic() +
#   theme(legend.position = "none")
# 
# all_rna_rel_site %>%
#   ggplot(aes(x = as.factor(Chr %in% c(4,9)), y = log(Counts), fill = Group)) + 
#   geom_boxplot() +
#   facet_wrap(~Source) +
#   scale_fill_manual(values = c("chr_4" = "red", "chr_9" = "red", "Other" = "gray")) +
#   labs(x = "Chromosome", y = "RNA Count") +
#   theme_classic() +
#   theme(legend.position = "none")
```
# Working with the Kallisto Counts
## Functions
```{r}
parse_columns <- function(table, sample){
  
  # Parse the columns
  parse.table <- table %>%
    mutate(rna = sub("rna-([^=]+).*", "\\1", target_id),
           gene = sub(".*gene=([^=]+).*", "\\1", target_id),
           name = sub(".*name=([^=]+).*", "\\1", target_id),
           seq_id = sub(".*seq_id=([^=]+).*", "\\1", target_id),
           type = sub(".*type=([^=]+)", "\\1", target_id),
           rna = gsub("gene", "", rna),
           gene = gsub("gene-|name", "", gene),
           name = gsub("seq_id", "", name)) %>%
    select(-target_id) %>%
    select(seq_id, everything()) %>%
    mutate(Source = sample)
  
  # Make the chromosome column numeric
  parse.table$seq_id <- as.numeric(gsub("NC_0499.*?([0-9]+).*", "\\1", parse.table$seq_id))

  # Add groups for later coloring
  final.table <- parse.table %>%
  mutate(Dup_Group = case_when(
    seq_id == 4 ~ "chr_4",
    seq_id == 9 ~ "chr_9",
    TRUE ~ "Other")) %>%
  mutate(Hom_Group = case_when(
    seq_id == 11 ~ "chr_11",
    seq_id == 12 ~ "chr_12",
    seq_id == 16 ~ "chr_16",
    TRUE ~ "Other"))
  return(final.table)
}
  

read_gff <- function(file){
  
  gff <- fread(file, skip = 3)
  
  colnames(gff) <- c("CHROM", "SOURCE", "ID", "START", "STOP", "SCORE", "STRAND", "FRAME", "ATTRIBUTE")
  
  # Parse the columns
  parse.table <- gff %>%
    mutate(gene = sub(".*gene-([^=]+).*", "\\1", ATTRIBUTE),
           POS = (STOP + START) / 2) %>%
    select(-ATTRIBUTE) %>%
    filter(!grepl(";", gene))
  
  gff <- parse.table[grep("^NC_0499", parse.table$CHROM), ]
  gff$CHROM <- as.numeric(gsub("NC_0499(..).1_RagTag", "\\1.", gff$CHROM))
  
  return(gff)  
}
```

## Read in data
```{r}
cr11a <- fread("./Data/Kallisto/11A_abundance.tsv")
cr12a <- fread("./Data/Kallisto/12A_abundance.tsv")
cr13a <- fread("./Data/Kallisto/13A_abundance.tsv")
cr15a <- fread("./Data/Kallisto/15A_abundance.tsv")
cr15a1 <- fread("./Data/Kallisto/15A1_abundance.tsv")
cr16a <- fread("./Data/Kallisto/16A_abundance.tsv")
cr16a1 <- fread("./Data/Kallisto/16A1_abundance.tsv")
cr16a2 <- fread("./Data/Kallisto/16A2_abundance.tsv")
cr17a1 <- fread("./Data/Kallisto/17A1_abundance.tsv")
cr17a2 <- fread("./Data/Kallisto/17A2_abundance.tsv")
cr18a <- fread("./Data/Kallisto/18A_abundance.tsv")
cr18a1 <- fread("./Data/Kallisto/18A1_abundance.tsv")
```

## Parse column information and create large data frame
```{r}
cr11a <- parse_columns(cr11a, "11A")
cr12a <- parse_columns(cr12a, "12A")
cr13a <- parse_columns(cr13a, "13A")
cr15a <- parse_columns(cr15a, "15A")
cr15a1 <- parse_columns(cr15a1, "15A1")
cr16a <- parse_columns(cr16a, "16A")
cr16a1 <- parse_columns(cr16a1, "16A1")
cr16a2 <- parse_columns(cr16a2, "16A2")
cr17a1 <- parse_columns(cr17a1, "17A1")
cr17a2 <- parse_columns(cr17a2, "17A2")
cr18a <- parse_columns(cr18a, "18A")
cr18a1 <- parse_columns(cr18a1, "18A1")

all_counts <- rbind(cr11a, cr12a, cr13a, cr15a, cr15a1, cr16a, cr16a1, cr16a2, cr17a1, cr17a2, cr18a, cr18a1)
```

## Calculate the counts per chromsome
```{r}
chr.table <- all_counts %>%
  filter(tpm > 1.0) %>%
  group_by(Source, seq_id) %>%
  summarise(counts_by_length = sum(est_counts)/sum(length), 
            counts_by_efflength = sum(est_counts)/sum(eff_length), 
            tpm_by_length = sum(tpm)/sum(length),
            tpm_by_efflength = sum(tpm)/sum(eff_length),
            mean_counts = mean(est_counts),
            mean_tpm = mean(tpm), 
            mean_length = mean(length)) %>%
  ungroup() %>%
  mutate(Dup_Group = case_when(
    seq_id == 4 ~ "chr_4",
    seq_id == 9 ~ "chr_9",
    TRUE ~ "Other")) %>%
  mutate(Hom_Group = case_when(
    seq_id == 11 ~ "chr_11",
    seq_id == 12 ~ "chr_12",
    seq_id == 16 ~ "chr_16",
    TRUE ~ "Other"))

```

## Plot
```{r}
chr.table %>%
  ggplot(aes(x = as.factor(seq_id), y = counts_by_length, fill = Dup_Group)) + 
  geom_bar(stat = "identity") +
  facet_wrap(~ Source, scales = "free") +
  scale_fill_manual(values = c("chr_4" = "red", "chr_9" = "red", "Other" = "gray")) +
  labs(x = "Chromosome", y = "Gene transcripts / total exon length") +
  theme_classic() +
  theme(legend.position = "none")
# ggsave("./Figures/RNA_Sequencing/kallisto_counts_over_length_per_chr.pdf", height = 8, width = 12)
# ggsave("./Figures/RNA_Sequencing/kallisto_counts_over_length_per_chr.png", height = 8, width = 12)

chr.table %>%
  ggplot(aes(x = as.factor(seq_id), y = counts_by_efflength, fill = Dup_Group)) + 
  geom_bar(stat = "identity") +
  facet_wrap(~ Source, scales = "free") +
  scale_fill_manual(values = c("chr_4" = "red", "chr_9" = "red", "Other" = "gray")) +
  labs(x = "Chromosome", y = "Gene transcripts / total effective exon length") +
  theme_classic() +
  theme(legend.position = "none")
# ggsave("./Figures/RNA_Sequencing/kallisto_counts_over_efflength_per_chr.pdf", height = 8, width = 12)
# ggsave("./Figures/RNA_Sequencing/kallisto_counts_over_efflength_per_chr.png", height = 8, width = 12)

chr.table %>%
  ggplot(aes(x = as.factor(seq_id), y = tpm_by_length, fill = Dup_Group)) + 
  geom_bar(stat = "identity") +
  facet_wrap(~ Source, scales = "free") +
  scale_fill_manual(values = c("chr_4" = "red", "chr_9" = "red", "Other" = "gray")) +
  labs(x = "Chromosome", y = "Transcript per Million / total exon length") +
  theme_classic() +
  theme(legend.position = "none")
# ggsave("./Figures/RNA_Sequencing/kallisto_tpm_over_length_per_chr.pdf", height = 8, width = 12)
# ggsave("./Figures/RNA_Sequencing/kallisto_tpm_over_length_per_chr.png", height = 8, width = 12)

chr.table %>%
  ggplot(aes(x = as.factor(seq_id), y = tpm_by_efflength, fill = Dup_Group)) + 
  geom_bar(stat = "identity") +
  facet_wrap(~ Source, scales = "free") +
  scale_fill_manual(values = c("chr_4" = "red", "chr_9" = "red", "Other" = "gray")) +
  labs(x = "Chromosome", y = "Transcript per Million / total  effective exon length") +
  theme_classic() +
  theme(legend.position = "none")
# ggsave("./Figures/RNA_Sequencing/kallisto_tpm_over_efflength_per_chr.pdf", height = 8, width = 12)
# ggsave("./Figures/RNA_Sequencing/kallisto_tpm_over_efflength_per_chr.png", height = 8, width = 12)

chr.table %>%
  ggplot(aes(x = as.factor(seq_id), y = mean_counts, fill = Dup_Group)) + 
  geom_bar(stat = "identity") +
  facet_wrap(~ Source, scales = "free") +
  scale_fill_manual(values = c("chr_4" = "red", "chr_9" = "red", "Other" = "gray")) +
  labs(x = "Chromosome", y = "Mean Counts") +
  theme_classic() +
  theme(legend.position = "none")
# ggsave("./Figures/RNA_Sequencing/kallisto_mean_counts_per_chr.pdf", height = 8, width = 12)
# ggsave("./Figures/RNA_Sequencing/kallisto_mean_counts_per_chr.png", height = 8, width = 12)

chr.table %>%
  ggplot(aes(x = as.factor(seq_id), y = mean_tpm, fill = Dup_Group)) + 
  geom_bar(stat = "identity") +
  facet_wrap(~ Source, scales = "free") +
  scale_fill_manual(values = c("chr_4" = "red", "chr_9" = "red", "Other" = "gray")) +
  labs(x = "Chromosome", y = "Mean TPM") +
  theme_classic() +
  theme(legend.position = "none")
# ggsave("./Figures/RNA_Sequencing/kallisto_mean_tpm_per_chr.pdf", height = 8, width = 12)
# ggsave("./Figures/RNA_Sequencing/kallisto_mean_tpm_per_chr.png", height = 8, width = 12)

chr.table %>%
  ggplot(aes(x = as.factor(seq_id), y = mean_length, fill = Dup_Group)) + 
  geom_bar(stat = "identity") +
  facet_wrap(~ Source, scales = "free") +
  scale_fill_manual(values = c("chr_4" = "red", "chr_9" = "red", "Other" = "gray")) +
  labs(x = "Chromosome", y = "Mean Length") +
  theme_classic() +
  theme(legend.position = "none")
ggsave("./Figures/RNA_Sequencing/kallisto_mean_length_per_chr.pdf", height = 8, width = 12)
ggsave("./Figures/RNA_Sequencing/kallisto_mean_length_per_chr.png", height = 8, width = 12)


all_counts %>%
  filter(tpm > 1.0) %>%
  ggplot(aes(x = as.factor(seq_id), y = log(est_counts), fill = Dup_Group)) + 
  geom_boxplot() +
  facet_wrap(~ Source, scales = "free") +
  scale_fill_manual(values = c("chr_4" = "red", "chr_9" = "red", "Other" = "gray")) +
  labs(x = "Chromosome", y = "Estimated Counts") +
  theme_classic() +
  theme(legend.position = "none")
# ggsave("./Figures/RNA_Sequencing/kallisto_boxplot_counts_per_chr.pdf", height = 8, width = 12)
# ggsave("./Figures/RNA_Sequencing/kallisto_boxplot_counts_per_chr.png", height = 8, width = 12)

all_counts %>%
  filter(tpm > 1.0) %>%
  ggplot(aes(x = as.factor(seq_id), y = log(est_counts/length), fill = Dup_Group)) + 
  geom_boxplot() +
  facet_wrap(~Source, scales = "free") +
  scale_fill_manual(values = c("chr_4" = "red", "chr_9" = "red", "Other" = "gray")) +
  labs(x = "Chromosome", y = "Count / total exon length (log)") +
  theme_classic() +
  theme(legend.position = "none")
# ggsave("./Figures/RNA_Sequencing/kallisto_boxplot_counts_over_length_per_chr.pdf", height = 8, width = 12)
# ggsave("./Figures/RNA_Sequencing/kallisto_boxplot_counts_over_length_per_chr.png", height = 8, width = 12)

all_counts %>%
  filter(tpm > 1.0) %>%
  ggplot(aes(x = as.factor(seq_id), y = log(tpm), fill = Dup_Group)) + 
  geom_boxplot() +
  facet_wrap(~Source, scales = "free") +
  scale_fill_manual(values = c("chr_4" = "red", "chr_9" = "red", "Other" = "gray")) +
  labs(x = "Chromosome", y = "TPM") +
  theme_classic() +
  theme(legend.position = "none")
# ggsave("./Figures/RNA_Sequencing/kallisto_boxplot_tpm_per_chr.pdf", height = 8, width = 12)
# ggsave("./Figures/RNA_Sequencing/kallisto_boxplot_tpm_per_chr.png", height = 8, width = 12)

all_counts %>%
  filter(tpm > 1.0) %>%
  ggplot(aes(x = as.factor(seq_id %in% c(4,9)), y = log(tpm), fill = Dup_Group)) + 
  geom_boxplot() +
  facet_wrap(~Source, scales = "free") +
  scale_fill_manual(values = c("chr_4" = "red", "chr_9" = "red", "Other" = "gray")) +
  labs(x = "Chromosome", y = "Log TPM") +
  theme_classic() +
  theme(legend.position = "none")
```

# Plot to compare all chr to 4 and 9
```{r}
# Set levels
levels <- c("11A", "12A", "13A", "15A",
            "15A1", "16A1", "16A2", "17A1",
            "17A2", "18A", "16A", "18A1")
grp_levels <- c("Other", "chr_4", "chr_9")
all_counts$Source <- factor(all_counts$Source, levels = levels)
all_counts$Dup_Group <- factor(all_counts$Dup_Group, levels = grp_levels)

# Specifying t.test comparisons
comparisons <- list(c("chr_4", "Other"), c("chr_9", "Other"))


all_counts %>%
  filter(tpm > 1.0) %>%
  ggplot(aes(x = Dup_Group, y = log(tpm), fill = Dup_Group)) + 
  geom_boxplot() +
  facet_wrap(~ Source, scales = "free") +
  scale_fill_manual(values = c("chr_4" = "firebrick", "chr_9" = "firebrick", "Other" = "gray90")) +
  scale_x_discrete(labels = c("Other" = "Others", "chr_4" = "4", "chr_9" = "9")) +
  labs(x = "Chromosome", y = "TPM") +
  stat_compare_means(method = "t.test", comparisons = comparisons, label = "p.signif") +
  theme_classic(base_size = 6) +
  theme(legend.position = "none",
        axis.title = element_text(face = "bold"),
        strip.background = element_blank(),
        strip.text = element_text(face = "bold"))

all_counts %>%
  filter(tpm > 1.0) %>%
  ggplot(aes(x = Dup_Group, y = log(tpm), fill = Dup_Group)) + 
  geom_boxplot() +
  facet_wrap(~ Source, scales = "free") +
  scale_fill_manual(values = c("chr_4" = "firebrick", "chr_9" = "firebrick", "Other" = "gray90")) +
  scale_x_discrete(labels = c("Other" = "Others", "chr_4" = "4", "chr_9" = "9")) +
  labs(x = "Chromosome", y = "TPM") +
  stat_compare_means(comparisons = comparisons) +
  theme_classic(base_size = 6) +
  theme(legend.position = "none",
        axis.title = element_text(face = "bold"),
        strip.background = element_blank(),
        strip.text = element_text(face = "bold"))

all_counts %>% 
  filter(tpm > 1.0) %>%
  ggplot(aes(x = Dup_Group, y = log(tpm), fill = Dup_Group)) + 
  geom_boxplot() +
  facet_grid(~ Source, scales = "free") +
  scale_fill_manual(values = c("chr_4" = "firebrick", "chr_9" = "firebrick", "Other" = "gray90")) +
  scale_x_discrete(labels = c("Other" = "Others", "chr_4" = "4", "chr_9" = "9")) +
  labs(x = "Chromosome", y = "Natural Log TPM") +
  stat_compare_means(method = "t.test", comparisons = comparisons, label = "p.signif", bracket.size = .3, size = 2) +
  theme_classic(base_size = 6) +
  theme(legend.position = "none",
        axis.title = element_text(face = "bold"),
        strip.background = element_blank(),
        strip.text = element_text(face = "bold"))
#ggsave("./Figures/RNA_Sequencing/compare_chr_49_to_all_ttest.pdf", height = 2.2, width = 5.8)
#ggsave("./Figures/RNA_Sequencing/compare_chr_49_to_all_ttest.png", height = 2.2, width = 5.8)
```

## Look at homozygous regions of the genome
```{r}
comparisons <- list(c("chr_11", "Other"), c("chr_12", "Other"), c("chr_16", "Other"))

all_counts %>%
  filter(!seq_id %in% c(4,9)) %>%
  filter(tpm >= 1.0) %>%
  ggplot(aes(x = Hom_Group, y = log(tpm), fill = Hom_Group)) + 
  geom_boxplot(outlier.size = .25) +
  facet_grid(~ Source, scales = "free") +
  scale_fill_manual(values = c("chr_11" = "firebrick", "chr_12" = "dodgerblue",  "chr_16" = "firebrick", "Other" = "gray90")) +
  scale_x_discrete(labels = c("Other" = "Others", "chr_11" = "11", "chr_12" = "12", "chr_16" = "16")) +
  labs(x = "Chromosome", y = "Natural Log TPM") +
  stat_compare_means(method = "t.test", comparisons = comparisons, label = "p.signif", bracket.size = .3, size = 2) +
  theme_classic(base_size = 6) +
  theme(legend.position = "none",
        axis.title = element_text(face = "bold"),
        strip.background = element_blank(),
        strip.text = element_text(face = "bold"))

#ggsave("./Figures/RNA_Sequencing/compare_chr_111216_to_all_ttest.pdf", height = 2.2, width = 5.8)
#ggsave("./Figures/RNA_Sequencing/compare_chr_111216_to_all_ttest.png", height = 2.2, width = 5.8)
```


## Look for regions of lower expression along the Chromosomes
# Get gene position
```{r}
gff <- read_gff("./Data/GFFs/Chromosome_Only_Liftoff/ref_chandler_primary_default_scaffold_chr_only_liftoff_a99s99.gff")

gene_pos <- gff %>%
  select(CHROM, POS, gene)

gene_pos <- gene_pos %>%
  mutate(gene = sub("_.*", "", gene))

all_counts <- all_counts %>%
  rename(CHROM = seq_id)

all_counts <- left_join(all_counts, gene_pos)


all_counts %>%
  filter(tpm > 1.0) %>%
  filter(Source == "12A") %>%
  ggplot(aes(x = POS, y = tpm)) + 
  geom_point() +
  facet_wrap(~ CHROM, scales = "free") +
  labs(x = "Position", y = "Log TPM") +
  theme_classic(base_size = 6) +
  theme()

str(all_counts)
```
# Creating windows
```{r}
trial <- all_counts %>%
  filter(tpm > 1) %>%
  filter(!is.na(POS))

# Define the window size
window_size <- 1e6

# Create a new column to define windows based on the POS column
trial[, window := floor((POS - 1) / window_size) + 1, by = CHROM]

# Calculate the mean TPM for each window for every chromosome, keeping the Source information
mean_tpm_windows <- trial[, .(mean_tpm = mean(tpm, na.rm = TRUE)), by = .(CHROM, window, Source)]

# Add a group column
mean_tpm_windows <- mean_tpm_windows %>%
    mutate(Dup_Group = case_when(
      CHROM == 4 ~ "chr_4",
      CHROM == 9 ~ "chr_9",
      TRUE ~ "Other"))

# Plotting
mean_tpm_windows %>%
  ggplot(aes(x = window, y = log(mean_tpm), color = Source)) + 
  geom_line() +
  facet_wrap(~ CHROM, scales = "free") +
  labs(x = "Position", y = "Log TPM") +
  scale_color_manual(values = c("16A" = "firebrick", "18A1" = "gold")) +
  theme_classic(base_size = 6) +
  theme()

mean_tpm_windows %>%
  ggplot(aes(x = window, y = log(mean_tpm), color = Source)) + 
  geom_point() +
  facet_wrap(~ CHROM, scales = "free") +
  labs(x = "Position", y = "Log TPM") +
  scale_color_manual(values = c("16A" = "firebrick", "18A1" = "gold")) +
  theme_classic(base_size = 6) +
  theme()
```

