---
title: "testing_filtering"
author: "Matthew Davis"
date: "2024-06-07"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

# Setup
## Libraries
```{r, message=FALSE}
library(tidyverse)
library(data.table)
library(vcfR)
library(viridis)
```

## Functions
```{r}
import_vcf <- function(file, col_to_separate = "UnnamedSample", name){
  # Read the VCF file using vcfR with minimal data
  vcf <- read.vcfR(file, verbose = FALSE)
  
  # Convert @fix and @gt slots to data.tables
  fix <- as.data.table(vcf@fix)
  gt <- as.data.table(vcf@gt)
  all <- cbind(fix, gt)
  
  # Apply filters as early as possible
  vcf.table <- all[grepl("^NC_0499", CHROM), ]
  
  # Extract Hap if present
  vcf.table[, Hap := ifelse(grepl("_hap\\d+", CHROM), as.numeric(gsub(".*_hap(\\d+).*", "\\1", CHROM)), NA)]
  
  # Extract CHROM number after extracting Hap
  vcf.table[, CHROM := as.numeric(gsub("NC_0499(..).1_RagTag(_hap\\d+)?", "\\1", CHROM))]
  
  # Separate genotype information into distinct columns
  vcf.table <- vcf.table[, c("GT", "GQ", "DP", "AD", "VAF", "PL") := tstrsplit(get(col_to_separate), ":", type.convert=TRUE)]
  
  # Create unique identifier
  vcf.table[, unique := ifelse(is.na(Hap), paste(CHROM, POS, sep = "_"), paste(CHROM, Hap, POS, sep = "_"))]
  
  # Convert to numeric
  vcf.table[, QUAL := as.numeric(QUAL)]
  vcf.table[, DP := as.numeric(DP)]
  vcf.table[, VAF := as.numeric(VAF)]
  vcf.table[, GQ := as.numeric(GQ)]
  
  # Add source column
  vcf.table[, Source := name]
  
  return(vcf.table)
}


comp_3way <- function(sample) {
  

  ref_dn <- ref_chandler %>%
    anti_join(tree2, by = "unique") %>%
    anti_join(sample, by = "unique")

  tree2_dn <- tree2 %>%
    anti_join(ref_chandler, by = "unique") %>%
    anti_join(sample, by = "unique")
  
  sample_dn <- sample %>%
    anti_join(ref_chandler, by = "unique") %>%
    anti_join(tree2, by = "unique")

  combo_table <- bind_rows(ref_dn, tree2_dn, sample_dn)

  # Function to generate summary for a specific QUAL value
  generate_summary <- function(data, qual_value) {
    summary <- data %>%
      filter(QUAL >= qual_value) %>%
      group_by(Source) %>%
      summarize(count = n(), .groups = 'drop')
    return(summary)
    }

  # Generate summary for each QUAL value from 0 to 40
  qual_range <- 1:40
  summary_list <- lapply(qual_range, generate_summary, data = combo_table)

  # Combine summaries into a single data frame
  summary_dn <- bind_rows(summary_list, .id = "qual") %>%
    mutate(qual = as.numeric(qual))

  plot <- summary_dn %>%
    ggplot(aes(x = Source, y = count, col = qual)) +
    geom_point() +
    scale_color_viridis(name = "Quality\nThreshold", option = "mako" ) +
    theme_classic()

  return(list(summary_dn = summary_dn, plot = plot))
}

denovo_id <- function(target_df, target_column, ...) {
  
  not_in_values <- Reduce(`c`, lapply(list(...), function(df) df[[target_column]]))
  
  final_df <- target_df %>%
    mutate(denovo = !(!!rlang::sym(target_column) %in% not_in_values))
  
  return(final_df)
}

denovo_snps <- function(vcf.table){
  
  valid_types <- c("0/0", "0/1", "1/1")

  snps <- vcf.table %>%
    filter(FILTER == "PASS") %>%
    filter(nchar(REF) == 1 & nchar(ALT) == 1) %>%
    filter(denovo == TRUE) %>%
    filter(GT %in% valid_types)
  
  return(snps)
}

ancestral_snps <- function(vcf.table){
  
  valid_types <- c("0/0", "0/1", "1/1")

  snps <- vcf.table %>%
    filter(FILTER == "PASS") %>%
    filter(nchar(REF) == 1 & nchar(ALT) == 1) %>%
    filter(denovo == FALSE) %>%
    filter(GT %in% valid_types)
  
  return(snps)
}
```

# Read in data
## VCFs
```{r}
# # HiFi reads
# ref_chandler <- import_vcf("./Data/VCFs/Longread_VCF/Ref_Primary_Aligned/ref_chandler_primary_deepvariant.vcf.gz", name = "ref_chandler")
# tree2 <- import_vcf("./Data/VCFs/Longread_VCF/Ref_Primary_Aligned/term_chandler_primary_deepvariant.vcf.gz", name = "tree2")
# cr2_hifi <- import_vcf("./Data/VCFs/Longread_VCF/Ref_Primary_Aligned/cr2_primary_deepvariant.vcf.gz", name = "cr2_hifi")
# 
# # Chandler shoots short reads
# cr85 <- import_vcf("./Data/VCFs/Shoot_Shortread_VCF/Ref_Primary_Aligned/CR85_S28_deepvariant.vcf.gz", "default", name = "cr85")
# cr10 <- import_vcf("./Data/VCFs/Shoot_Shortread_VCF/Ref_Primary_Aligned/CR10_S29_deepvariant.vcf.gz", "default", name = "cr10")
# cr13 <- import_vcf("./Data/VCFs/Shoot_Shortread_VCF/Ref_Primary_Aligned/CR13_S30_deepvariant.vcf.gz", "default", name = "cr13")
# cr22 <- import_vcf("./Data/VCFs/Shoot_Shortread_VCF/Ref_Primary_Aligned/CR22_S31_deepvariant.vcf.gz", "default", name = "cr22")
# 
# # CR2 original short reads
# cr21_1 <- import_vcf("./Data/VCFs/Original_CR2/Ref_Primary_Aligned/CR21-1_S7_deepvariant.vcf.gz", "default", name = "cr21_1")
# cr21_2 <- import_vcf("./Data/VCFs/Original_CR2/Ref_Primary_Aligned/CR21-2_S8_deepvariant.vcf.gz", "default", name = "cr21_2")
# 
# # CR2 stacks short reads
# cr2_11A <- import_vcf("./Data/VCFs/CR2_Stacks_VCF/Ref_Primary_Aligned/11A_deepvariant.vcf.gz", "default", name = "cr2_11A")
# cr2_12A <- import_vcf("./Data/VCFs/CR2_Stacks_VCF/Ref_Primary_Aligned/12A_deepvariant.vcf.gz", "default", name = "cr2_12A")
# cr2_13A <- import_vcf("./Data/VCFs/CR2_Stacks_VCF/Ref_Primary_Aligned/13A_deepvariant.vcf.gz", "default", name = "cr2_13A")
# cr2_15A <- import_vcf("./Data/VCFs/CR2_Stacks_VCF/Ref_Primary_Aligned/15A_deepvariant.vcf.gz", "default", name = "cr2_15A")
# cr2_15A1 <- import_vcf("./Data/VCFs/CR2_Stacks_VCF/Ref_Primary_Aligned/15A1_deepvariant.vcf.gz", "default", name = "cr2_15A1")
# cr2_16A <- import_vcf("./Data/VCFs/CR2_Stacks_VCF/Ref_Primary_Aligned/16A_deepvariant.vcf.gz", "default", name = "cr2_16A")
# cr2_16A1 <- import_vcf("./Data/VCFs/CR2_Stacks_VCF/Ref_Primary_Aligned/16A1_deepvariant.vcf.gz", "default", name = "cr2_16A1")
# cr2_16A2 <- import_vcf("./Data/VCFs/CR2_Stacks_VCF/Ref_Primary_Aligned/16A2_deepvariant.vcf.gz", "default", name = "cr2_16A2")
# cr2_17A1 <- import_vcf("./Data/VCFs/CR2_Stacks_VCF/Ref_Primary_Aligned/17A1_deepvariant.vcf.gz", "default", name = "cr2_17A1")
# cr2_17A2 <- import_vcf("./Data/VCFs/CR2_Stacks_VCF/Ref_Primary_Aligned/17A2_deepvariant.vcf.gz", "default", name = "cr2_17A2")
# cr2_18A <- import_vcf("./Data/VCFs/CR2_Stacks_VCF/Ref_Primary_Aligned/18A_deepvariant.vcf.gz", "default", name = "cr2_18A")
# cr2_18A1 <- import_vcf("./Data/VCFs/CR2_Stacks_VCF/Ref_Primary_Aligned/18A1_deepvariant.vcf.gz", "default", name = "cr2_18A1")
# 
# # Change UnnamedSample column name for combining data later
# setnames(ref_chandler, "UnnamedSample", "default")
# setnames(tree2, "UnnamedSample", "default")
# setnames(cr2_hifi, "UnnamedSample", "default")
```

# Identify de novo snps
## Three way comparisons only.
Reference and terminal tree as the two constants, rotation of all other samples.
```{r}
# # HiFi
# cr2_hifi_3way <- comp_3way(cr2_hifi)
# 
# # Shoots
# cr85_3way <-comp_3way(cr85)
# cr10_3way <-comp_3way(cr10)
# cr13_3way <-comp_3way(cr13)
# cr22_3way <-comp_3way(cr22)
# 
# # Original short read embryos
# cr21_1_3way <-comp_3way(cr21_1)
# cr21_2_3way <-comp_3way(cr21_2)
# 
# # Embryo stacks
# cr2_11A_3way <-comp_3way(cr2_11A)
# cr2_12A_3way <-comp_3way(cr2_12A)
# cr2_13A_3way <-comp_3way(cr2_13A)
# cr2_15A_3way <-comp_3way(cr2_15A)
# cr2_15A1_3way <-comp_3way(cr2_15A1)
# cr2_16A_3way <-comp_3way(cr2_16A)
# cr2_16A1_3way <-comp_3way(cr2_16A1)
# cr2_16A2_3way <-comp_3way(cr2_16A2)
# cr2_17A1_3way <-comp_3way(cr2_17A1)
# cr2_17A2_3way <-comp_3way(cr2_17A2)
# cr2_18A_3way <-comp_3way(cr2_18A)
# cr2_18A1_3way <-comp_3way(cr2_18A1)
```

## Combining the data frames into a list
```{r}
# cr2_hifi_3way$summary_dn %>%
#   filter(!Source %in% c("ref_chandler", "tree2"))
# 
# # Function to filter the dataframes
# filter_data <- function(df) {
#   df %>% filter(!Source %in% c("ref_chandler", "tree2"))
# }
# 
# # Filter each dataframe
# cr2_hifi_3way_filtered <- filter_data(cr2_hifi_3way$summary_dn)
# cr85_3way_filtered <- filter_data(cr85_3way$summary_dn)
# cr10_3way_filtered <- filter_data(cr10_3way$summary_dn)
# cr13_3way_filtered <- filter_data(cr13_3way$summary_dn)
# cr22_3way_filtered <- filter_data(cr22_3way$summary_dn)
# cr21_1_3way_filtered <- filter_data(cr21_1_3way$summary_dn)
# cr21_2_3way_filtered <- filter_data(cr21_2_3way$summary_dn)
# cr2_11A_3way_filtered <- filter_data(cr2_11A_3way$summary_dn)
# cr2_12A_3way_filtered <- filter_data(cr2_12A_3way$summary_dn)
# cr2_13A_3way_filtered <- filter_data(cr2_13A_3way$summary_dn)
# cr2_15A_3way_filtered <- filter_data(cr2_15A_3way$summary_dn)
# cr2_15A1_3way_filtered <- filter_data(cr2_15A1_3way$summary_dn)
# cr2_16A_3way_filtered <- filter_data(cr2_16A_3way$summary_dn)
# cr2_16A1_3way_filtered <- filter_data(cr2_16A1_3way$summary_dn)
# cr2_16A2_3way_filtered <- filter_data(cr2_16A2_3way$summary_dn)
# cr2_17A1_3way_filtered <- filter_data(cr2_17A1_3way$summary_dn)
# cr2_17A2_3way_filtered <- filter_data(cr2_17A2_3way$summary_dn)
# cr2_18A_3way_filtered <- filter_data(cr2_18A_3way$summary_dn)
# cr2_18A1_3way_filtered <- filter_data(cr2_18A1_3way$summary_dn)
# 
# # Combine all filtered dataframes into one
# combined_filtered_df <- bind_rows(
#   cr2_hifi_3way_filtered, cr85_3way_filtered, cr10_3way_filtered, cr13_3way_filtered, cr22_3way_filtered,
#   cr21_1_3way_filtered, cr21_2_3way_filtered, cr2_11A_3way_filtered, cr2_12A_3way_filtered, cr2_13A_3way_filtered,
#   cr2_15A_3way_filtered, cr2_15A1_3way_filtered, cr2_16A_3way_filtered, cr2_16A1_3way_filtered, cr2_16A2_3way_filtered,
#   cr2_17A1_3way_filtered, cr2_17A2_3way_filtered, cr2_18A_3way_filtered, cr2_18A1_3way_filtered
# )
# 
# desired_levels <- c(
#   "cr2_hifi", "cr85", "cr10", "cr13", "cr22",
#   "cr21_1", "cr21_2", "cr2_11A", "cr2_12A", "cr2_13A",
#   "cr2_15A", "cr2_15A1", "cr2_16A", "cr2_16A1", "cr2_16A2",
#   "cr2_17A1", "cr2_17A2", "cr2_18A", "cr2_18A1"
# )
# 
# combined_filtered_df$Source <- factor(combined_filtered_df$Source, levels = desired_levels)
# 
# # Check the levels
# levels(combined_filtered_df$Source)
# 
# combined_filtered_df %>%
#     ggplot(aes(x = Source, y = count, col = qual)) +
#     geom_point() +
#     scale_color_viridis(name = "Quality\nThreshold", option = "mako" ) +
#     theme_classic()
```

# Finding overlaps between the 3 major groups
The venndiagram. remove the overlaps between embryo, tree, and shoots

## Read in the vcfs
```{r}
# HiFi reads
ref_chandler <- import_vcf("./Data/VCFs/Deepvariantv1.6.1/HiFi/ref_chandler_deepvariant.vcf.gz", name = "ref_chandler")
tree2 <- import_vcf("./Data/VCFs/Deepvariantv1.6.1/HiFi/term_chandler_deepvariant.vcf.gz", name = "tree2")
cr2_hifi <- import_vcf("./Data/VCFs/Deepvariantv1.6.1/HiFi/cr2_deepvariant.vcf.gz", name = "cr2_hifi")

# Chandler shoots short reads
cr85 <- import_vcf("./Data/VCFs/Deepvariantv1.6.1/Short_Read/CR85_deepvariant.vcf.gz", "default", name = "cr85")
cr10 <- import_vcf("./Data/VCFs/Deepvariantv1.6.1/Short_Read/CR10_deepvariant.vcf.gz", "default", name = "cr10")
cr13 <- import_vcf("./Data/VCFs/Deepvariantv1.6.1/Short_Read/CR13_deepvariant.vcf.gz", "default", name = "cr13")
cr22 <- import_vcf("./Data/VCFs/Deepvariantv1.6.1/Short_Read/CR22_deepvariant.vcf.gz", "default", name = "cr22")

# CR2 original short reads
cr21_1 <- import_vcf("./Data/VCFs/Deepvariantv1.6.1/Short_Read/CR21-1_deepvariant.vcf.gz", "default", name = "cr21_1")
cr21_2 <- import_vcf("./Data/VCFs/Deepvariantv1.6.1/Short_Read/CR21-2_deepvariant.vcf.gz", "default", name = "cr21_2")

# CR2 stacks short reads
cr2_11A <- import_vcf("./Data/VCFs/Deepvariantv1.6.1/Short_Read/11A_deepvariant.vcf.gz", "default", name = "cr2_11A")
cr2_12A <- import_vcf("./Data/VCFs/Deepvariantv1.6.1/Short_Read/12A_deepvariant.vcf.gz", "default", name = "cr2_12A")
cr2_13A <- import_vcf("./Data/VCFs/Deepvariantv1.6.1/Short_Read/13A_deepvariant.vcf.gz", "default", name = "cr2_13A")
cr2_15A <- import_vcf("./Data/VCFs/Deepvariantv1.6.1/Short_Read/15A_deepvariant.vcf.gz", "default", name = "cr2_15A")
cr2_15A1 <- import_vcf("./Data/VCFs/Deepvariantv1.6.1/Short_Read/15A1_deepvariant.vcf.gz", "default", name = "cr2_15A1")
cr2_16A <- import_vcf("./Data/VCFs/Deepvariantv1.6.1/Short_Read/16A_deepvariant.vcf.gz", "default", name = "cr2_16A")
cr2_16A1 <- import_vcf("./Data/VCFs/Deepvariantv1.6.1/Short_Read/16A1_deepvariant.vcf.gz", "default", name = "cr2_16A1")
cr2_16A2 <- import_vcf("./Data/VCFs/Deepvariantv1.6.1/Short_Read/16A2_deepvariant.vcf.gz", "default", name = "cr2_16A2")
cr2_17A1 <- import_vcf("./Data/VCFs/Deepvariantv1.6.1/Short_Read/17A1_deepvariant.vcf.gz", "default", name = "cr2_17A1")
cr2_17A2 <- import_vcf("./Data/VCFs/Deepvariantv1.6.1/Short_Read/17A2_deepvariant.vcf.gz", "default", name = "cr2_17A2")
cr2_18A <- import_vcf("./Data/VCFs/Deepvariantv1.6.1/Short_Read/18A_deepvariant.vcf.gz", "default", name = "cr2_18A")
cr2_18A1 <- import_vcf("./Data/VCFs/Deepvariantv1.6.1/Short_Read/18A1_deepvariant.vcf.gz", "default", name = "cr2_18A1")

# Change UnnamedSample column name for combining data later
setnames(ref_chandler, "UnnamedSample", "default")
setnames(tree2, "UnnamedSample", "default")
setnames(cr2_hifi, "UnnamedSample", "default")
```

## Combine vcfs by "type" 
the types are embryo, shoot, tree
```{r}
# tree_list <- list(ref_chandler, tree2)
# #rm(ref_chandler, tree2) # remove from environment to save memory
# 
# shoot_list <- list(cr85, cr10, cr13, cr22)
# #rm(cr85, cr10, cr13, cr22)
# 
# embryo_list <- list(cr2_hifi, cr21_1, cr21_2, cr2_11A, cr2_12A, cr2_13A, cr2_15A, cr2_15A1, cr2_16A, cr2_16A1, cr2_16A2, cr2_17A1, cr2_17A2, cr2_18A, cr2_18A1) # trying a list to reduce memory
# #rm(cr2_hifi, cr21_1, cr21_2, cr2_11A, cr2_12A, cr2_13A, cr2_15A, cr2_15A1, cr2_16A, cr2_16A1, cr2_16A2, cr2_17A1, 
```

## Find the intersections within the lists
```{r}
# # Function to get all unique values in a list of dataframes, ignoring duplicates within each dataframe
# get_all_unique_values <- function(df_list) {
#   unique(unlist(lapply(df_list, function(df) unique(df$unique))))
# }
# 
# # Get all unique values for each list
# tree_unique_values <- get_all_unique_values(tree_list)
# shoot_unique_values <- get_all_unique_values(shoot_list)
# embryo_unique_values <- get_all_unique_values(embryo_list)
# 
# # Find any common 'unique' values between the lists
# common_tree_shoot <- intersect(tree_unique_values, shoot_unique_values)
# common_tree_embryo <- intersect(tree_unique_values, embryo_unique_values)
# common_shoot_embryo <- intersect(shoot_unique_values, embryo_unique_values)
# 
# # Combine all common values
# all_common_values <- unique(c(common_tree_shoot, common_tree_embryo, common_shoot_embryo))
```

## Filter out these values from dataframes
```{r}
# # Function to filter out multiple sets of common values from a list of dataframes
# filter_out_common_values <- function(df_list, common_values_list) {
#   combined_common_values <- unique(unlist(common_values_list))
#   lapply(df_list, function(df) df[!df$unique %in% combined_common_values, ])
# }
# 
# # Filter each list to remove the common 'unique' values
# filtered_tree_list <- filter_out_common_values(tree_list, list(all_common_values, common_tree_shoot, common_tree_embryo))
# filtered_shoot_list <- filter_out_common_values(shoot_list, list(all_common_values, common_tree_shoot, common_shoot_embryo))
# filtered_embryo_list <- filter_out_common_values(embryo_list, list(all_common_values, common_tree_embryo, common_shoot_embryo))

```

## Try using only unique column
```{r}
# Function to filter for source and unique
unique_col <- function(df){
  select_df <- df %>%
  select(unique)
  
  return(select_df)
  
}

ref_chandler_unique <- unique_col(ref_chandler)
tree2_unique <- unique_col(tree2)
trees_snps_unique <- unique(bind_rows(ref_chandler_unique, tree2_unique))

cr85_unique <- unique_col(cr85)
cr10_unique <- unique_col(cr10)
cr13_unique <- unique_col(cr13)
cr22_unique <- unique_col(cr22)
shoots_snps_unique <- unique(bind_rows(cr85_unique, cr10_unique, cr13_unique, cr22_unique))

cr2_hifi_unique <- unique_col(cr2_hifi)
cr21_1_unique <- unique_col(cr21_1)
cr21_2_unique <- unique_col(cr21_2)
cr2_11A_unique <- unique_col(cr2_11A)
cr2_12A_unique <- unique_col(cr2_12A)
cr2_13A_unique <- unique_col(cr2_13A)
cr2_15A_unique <- unique_col(cr2_15A)
cr2_15A1_unique <- unique_col(cr2_15A1)
cr2_16A_unique <- unique_col(cr2_16A)
cr2_16A1_unique <- unique_col(cr2_16A1)
cr2_16A2_unique <- unique_col(cr2_16A2)
cr2_17A1_unique <- unique_col(cr2_17A1)
cr2_17A2_unique <- unique_col(cr2_17A2)
cr2_18A_unique <- unique_col(cr2_18A)
cr2_18A1_unique <- unique_col(cr2_18A1)
embryos_snps_unique_hifi <- unique(cr2_hifi_unique)
embryos_snps_unique_short <- unique(bind_rows(cr21_1_unique, cr21_2_unique, cr2_11A_unique, cr2_12A_unique, cr2_13A_unique, cr2_15A_unique, cr2_15A1_unique, cr2_16A_unique, cr2_16A1_unique, cr2_16A2_unique, cr2_17A1_unique, cr2_17A2_unique, cr2_18A_unique, cr2_18A1_unique))
embryos_snps_unique <- unique(bind_rows(cr2_hifi_unique, cr21_1_unique, cr21_2_unique, cr2_11A_unique, cr2_12A_unique, cr2_13A_unique, cr2_15A_unique, cr2_15A1_unique, cr2_16A_unique, cr2_16A1_unique, cr2_16A2_unique, cr2_17A1_unique, cr2_17A2_unique, cr2_18A_unique, cr2_18A1_unique))
```

## Making the Venn Diagram
```{r}
library(ggVennDiagram)
unique_snps.list <- list("embryo" = embryos_snps_unique$unique,
                         "shoot" = shoots_snps_unique$unique,
                         "tree" = trees_snps_unique$unique)

unique_snps_hifivshort.list <- list("embryo hifi" = embryos_snps_unique_hifi$unique,
                                    "embryo short" = embryos_snps_unique_short$unique,
                                    "shoot" = shoots_snps_unique$unique,
                                    "tree" = trees_snps_unique$unique)
# 
# trees_snps_unique <- trees_snps_unique %>%
#   mutate(id = "tree")
# 
# shoots_snps_unique <- shoots_snps_unique %>%
#   mutate(id = "shoot")
# 
# embryos_snps_unique <- embryos_snps_unique %>%
#   mutate(id = "embryo")

# unique_snps.df <- rbind(trees_snps_unique, shoots_snps_unique, embryos_snps_unique)
ggVennDiagram(x = unique_snps_hifivshort.list,
              category.names = c("Embryo HiFi", "Embryo Short", "Shoots","Trees"),
              set_color = c("goldenrod1","goldenrod3", "darkseagreen","darkseagreen4"),
              edge_size = 2,
              label = "percent",
              label_percent_digit = 1,
              label_alpha = 0,
              label_color = "black") +
  scale_fill_distiller(palette = "Greys", direction = 1)
ggsave("~/Desktop/venndiagram_hifi_vs_short_snps.pdf", height = 5, width = 7)
ggsave("~/Desktop/venndiagram_hifi_vs_short_snps.png", height = 5, width = 7)

ggVennDiagram(x = unique_snps.list,
              category.names = c("Embryos","Shoots","Trees"),
              set_color = c("goldenrod1","darkseagreen","darkseagreen4"),
              edge_size = 2,
              label = "percent",
              label_percent_digit = 1,
              label_alpha = 0,
              label_color = "black") +
  scale_fill_distiller(palette = "Greys", direction = 1)
ggsave("~/Desktop/venndiagram_all_snps.pdf", height = 5, width = 7)
ggsave("~/Desktop/venndiagram_all_snps.png", height = 5, width = 7)
```

## Identifying snps in the intersections of the venn diagram
```{r}
tree_shoot_snps <- inner_join(trees_snps_unique, shoots_snps_unique)
tree_embryo_snps <- inner_join(trees_snps_unique, embryos_snps_unique)
shoot_embryo_snps <- inner_join(shoots_snps_unique, embryos_snps_unique)

tree_shoot_embryo_snps <- tree_shoot_snps %>%
  inner_join(shoots_snps_unique) %>%
  inner_join(embryos_snps_unique)

# combining all to identify all ancestral snps
all_anc_snps <- unique(bind_rows(tree_shoot_snps, tree_embryo_snps, shoot_embryo_snps, tree_shoot_embryo_snps))

rm(embryos_snps_unique, embryos_snps_unique_hifi, embryos_snps_unique_short, trees_snps_unique,shoots_snps_unique, 
   tree_shoot_snps, tree_embryo_snps, shoot_embryo_snps, tree_shoot_embryo_snps, unique_snps.list,
   unique_snps_hifivshort.list)
```

## Filtering all for denovo snps
```{r}
# Hifi
ref_chandler_dn <- ref_chandler %>%
  anti_join(all_anc_snps)

tree2_dn <- tree2 %>%
  anti_join(all_anc_snps)

cr2_hifi_dn <- cr2_hifi %>%
  anti_join(all_anc_snps)

# Short read shoots
cr85_dn <- cr85 %>%
  anti_join(all_anc_snps)

cr10_dn <- cr10 %>%
  anti_join(all_anc_snps)

cr13_dn <- cr13 %>%
  anti_join(all_anc_snps)

cr22_dn <- cr22 %>%
  anti_join(all_anc_snps)

# Original embryo cultures
cr21_1_dn <- cr21_1 %>%
  anti_join(all_anc_snps)

cr21_2_dn <- cr21_2 %>%
  anti_join(all_anc_snps)

# Embryo stacks
cr2_11A_dn <- cr2_11A %>%
  anti_join(all_anc_snps)

cr2_12A_dn <- cr2_12A %>%
  anti_join(all_anc_snps)

cr2_13A_dn <- cr2_13A %>%
  anti_join(all_anc_snps)

cr2_15A_dn <- cr2_15A %>%
  anti_join(all_anc_snps)

cr2_15A1_dn <- cr2_15A1 %>%
  anti_join(all_anc_snps)

cr2_16A_dn <- cr2_16A %>%
  anti_join(all_anc_snps)

cr2_16A1_dn <- cr2_16A1 %>%
  anti_join(all_anc_snps)

cr2_16A2_dn <- cr2_16A2 %>%
  anti_join(all_anc_snps)

cr2_17A1_dn <- cr2_17A1 %>%
  anti_join(all_anc_snps)

cr2_17A2_dn <- cr2_17A2 %>%
  anti_join(all_anc_snps)

cr2_18A_dn <- cr2_18A %>%
  anti_join(all_anc_snps)

cr2_18A1_dn <- cr2_18A1 %>%
  anti_join(all_anc_snps)

all_dn_snps <- bind_rows(ref_chandler_dn, tree2_dn, cr2_hifi_dn,
                         cr85_dn, cr10_dn, cr13_dn, cr22_dn,
                         cr21_1_dn, cr21_2_dn, cr2_11A_dn, cr2_12A_dn,
                         cr2_13A_dn, cr2_15A_dn, cr2_15A1_dn, cr2_16A_dn,
                         cr2_16A1_dn, cr2_16A2_dn, cr2_17A1_dn, cr2_17A2_dn,
                         cr2_18A_dn, cr2_18A1_dn)
rm(ref_chandler_dn, tree2_dn, cr2_hifi_dn, cr85_dn, cr10_dn, cr13_dn, 
   cr22_dn, cr21_1_dn, cr21_2_dn, cr2_11A_dn, cr2_12A_dn, cr2_13A_dn, 
   cr2_15A_dn, cr2_15A1_dn, cr2_16A_dn, cr2_16A1_dn, cr2_16A2_dn, 
   cr2_17A1_dn, cr2_17A2_dn, cr2_18A_dn, cr2_18A1_dn)
```

## Write out vcf
```{eval = FALSE}
all_dn_snps
write_tsv(x = all_dn_snps, file = "./Data/VCFs/Filtered/all_dn_snps.vcf")
```

## Filter SNPs shared between independent samples, trees and shoots
```{r}
# Looking for shared SNPs among tree samples
ref_chandler_data <- all_dn_snps %>% filter(Source == 'ref_chandler')
tree2_data <- all_dn_snps %>% filter(Source == 'tree2')
unique_tree.list <- list("ref" = ref_chandler_data$unique,
                         "tree2" = tree2_data$unique)

# Plot venndiagram to see SNP overlap
ggVennDiagram(x = unique_tree.list,
              edge_size = 2,
              label = "count",
              label_alpha = 0,
              label_color = "black") +
  scale_fill_distiller(palette = "Greys", direction = 1)

# Find intersection of tree values
common_unique_tree <- intersect(ref_chandler_data$unique, tree2_data$unique)
length(common_unique_tree)

# Look into shared SNPs between the shoots
cr85_data <- all_dn_snps %>% filter(Source == 'cr85')
cr10_data <- all_dn_snps %>% filter(Source == 'cr10')
cr13_data <- all_dn_snps %>% filter(Source == 'cr13')
cr22_data <- all_dn_snps %>% filter(Source == 'cr22')
unique_shoot.list <- list("cr85" = cr85_data$unique,
                          "cr10" = cr10_data$unique,
                          "cr13" = cr13_data$unique,
                          "cr22" = cr22_data$unique)

#Plot
ggVennDiagram(x = unique_shoot.list,
              edge_size = 2,
              label = "count",
              label_alpha = 0,
              label_color = "black") +
  scale_fill_distiller(palette = "Greys", direction = 1)

# Function to find intersections between multiple data frames
intersect_multiple <- function(...) {
  Reduce(intersect, list(...))
}

# Pairwise intersections
intersect_cr85_cr10 <- intersect(cr85_data$unique, cr10_data$unique)
intersect_cr85_cr13 <- intersect(cr85_data$unique, cr13_data$unique)
intersect_cr85_cr22 <- intersect(cr85_data$unique, cr22_data$unique)
intersect_cr10_cr13 <- intersect(cr10_data$unique, cr13_data$unique)
intersect_cr10_cr22 <- intersect(cr10_data$unique, cr22_data$unique)
intersect_cr13_cr22 <- intersect(cr13_data$unique, cr22_data$unique)

# Intersections involving three data frames
intersect_cr85_cr10_cr13 <- intersect_multiple(cr85_data$unique, cr10_data$unique, cr13_data$unique)
intersect_cr85_cr10_cr22 <- intersect_multiple(cr85_data$unique, cr10_data$unique, cr22_data$unique)
intersect_cr85_cr13_cr22 <- intersect_multiple(cr85_data$unique, cr13_data$unique, cr22_data$unique)
intersect_cr10_cr13_cr22 <- intersect_multiple(cr10_data$unique, cr13_data$unique, cr22_data$unique)

# Intersection of all four data frames
intersect_all <- intersect_multiple(cr85_data$unique, cr10_data$unique, cr13_data$unique, cr22_data$unique)

# List of all intersections
intersections_list <- list(
  intersect_cr85_cr10 = intersect_cr85_cr10,
  intersect_cr85_cr13 = intersect_cr85_cr13,
  intersect_cr85_cr22 = intersect_cr85_cr22,
  intersect_cr10_cr13 = intersect_cr10_cr13,
  intersect_cr10_cr22 = intersect_cr10_cr22,
  intersect_cr13_cr22 = intersect_cr13_cr22,
  intersect_cr85_cr10_cr13 = intersect_cr85_cr10_cr13,
  intersect_cr85_cr10_cr22 = intersect_cr85_cr10_cr22,
  intersect_cr85_cr13_cr22 = intersect_cr85_cr13_cr22,
  intersect_cr10_cr13_cr22 = intersect_cr10_cr13_cr22,
  intersect_all = intersect_all
)

# Create a table of sites to remove
shoot_instersects <- do.call(rbind, intersections_list)

shoot_instersects <- as.data.table(shoot_instersects) %>%
  pivot_longer(cols = everything())

shoot_instersects_unique <- unique(shoot_instersects$value)

within_shoot_tree_intersects <- unique(c(shoot_instersects_unique,common_unique_tree))

within_shoot_tree_intersects <- as.data.frame(within_shoot_tree_intersects)

colnames(within_shoot_tree_intersects)[colnames(within_shoot_tree_intersects) == "within_shoot_tree_intersects"] <- "unique"

# Remove these sites from the de novo snp list
filt_all_dn_snp <- all_dn_snps %>%
  anti_join(within_shoot_tree_intersects)

# Verify the sites were removed from trees
ref_chandler_data <- filt_all_dn_snp %>% filter(Source == 'ref_chandler')
tree2_data <- filt_all_dn_snp %>% filter(Source == 'tree2')

unique_tree.list <- list("ref" = ref_chandler_data$unique,
                         "tree2" = tree2_data$unique)

ggVennDiagram(x = unique_tree.list,
              edge_size = 2,
              label = "count",
              label_alpha = 0,
              label_color = "black") +
  scale_fill_distiller(palette = "Greys", direction = 1)

# Verify the sites were removed from shoots
cr85_data <- filt_all_dn_snp %>% filter(Source == 'cr85')
cr10_data <- filt_all_dn_snp %>% filter(Source == 'cr10')
cr13_data <- filt_all_dn_snp %>% filter(Source == 'cr13')
cr22_data <- filt_all_dn_snp %>% filter(Source == 'cr22')
unique_shoot.list <- list("cr85" = cr85_data$unique,
                          "cr10" = cr10_data$unique,
                          "cr13" = cr13_data$unique,
                          "cr22" = cr22_data$unique)


ggVennDiagram(x = unique_shoot.list,
              edge_size = 2,
              label = "count",
              label_alpha = 0,
              label_color = "black") +
  scale_fill_distiller(palette = "Greys", direction = 1)
```

## Write out vcf
```{r, eval=FALSE}
filt_all_dn_snp
write_tsv(x = filt_all_dn_snp, file = "./Data/VCFs/Filtered/dn_snps_no_share.vcf")
```


## Filtering for all ancestral snps
```{r}
rm(all_dn_snps,filt_all_dn_snp)

# Hifi
ref_chandler_an <- ref_chandler %>%
  inner_join(all_anc_snps)
rm(ref_chandler, ref_chandler_unique)

tree2_an <- tree2 %>%
  inner_join(all_anc_snps)
rm(tree2, tree2_unique)

cr2_hifi_an <- cr2_hifi %>%
  inner_join(all_anc_snps)
rm(cr2_hifi, cr2_hifi_unique)

# Short read shoots
cr85_an <- cr85 %>%
  inner_join(all_anc_snps)
rm(cr85, cr85_unique)

cr10_an <- cr10 %>%
  inner_join(all_anc_snps)
rm(cr10, cr10_unique)

cr13_an <- cr13 %>%
  inner_join(all_anc_snps)
rm(cr13, cr13_unique)

cr22_an <- cr22 %>%
  inner_join(all_anc_snps)
rm(cr22, cr22_unique)

# Original embryo cultures
cr21_1_an <- cr21_1 %>%
  inner_join(all_anc_snps)
rm(cr21_1, cr21_1_unique)

cr21_2_an <- cr21_2 %>%
  inner_join(all_anc_snps)
rm(cr21_2, cr21_2_unique)

# Embryo stacks
cr2_11A_an <- cr2_11A %>%
  inner_join(all_anc_snps)
rm(cr2_11A, cr2_11A_unique)

cr2_12A_an <- cr2_12A %>%
  inner_join(all_anc_snps)
rm(cr2_12A, cr2_12A_unique)

cr2_13A_an <- cr2_13A %>%
  inner_join(all_anc_snps)
rm(cr2_13A, cr2_13A_unique)

cr2_15A_an <- cr2_15A %>%
  inner_join(all_anc_snps)
rm(cr2_15A, cr2_15A_unique)

cr2_15A1_an <- cr2_15A1 %>%
  inner_join(all_anc_snps)
rm(cr2_15A1, cr2_15A1_unique)

cr2_16A_an <- cr2_16A %>%
  inner_join(all_anc_snps)
rm(cr2_16A, cr2_16A_unique)

cr2_16A1_an <- cr2_16A1 %>%
  inner_join(all_anc_snps)
rm(cr2_16A1, cr2_16A1_unique)

cr2_16A2_an <- cr2_16A2 %>%
  inner_join(all_anc_snps)
rm(cr2_16A2, cr2_16A2_unique)

cr2_17A1_an <- cr2_17A1 %>%
  inner_join(all_anc_snps)
rm(cr2_17A1, cr2_17A1_unique)

cr2_17A2_an <- cr2_17A2 %>%
  inner_join(all_anc_snps)
rm(cr2_17A2, cr2_17A2_unique)

cr2_18A_an <- cr2_18A %>%
  inner_join(all_anc_snps)
rm(cr2_18A, cr2_18A_unique)

cr2_18A1_an <- cr2_18A1 %>%
  inner_join(all_anc_snps)
rm(cr2_18A1, cr2_18A1_unique)

all_an_snps <- list(ref_chandler_an, tree2_an, cr2_hifi_an,
                         cr85_an, cr10_an, cr13_an, cr22_an,
                         cr21_1_an, cr21_2_an, cr2_11A_an, cr2_12A_an,
                         cr2_13A_an, cr2_15A_an, cr2_15A1_an, cr2_16A_an,
                         cr2_16A1_an, cr2_16A2_an, cr2_17A1_an, cr2_17A2_an,
                         cr2_18A_an, cr2_18A1_an)


rm(ref_chandler_an, tree2_an, cr2_hifi_an, cr85_an, cr10_an, cr13_an, 
   cr22_an, cr21_1_an, cr21_2_an, cr2_11A_an, cr2_12A_an, cr2_13A_an, 
   cr2_15A_an, cr2_15A1_an, cr2_16A_an, cr2_16A1_an, cr2_16A2_an, 
   cr2_17A1_an, cr2_17A2_an, cr2_18A_an, cr2_18A1_an, all_anc_snps)
```
## Separate the ancestral snps by chromosome and save the file. CAUTION, see note below
NOTE: Only run this one time! The files are being appended. If it needs to be rerun, files must be deleted.
```{r, eval=FALSE}
# NOTE: Only run this one time! The files are being appended. If it needs to be rerun, files must be deleted.
# Create separate files for each CHROM value and write the header
chrom_values <- 1:16
file_paths <- paste0("./Data/VCFs/Filtered/all_an_snps_chr_", chrom_values, ".vcf")

# Initialize empty files with header
initialize_files_with_header <- function(df_list, file_paths) {
  header <- colnames(df_list[[1]])
  lapply(file_paths, function(fp) write.table(t(header), fp, col.names = FALSE, row.names = FALSE, sep = "\t", quote = FALSE))
}

initialize_files_with_header(all_an_snps, file_paths)

# Function to append data to the corresponding file based on CHROM value
append_to_file <- function(df) {
  split_dfs <- split(df, df$CHROM)
  for (chrom in names(split_dfs)) {
    file_path <- paste0("./Data/VCFs/Filtered/all_an_snps_chr_", chrom, ".vcf")
    write.table(split_dfs[[chrom]], file_path, append = TRUE, col.names = FALSE, row.names = FALSE, sep = "\t", quote = FALSE)
  }
}

# Process each data frame and append to the appropriate files
lapply(all_an_snps, append_to_file)
```


