---
title: "tree_creation"
author: "Matthew Davis"
date: "2024-07-11"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

# Setup
## Libraries
```{r, message=FALSE}
library(tidyverse)
library(data.table)
library(janitor)
library(ggdendro)
library(dendextend)
library(Biostrings)
library(DECIPHER)
```

## Functions
```{r}
snp_matrix <- function(vcf.data, qual){
  
  snp.matrix <- vcf.data %>%
    filter(QUAL >= qual) %>%
    select(Source, unique) %>%
    mutate(present = 1) %>%
    pivot_wider(names_from = unique, values_from = present, values_fill = list(present = 0))
  
  snp.matrix <- snp.matrix %>%
    column_to_rownames("Source")

  return(snp.matrix)
}

dendro_plot <- function(clust.data, clusters = 3, height = 3, color = "red", qual){
  
  main_title <- paste("Dendrogram -", qual)
  plot(clust.data, main = main_title, xlab = "Sources", sub = "", cex = 0.6)
  rect.hclust(clust.data , k = clusters, border = 2:6)
  abline(h = height, col = color)
}

read_ancestral_snps <- function(path){
  # Read in data
  data <- fread(path)
  
  table <- data %>%
    select(Source, unique, QUAL)
  
  return(table)
}
```


## Read in files and create  chr 4 and 9 files
```{r}
dn_snps <- fread("./Data/VCFs/Filtered/all_dn_snps.vcf")

chr4_dn_snps <- dn_snps %>%
  filter(CHROM %in% 4) %>%
  filter(!Source %in% c("ref_chandler", "tree2", "cr85", "cr10", "cr13", "cr22"))

chr9_dn_snps <- dn_snps %>%
  filter(CHROM %in% 9) %>%
  filter(!Source %in% c("ref_chandler", "tree2", "cr85", "cr10", "cr13", "cr22"))
```

# Make SNP tree for denovo variants
## Make matrix
```{r}
# Create matrix
snp.matrix0 <- snp_matrix(dn_snps, qual = 0)
snp.matrix20 <- snp_matrix(dn_snps, qual = 20)
snp.matrix30 <- snp_matrix(dn_snps, qual = 30)
snp.matrix40 <- snp_matrix(dn_snps, qual = 40)

chr4_snp.matrix20 <- snp_matrix(chr4_dn_snps, qual = 20)
chr9_snp.matrix20 <- snp_matrix(chr9_dn_snps, qual = 20)
```

## Calculate the bindary distance between samples and cluster the matrix
```{r}
## distance
dist.matrix0 <- dist(snp.matrix0, method = "binary")
dist.matrix20 <- dist(snp.matrix20, method = "binary")
dist.matrix30 <- dist(snp.matrix30, method = "binary")
dist.matrix40 <- dist(snp.matrix40, method = "binary")

chr4_dist.matrix20 <- dist(chr4_snp.matrix20, method = "binary")
chr9_dist.matrix20 <- dist(chr9_snp.matrix20, method = "binary")

## cluster
clust_avg0 <- hclust(dist.matrix0, method = "average")
clust_avg20 <- hclust(dist.matrix20, method = "average")
clust_avg30 <- hclust(dist.matrix30, method = "average")
clust_avg40 <- hclust(dist.matrix40, method = "average")

chr4_clust_avg20 <- hclust(chr4_dist.matrix20, method = "average")
chr9_clust_avg20 <- hclust(chr9_dist.matrix20, method = "average")


# Define expected clusters. These are the different sample groups
#clusters <- cutree(clust_avg, k = 3)  # Adjust 'k' as needed
```

## Plot the tree
```{r}
## Base R
pdf("./Figures/Trees/dn_snps_dendrogram_q>=0.pdf", height = 5, width = 8)
dendro_plot(clust_avg0, qual = "q>=0")
dev.off()
pdf("./Figures/Trees/dn_snps_dendrogram_q>=20.pdf", height = 5, width = 8)
dendro_plot(clust_avg20, qual = "q>=20")
dev.off()
pdf("./Figures/Trees/dn_snps_dendrogram_q>=30.pdf", height = 5, width = 8)
dendro_plot(clust_avg30, qual = "q>=30")
dev.off()
pdf("./Figures/Trees/dn_snps_dendrogram_q>=40.pdf", height = 5, width = 8)
dendro_plot(clust_avg40, qual = "q>=40")
dev.off()

dendro_plot(chr4_clust_avg20, qual = "q>=20", clusters = 2)
dendro_plot(chr9_clust_avg20, qual = "q>=20", clusters = 2)

# ## GG Dendro
# dendro <- as.dendrogram(clust_avg)
# dendro_dn_snps <- dendro_data(dendro)
# 
# ## Cluster test
# # Convert dendrogram labels to data frame for ggplot
# # Convert dendrogram to dendro_data
# dendro <- as.dendrogram(clust_avg)
# dendro_dn_snps <- dendro_data(dendro)
# 
# # Convert dendrogram labels to data frame for ggplot
# label_data <- dendro_dn_snps$labels %>%
#   as.data.frame() %>%
#   mutate(cluster = clusters[match(label, names(clusters))])
# 
# # Calculate x positions for segments based on dendrogram structure
# segment_data <- dendro_dn_snps$segments %>%
#   as.data.frame()
# 
# # Define colors for each cluster
# cluster_colors <- c("red", "blue", "green")  # Add more colors as needed
# 
# ggplot(segment_data) +
#   geom_segment(aes(x = x, y = y, xend = xend, yend = yend, color = label_data$cluster), size = 1) +
#   coord_flip() + 
#   scale_color_manual(values = cluster_colors) +
#   geom_text(aes(x = x, y = y, label = label, hjust = 0), data = label_data) +
#   scale_y_reverse(expand = c(0.2, 0)) +
#   theme_dendro()
```

# make ancestral sites tree
```{r}
# Read in only necessary columns
chr1 <- read_ancestral_snps("./Data/VCFs/Filtered/all_an_snps_chr_1.vcf")
chr2 <- read_ancestral_snps("./Data/VCFs/Filtered/all_an_snps_chr_2.vcf")
chr3 <- read_ancestral_snps("./Data/VCFs/Filtered/all_an_snps_chr_3.vcf")
chr4 <- read_ancestral_snps("./Data/VCFs/Filtered/all_an_snps_chr_4.vcf")
chr5 <- read_ancestral_snps("./Data/VCFs/Filtered/all_an_snps_chr_5.vcf")
chr6 <- read_ancestral_snps("./Data/VCFs/Filtered/all_an_snps_chr_6.vcf")
chr7 <- read_ancestral_snps("./Data/VCFs/Filtered/all_an_snps_chr_7.vcf")
chr8 <- read_ancestral_snps("./Data/VCFs/Filtered/all_an_snps_chr_8.vcf")
chr9 <- read_ancestral_snps("./Data/VCFs/Filtered/all_an_snps_chr_9.vcf")
chr10 <- read_ancestral_snps("./Data/VCFs/Filtered/all_an_snps_chr_10.vcf")
chr11 <- read_ancestral_snps("./Data/VCFs/Filtered/all_an_snps_chr_11.vcf")
chr12 <- read_ancestral_snps("./Data/VCFs/Filtered/all_an_snps_chr_12.vcf")
chr13 <- read_ancestral_snps("./Data/VCFs/Filtered/all_an_snps_chr_13.vcf")
chr14 <- read_ancestral_snps("./Data/VCFs/Filtered/all_an_snps_chr_14.vcf")
chr15 <- read_ancestral_snps("./Data/VCFs/Filtered/all_an_snps_chr_15.vcf")
chr16 <- read_ancestral_snps("./Data/VCFs/Filtered/all_an_snps_chr_16.vcf") 

#Bind all columns
an_snps <- rbind(chr1, chr2, chr3, chr4, chr5, chr6, chr7, chr8, chr9, chr10, chr11, chr12, chr13, chr14, chr15, chr16)

# Remove unnecessary files
rm(chr1, chr2, chr3, chr4, chr5, chr6, chr7, chr8, chr9, chr10, chr11, chr12, chr13, chr14, chr15, chr16)
```

## Make matrix
```{r}
# Create matrix
an.matrix0 <- snp_matrix(an_snps, qual = 0)
an.matrix20 <- snp_matrix(an_snps, qual = 20)
an.matrix30 <- snp_matrix(an_snps, qual = 30)
an.matrix40 <- snp_matrix(an_snps, qual = 40)
```
## Calculate the bindary distance between samples and cluster the matrix
```{r}
## distance
an_dist.matrix0 <- dist(an.matrix0, method = "binary")
an_dist.matrix20 <- dist(an.matrix20, method = "binary")
an_dist.matrix30 <- dist(an.matrix30, method = "binary")
an_dist.matrix40 <- dist(an.matrix40, method = "binary")

## cluster
an_clust_avg0 <- hclust(an_dist.matrix0, method = "average")
an_clust_avg20 <- hclust(an_dist.matrix20, method = "average")
an_clust_avg30 <- hclust(an_dist.matrix30, method = "average")
an_clust_avg40 <- hclust(an_dist.matrix40, method = "average")

# Define expected clusters. These are the different sample groups
#clusters <- cutree(clust_avg, k = 3)  # Adjust 'k' as needed
```

## Plot the tree
```{r}
## Base R
pdf("./Figures/Trees/an_snps_dendrogram_q>=0.pdf", height = 5, width = 8)
dendro_plot(an_clust_avg0, qual = "q>=0")
dev.off()
pdf("./Figures/Trees/an_snps_dendrogram_q>=20.pdf", height = 5, width = 8)
dendro_plot(an_clust_avg20, qual = "q>=20")
dev.off()
pdf("./Figures/Trees/an_snps_dendrogram_q>=30.pdf", height = 5, width = 8)
dendro_plot(an_clust_avg30, qual = "q>=30")
dev.off()
pdf("./Figures/Trees/an_snps_dendrogram_q>=40.pdf", height = 5, width = 8)
dendro_plot(an_clust_avg40, qual = "q>=40")
dev.off()
# Having some trouble interpreting some of these clusters
```

# Making a TE tree
## Read in data with BioStrings and Decipher
```{r}
te_5453 <- readDNAStringSet("./Data/SV_Outputs/cr2_5453sv.fasta", format = "fasta")
te_5454 <- readDNAStringSet("./Data/SV_Outputs/cr2_5454sv.fasta", format = "fasta")
te_5455 <- readDNAStringSet("./Data/SV_Outputs/cr2_5455sv.fasta", format = "fasta")
te_899 <- readDNAStringSet("./Data/SV_Outputs/cr2_899sv.fasta", format = "fasta")

# All fastas
te <- c(te_899, te_5453, te_5454, te_5455)
```

## Make matrix and cluster
```{r}
# Make distance matrix
dist_matrix_5453 <- DistanceMatrix(te_5453)
dist_matrix_5454 <- DistanceMatrix(te_5454)
dist_matrix_5455 <- DistanceMatrix(te_5455)
dist_matrix_899 <- DistanceMatrix(te_899)

dist_matrix_te <- DistanceMatrix(te)

# Make the matrix a dist object
dist_matrix_5453 <- as.dist(dist_matrix_5453)
dist_matrix_5454 <- as.dist(dist_matrix_5454)
dist_matrix_5455 <- as.dist(dist_matrix_5455)
dist_matrix_899 <- as.dist(dist_matrix_899)

dist_matrix_te <- as.dist(dist_matrix_te)

# Cluster
clust_5453 <- hclust(dist_matrix_5453, method = "average")
clust_5454 <- hclust(dist_matrix_5454, method = "average")
clust_5455 <- hclust(dist_matrix_5455, method = "average")
clust_899 <- hclust(dist_matrix_899, method = "average")

clust_te <- hclust(dist_matrix_te, method = "average")
```

## Plot
```{r}
plot(clust_5453)
plot(clust_5454)
plot(clust_5455)
plot(clust_899)

plot(clust_te)
rect.hclust(clust_te , k = 4, border = 2:6)
abline(h = 3, col = 'red')
```

