---
title: "snp_tricontext_analysis"
author: "Matthew Davis"
date: "2024-11-15"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

# Setup
## Libraries
```{r, message=FALSE}
library(polymorphology2)
library(parallel)
library(ggenomics)
library(ggtext)
library(ggh4x)
```

## Functions
```{r}
fasta_trimer_freq<-function(fasta, stop = NULL){
  fasta_windows<-rbindlist(lapply(1:length(fasta), function(i){
    data.table(CHROM=names(fasta)[i], START=1, STOP=length(fasta[[i]]), ID=i)
    data.table(CHROM=names(fasta)[i], START=1, STOP=ifelse(is.null(stop), length(fasta[[i]]), stop), ID=i)
    
  }))[!is.na(CHROM)]
  trimer_freq<-Nmerfrequency(features = fasta_windows, fasta=fasta, Nmer = 3, mode="counts")
  trimer_freq<-rbindlist(lapply(1:ncol(trimer_freq), function(i){
    data.table(TRI=names(trimer_freq)[i], N=sum(trimer_freq[[i]]))
  }))
  
  trimer_freq<-trimer_freq[!grepl("ID|N|V", TRI)]
}

```

# Calculate trimer frequency
## Read in fasta
```{r, eval=FALSE}
fasta <- read.fasta("./Data/Fastas/ref_chandler_primary_default_scaffold.fasta")
```

## Filter fasta for only chromosomes
```{r, eval=FALSE}
names(fasta)<-as.numeric(gsub("NC_0499(..).1_RagTag", "\\1", names(fasta)))
fasta <- fasta[!is.na(names(fasta))]
```

## Calculate trimer frequency
This takes about 2 hours
```{r, eval=FALSE}
## Needs library(parallel)
freq_trimer <- fasta_trimer_freq(fasta, stop = NULL)
```

## Write trimer frequency for ref_chandler_primary
```{r, eval=FALSE}
fwrite(freq_trimer, "./Data/Trimer_Frequency/ref_chandler_primary_chr_only_trimer_freq.csv")
```

## Read in data
```{r}
mutations <- fread("./Data/VCFs/Parsed/mutations_strict.tsv")
freq_trimer <- fread("./Data/Trimer_Frequency/ref_chandler_primary_chr_only_trimer_freq.csv")
```

# Prepare data
## Filter mutations for SNPs
```{r}
snps <- mutations[medQUAL >= 30 & medDPREAL >= 15 & MAPP == T & SNP == T]
indels <- mutations[medQUAL >= 30 & medDPREAL >= 15 & MAPP == T & SNP == F]
```

# Plotting data
## Polymorphlolgy plotting function full data
```{r}
tree_plot <- plot_tricontexts(snps[CULTURETYPE=="tree",.(UNIQUE, TRICONTEXT)]$TRICONTEXT, full=T, trimer_freq = freq_trimer, x_text = F)
tree_plot$plot +
  theme(
    strip.text = element_text(face = "plain"),
    plot.margin = ggplot2::margin(t = 0, r = 0.01, b = 0, l = 0.01, unit = "cm")
  )
ggsave("./Figures/Tricontext/tree_tricontext.pdf", height = 1, width = 3.75)
shoot_plot <- plot_tricontexts(snps[CULTURETYPE=="shoot",.(UNIQUE, TRICONTEXT)]$TRICONTEXT, full=T, trimer_freq = freq_trimer, x_text = F)
shoot_plot$plot +
  theme(
    strip.text = element_text(face = "plain"),
    plot.margin = ggplot2::margin(t = 0, r = 0.01, b = 0, l = 0.01, unit = "cm")
  )
ggsave("./Figures/Tricontext/shoot_tricontext.pdf", height =1, width = 3.75)
embryo_plot <- plot_tricontexts(snps[CULTURETYPE=="embryo",.(UNIQUE, TRICONTEXT)]$TRICONTEXT, full=T, trimer_freq = freq_trimer,  x_text = T)
embryo_plot$plot +
  theme(
    strip.text = element_text(face = "plain"),
    plot.margin = ggplot2::margin(t = 0, r = 0.01, b = 0, l = 0.01, unit = "cm")
  )
ggsave("./Figures/Tricontext/embryo_tricontext.pdf", height = 1.5, width = 7.5)
```

## Polymorphlolgy plotting function simple data
```{r}
tree_plot <- plot_tricontexts(snps[CULTURETYPE=="tree",.(UNIQUE, TRICONTEXT)]$TRICONTEXT, full=F, trimer_freq = freq_trimer, x_text = F)
tree_plot$plot +
  theme(plot.margin = ggplot2::margin(t = 0, r = 0.01, b = 0, l = 0.01, unit = "cm"))
ggsave("./Figures/Tricontext/tree_tricontext_simple.pdf", height = .9, width = 3.7)
shoot_plot <- plot_tricontexts(snps[CULTURETYPE=="shoot",.(UNIQUE, TRICONTEXT)]$TRICONTEXT, full=F, trimer_freq = freq_trimer, x_text = F)
shoot_plot$plot +
  theme(plot.margin = ggplot2::margin(t = 0, r = 0.01, b = 0, l = 0.01, unit = "cm"))
ggsave("./Figures/Tricontext/shoot_tricontext_simple.pdf", height =.9, width = 3.7)
embryo_plot <- plot_tricontexts(snps[CULTURETYPE=="embryo",.(UNIQUE, TRICONTEXT)]$TRICONTEXT, full=F, trimer_freq = freq_trimer,  x_text = T)
embryo_plot$plot +
  theme(plot.margin = ggplot2::margin(t = 0, r = 0.01, b = 0, l = 0.01, unit = "cm"))
ggsave("./Figures/Tricontext/embryo_tricontext_simple.pdf", height = 1.75, width = 7.4)
```

# Plotting all individuals in tricontext
## Colors, levels, labels
```{r}
mut_colors<-c(`C>A`="#03bdee",`C>G`="#000000",`C>T`="#e52a25",`T>A`="#cdc9ca",`T>C`="#a3ce62",`T>G`="#ecc6c5")

mut_colors2 <- c(`C>A`= "white",`C>G`="white",`C>T`="white",`T>A`="white",`T>C`="white",`T>G`="white",`C>A`="#03bdee",`C>G`="#000000",`C>T`="#e52a25",`T>A`="#cdc9ca",`T>C`="#a3ce62",`T>G`="#ecc6c5")

mut_fills_list <- as.list(mut_colors2)
mut_colors_list <- list(element_rect(color = "white"), element_rect(color = "white"), element_rect(color = "white"), element_rect(color = "white"), element_rect(color = "white"), element_rect(color = "white"),
                         element_rect(color = "white"), element_rect(color = "white"), element_rect(color = "white"), element_rect(color = "white"), element_rect(color = "white"), element_rect(color = "white"))

plot_levels <- c("tree2", "ref_chandler", "cr22", "cr13", "cr10", "cr85",
                 "cr2_11A", "cr2_13A", "cr2_12A", "cr21_1", "cr21_2",
                 "cr2_18A", "cr2_15A", "cr2_15A1", "cr2_hifi", "cr2_16A",
                 "cr2_18A1", "cr2_17A1", "cr2_17A2", "cr2_16A1", "cr2_16A2")

new_labels = c(
  "ref_chandler" = "Reference", "tree2" = "Tree",
  "cr2_hifi" = "E HiFi", "cr21_1" = "E 11", "cr21_2" = "E 10",
  "cr2_11A" = "E 14", "cr2_12A" = "E 12", "cr2_13A" = "E 13",
  "cr2_15A" = "E 8", "cr2_15A1" = "E 7", "cr2_16A" = "E 6",
  "cr2_16A1" = "E 2", "cr2_16A2" = "E 1", "cr2_17A1" = "E 4",
  "cr2_17A2" = "E 3", "cr2_18A" = "E 9", "cr2_18A1" = "E 5",
  "cr85" = "S 1985", "cr10" = "S 2010",
  "cr13" = "S 2013", "cr22" = "S 2022"
)
```

## Prepare data for plotting
```{r}
# Tricontext plots plotting
embryo_trimers <- data.table(table(SOURCE = snps$SOURCE, TRICONTEXT = snps$TRICONTEXT, MUT=snps$MUT, CULTURETYPE=snps$CULTURETYPE))[N > 0]
embryo_trimers$TRI <- substr(embryo_trimers$TRICONTEXT, 1, 3)
embryo_trimers$genome <- freq_trimer$N[match(embryo_trimers$TRI, freq_trimer$TRI)]

embryo_trimers[, genomeprop := N/(genome)]
embryo_trimers[, Nprop := genomeprop/sum(genomeprop), by= SOURCE]


embryo_trimers$SOURCE <- factor(embryo_trimers$SOURCE, levels=plot_levels)

## Plots no context
embryo_monomers<-data.table(table(SOURCE = snps$SOURCE, MUT = snps$MUT))[N>0]
embryo_monomers[,Nprop := N/sum(N), by = SOURCE]

embryo_monomers$SOURCE<-factor(embryo_monomers$SOURCE, levels = plot_levels)
```

## Create colored x axis
```{r}
embryo_trimers[, COLOR := fifelse(MUT %in% "C>A", "#03bdee", NA)]
embryo_trimers[, COLOR := fifelse(MUT %in% "C>G", "#000000", COLOR)]
embryo_trimers[, COLOR := fifelse(MUT %in% "C>T", "#e52a25", COLOR)]
embryo_trimers[, COLOR := fifelse(MUT %in% "T>A", "#cdc9ca", COLOR)]
embryo_trimers[, COLOR := fifelse(MUT %in% "T>C", "#a3ce62", COLOR)]
embryo_trimers[, COLOR := fifelse(MUT %in% "T>G", "#ecc6c5", COLOR)]

embryo_trimers[,TRI_COLOR := paste(substr(TRI, 1, 1), "<span style='color: ", COLOR, ";'>", substr(TRI, 2, 2), "</span>", substr(TRI, 3, 3), sep = "")]
```

## Create new column for plotting
This is a work around for labeling that is only for plotting
```{r}
embryo_trimers$MUT2 <- embryo_trimers$MUT
```


## Plot
```{r}
# Tricontext
ggplot(embryo_trimers, aes(x = TRI_COLOR, y = SOURCE, fill = MUT, size = genomeprop, alpha = genomeprop)) +
  geom_point(shape = 22) +
  facet_grid2(~MUT + MUT2, scales = "free_x",
              strip = strip_themed(
                background_x = elem_list_rect(fill = mut_fills_list, color = mut_colors_list),
                text_x = element_markdown(face = "bold")
              ),
              labeller = labeller(
                MUT = c(
                  `C>A` = "<span style = 'color: #03bdee;'>C>A</span>",
                  `C>G` = "<span style = 'color: #000000;'>C>G</span>",
                  `C>T` = "<span style = 'color: #e52a25;'>C>T</span>",
                  `T>A` = "<span style = 'color: #cdc9ca;'>T>A</span>",
                  `T>C` = "<span style = 'color: #a3ce62;'>T>C</span>",
                  `T>G` = "<span style = 'color: #ecc6c5;'>T>G</span>"
                ),
                MUT2 = c(
                  `C>A` = "<span style = 'color: #03bdee;'>C>A</span>",
                  `C>G` = "<span style = 'color: #000000;'>C>G</span>",
                  `C>T` = "<span style = 'color: #e52a25;'>C>T</span>",
                  `T>A` = "<span style = 'color: #cdc9ca;'>T>A</span>",
                  `T>C` = "<span style = 'color: #a3ce62;'>T>C</span>",
                  `T>G` = "<span style = 'color: #ecc6c5;'>T>G</span>"
                )
              )
  ) +
  scale_size_continuous(range = c(0.25, 3)) +
  scale_fill_manual(values = mut_colors) +
  scale_y_discrete(labels = new_labels) +
  theme_classic(base_size = 6) +
  theme(
    legend.position = "none",
    axis.title.x = element_blank(),
    axis.text.x = element_markdown(angle = 90, hjust = 1),
    axis.line.x = element_blank(),
    axis.ticks.x = element_blank(),
    axis.title.y = element_blank(),
    axis.text.y = element_blank(),
    axis.line.y = element_blank(),
    axis.ticks.y = element_blank(),
    panel.spacing.y = unit(0.05, "cm"),
    panel.spacing.x = unit(0.05, "cm"),
    panel.background = element_rect(fill = NA, color = NA),
    plot.background = element_rect(fill = NA, color = NA),
    plot.margin = ggplot2::margin(t = 0, r = 0.05, b = 0, l = 0, unit = "cm")
  ) 
ggsave("./Figures/Tricontext/sample_individual_tricontext.pdf", height = 2.66, width = 6.525)

# Simple no context
ggplot(embryo_monomers, aes(x = MUT, y = SOURCE, fill = MUT, size = N)) +
  geom_point(shape = 22) +
  theme_minimal(base_size = 6) +
  theme(legend.position = "top")+
  scale_size_continuous(range = c(0.25,5))+
  facet_grid(~MUT, scales = "free_x")+
  scale_fill_manual(values = mut_colors, guide = "none")
```

